{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Woodchuck Deltavirus (WDV)\n",
    "```\n",
    "Lead     : ababaian\n",
    "Issue    : x\n",
    "start    : 2020 07 06\n",
    "complete : YYYY MM DD\n",
    "files    : ~/serratus/notebook/200702_ab/\n",
    "s3 files : s3://serratus-public/notebook/200706_ab/\n",
    "```\n",
    "\n",
    "## Introduction\n",
    "\n",
    "While coronaviruses are well characterized, other viral families offer higher probability for identifying novel strains. In a personal correspondance with D. Streiker following a [comment on the Serratus virological post](https://virological.org/t/serratus-the-ultra-deep-search-to-discover-novel-coronaviruses/516) and in extension to the recent pre-print from [Bergner et al.](https://www.biorxiv.org/content/10.1101/2020.06.17.156745v2) we were interested if we could identify novel deltaviruses.\n",
    "\n",
    "A recent develpoment to `Serratus` is the use of the `blastx` algorithm for a search against a protein reference as opposed to nucleotide `bowtie2`. This in theory, and in practice allows for the detection of deeper homologies of distant strain/species. The pilot search included viral UniProt sequences and all known deltavirus antigen proteins, as well as two representative dAg from vampire bats provided by Streiker and collegues, and dAg from the bambi sample. 100K sequencing libraries from non-human mammals were selected as the pilot search space [mamm_SraRunInfo.csv](SRA-queries).\n",
    "\n",
    "With this we were able to identify two novel deltaviruses; Murray (`SRR2136906`) in _Marmota monax_ (Eastern woodchuck) and Lassie (`SRR7910143`) in _Peropteryx macrotis_ (Lesser dog-like bat).\n",
    "\n",
    "### Objectives\n",
    "< What are the objectives / deliverable items from this experiment >"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Murray Virus / Woodchuck Deltavirus (WDV)\n",
    "\n",
    "The murray virus was assembled from `SRR2136906` to yield [a 1753 nt](200706_ab/murray.fa) circular genome. The virus has the characteristic 'bar' folding of a deltavirus.\n",
    "\n",
    "![Figure 2: Mfold base-pairing of Murray Virus](200706_ab/murray_mfold.png)\n",
    "\n",
    "The groundhog samples from this study by [Fletcher et al., 2015](https://journals.plos.org/plospathogens/article?id=10.1371/journal.ppat.1005103) on the effects of IFNa on experimental infection with Woodchuck Hepatitis Virus (WHV). `SRR2136906` and `SRR2136907` samples with high titre Murray are replicate liver samples from the placebo-treated (WHV, no IFNa) animal designated `1008` taken on week 1 of the study. The same animal `1008` shows trace amounts of Murray virus at weeks 4, 8, 12, and 16 and no deltavirus reads on weeks 3, 20 and 24 suggesting waning viral levels following WHV challenge.  WHV infection is confirmed in the [summary](https://lovelywater.s3.amazonaws.com/summary/SRR2136906.summary) file for `1008`. Within this study the placebo treated animal `1016` shows trace Murray virus on week 24.\n",
    "\n",
    "Similiar to how human deltavirus (HDV) is a satellite virus that requires hepetitis virus B (HBV) for it's replication, it is likely that Murray is a satellite virus requiring WHV for it's replication, although this will have to be tested experimentally to be validated. The woodchuck is a model organism for hepatocellular carcionoma (HCC) caused by WHV, the inconsistent precence of Murray virus and it's possible effects on HCC development in this model system may be a confounding variable complicating studies using this model system.\n",
    "\n",
    "- [SRR2136906.summary](https://lovelywater.s3.amazonaws.com/summary/SRR2136906.summary) \n",
    "- [SRR2136906.bam alignment](https://lovelywater.s3.amazonaws.com/summary/SRR2136906.bam)\n",
    "- [SRR2136906.psummary](https://serratus-public.s3.amazonaws.com/out/200628_pmamm/psummary/SRR2136906.psummary)\n",
    "\n",
    "See also:\n",
    "- [Characterization of newt, toad, termite deltavirus](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6628682/#!po=62.5000)\n",
    "- [Detailed review article on HDV infection](https://www.mdpi.com/1999-4915/12/6/648/htm)\n",
    "- [Ponzetto et al., 1984. Human deltavirus infection worsens WHV pathology in groundhogs](https://www.pnas.org/content/81/7/2208.short)\n",
    "- [Human Deltavirus Wiki](https://en.wikipedia.org/wiki/Hepatitis_D)\n",
    "- [Woodchuck HDV adaptation mutations](https://jvi.asm.org/content/80/13/6469.short)\n",
    "- [Woodchuck RNA-seq](https://journals.plos.org/plospathogens/article/authors?id=10.1371/journal.ppat.1005103)\n",
    "- [Woodchuck Genome](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6893209/)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    ">murray_deltavirus_1712nt_cov_254\n",
    "TCGGGTAAGTTTGTAAATTTATTTGGTAATTATCTGAGGGGGGAGGGAGA\n",
    "AGGAATAGGGATCAAATTTTCAATTAGGGGAACTCCTGCTTTCCTCTAAT\n",
    "GTCCAGTCCTTCCCCTCTTCGGGAGTATGGACTTTCGGGGACCCCTTGAA\n",
    "GATTGGAGACGAAGCCTCCTCCGGGCGCTCCTCTGGGACCCCCTCCTCCG\n",
    "ACGTTCACATCCCCAGGTCCGTGGTCTCCCAATCTTCTTTCTCTAGCCTC\n",
    "ATCCGGACCGGTCAACCTCCGCAACTCCTCTTCCTCCTCCTGGGGCAGTT\n",
    "TCTTTCCATGTCCCTCCAGCTGTTTCTTCTTGTTCTCCAGCGCCTTCCTT\n",
    "CTGCGGTGGTCCCGCTTCTCCTCCTCCGAGAATCTCGGCCCTTGGGGTCC\n",
    "GCCGCTAGATCCTCCATCGACCTCCATGCTTTCTCCTCGTGGTCTCTTGG\n",
    "ACGGGGAGCTCTGCTCCTCTCCGCCTCTCTTTCTCACCATTCCCAGGATG\n",
    "TTCCCCAGCCAGGGATTCCTCTCCTCAAGCTTCTTTATTTTCTTGTTGAG\n",
    "GCGACGGAGTTCCTCCTCGAGCTCCTCCTTTCTCTCTCTTCCTTTGACCC\n",
    "ATTGTCTTAGAGTCTCCTCCCTCCCCTTCGAGTTTCCTTGCTTAGGATTC\n",
    "TCCATCGAGGAAATCCGGGATAAGGTCCCAAAAATTCTCTTACTTTCGGG\n",
    "AAGAAAGGAGGAATGTTGGGCACCCTCCCGGCGCCCCCGATGGATCGTTG\n",
    "CCTAGCATAGGGCTCGAAGAACGGAGAGGAGGAGCCCCAGAAATGGAAAA\n",
    "AAGGTCAAAGAGTGGAGGAACCCTCGAGAGGTAAACCCCCAATCCCTCAT\n",
    "CAATATGAAGAAGAGAGGGGGTTTCCCGTATGAGCTGGTGAAACTCCGGA\n",
    "ACGGATTTGAAAGTTAGAACGCGTGAGATAGAATAGAGCTTAGCCCCTCT\n",
    "ATTGCCAATTGACCGCGGAGCCAAGCAAGTGGAGTCCGGAGGGTCCCATC\n",
    "GGCGGACATTCCAGCGCCGGAGAACTCCGGGGGGAGAATGACGGAAAACC\n",
    "CCTCCAGAGGACGCTATCAGCGAACAGAAGTAGCTGGGAGGGGACTGCGA\n",
    "AAAGCTAGCCGCTAGGGTTGGATGCGAGCGAGATGGGGGAGACCGCCCGA\n",
    "GTGGGCCGGAAAGAAATGGTAGAGTATAACGTGACCGGATGTTCCGGAAG\n",
    "GAGAAGAGGACCAGTAAGGCTTATCCCGGAGTCCTCGGCAACGCCTCCTC\n",
    "TCCCTCAGGGCCCCCAGGCCATAGTTCATGAATTGGAGAGGAAAGGAATT\n",
    "GAGTAGGAAAGCTAAGGAAAGTACCCCAATTAAAAGAAAATGATCACTTT\n",
    "TTCCTTACATACCCACCCCCTCAAGAAACTATGCCAAATAAGATTTTACA\n",
    "ACTTCTTACATCGAAGTCTGAAATCTTATCTTTCTTACCTGATGGTTGGC\n",
    "ATGGCCCCAGCCTCCTCGCGGCGCCAACCGGGCAACGTTTCCGAAAGAGT\n",
    "CCCTCTTTCGAGAACGGCTAATGGGAGTCAGGAATTCTCGTAGACTTGCT\n",
    "AGTAGTTAGTCCTCTCTGAGGATATCTCGCAAGTCCGAGCAACCCTGGAT\n",
    "CTCCCTTAGCCTTCCGAAAGGAGGGTTCTCTTTTACGGAAGCCCAGGCTG\n",
    "GACCACGGGGAGGTGGTAGATCCATGCCAGCCCGAAGAAGGAAAGAAGAG\n",
    "TCAGACACAGAT\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Viral Origins\n",
    "\n",
    "Preliminary phylogeny was performed. This supports that Murray is not of HDV origin but more close to DrDVa and Bambi (Deer deltavirus). The alternative hypothesis was brought up by D. Stieker via personal comms but this is not neccesarily supported in the data\n",
    "\n",
    "![dAg Phylogeny](200706_ab/deltavirus_phylogram_v3.png)\n",
    "\n",
    "> \"DS:\"\n",
    "> Second, groundhog/woodchuck have been used as an experimental model for HDV and there has been some effort to make a woodchuck adapted deltavirus (https://jvi.asm.org/content/80/13/6469.short) - is what you found something entirely new or the lab-adapted strain? \n",
    "\n",
    "\n",
    "> \"AB:\"\n",
    ">  You did find an interesting paper, Murray is quite a bit different then HDV3 but this does raise the interesing zoonosis hypothesis, maybe even lab-borne. That paper you sent is an interesting read, the positions they identify as associated with human to groundhog host switch, E16D, Q20D and especially I42V are not found in Murray, but E16D and Q20D are found in Bambi and DrDVa. Fun exercise, takes me back to my undergrad days of playing with sequences at this level.\n",
    "\n",
    "![Adaptations WDV](200706_ab/hAG_adaptations.png)\n",
    "\n",
    "\n",
    "To address this in an independent measurement of the dAg protein, I decided to build a phylogeny based on the ribozyme sequences. This is a very rough alignment based on ClustalO standard output. I've since refined the alignment to correspond with \n",
    "\n",
    "![Ribozymes WDV Phylogeny](200706_ab/wdv_ribozymes_phylogeny.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Woodchuck RNA-seq analysis\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "#### SRA data list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hidden": true
   },
   "source": [
    "SRA accession list `sra.txt`\n",
    "\n",
    "```\n",
    "SRR2136852\n",
    "SRR2136853\n",
    "SRR2136854\n",
    "SRR2136855\n",
    "SRR2136856\n",
    "SRR2136857\n",
    "SRR2136858\n",
    "SRR2136859\n",
    "SRR2136860\n",
    "SRR2136861\n",
    "SRR2136862\n",
    "SRR2136863\n",
    "SRR2136864\n",
    "SRR2136865\n",
    "SRR2136866\n",
    "SRR2136867\n",
    "SRR2136868\n",
    "SRR2136869\n",
    "SRR2136870\n",
    "SRR2136871\n",
    "SRR2136872\n",
    "SRR2136873\n",
    "SRR2136874\n",
    "SRR2136875\n",
    "SRR2136876\n",
    "SRR2136877\n",
    "SRR2136878\n",
    "SRR2136879\n",
    "SRR2136880\n",
    "SRR2136881\n",
    "SRR2136882\n",
    "SRR2136883\n",
    "SRR2136884\n",
    "SRR2136885\n",
    "SRR2136886\n",
    "SRR2136887\n",
    "SRR2136888\n",
    "SRR2136889\n",
    "SRR2136890\n",
    "SRR2136891\n",
    "SRR2136892\n",
    "SRR2136893\n",
    "SRR2136894\n",
    "SRR2136895\n",
    "SRR2136896\n",
    "SRR2136897\n",
    "SRR2136898\n",
    "SRR2136899\n",
    "SRR2136900\n",
    "SRR2136901\n",
    "SRR2136902\n",
    "SRR2136903\n",
    "SRR2136904\n",
    "SRR2136905\n",
    "SRR2136906\n",
    "SRR2136907\n",
    "SRR2136908\n",
    "SRR2136909\n",
    "SRR2136910\n",
    "SRR2136911\n",
    "SRR2136912\n",
    "SRR2136913\n",
    "SRR2136914\n",
    "SRR2136915\n",
    "SRR2136916\n",
    "SRR2136917\n",
    "SRR2136918\n",
    "SRR2136919\n",
    "SRR2136920\n",
    "SRR2136921\n",
    "SRR2136922\n",
    "SRR2136923\n",
    "SRR2136924\n",
    "SRR2136925\n",
    "SRR2136926\n",
    "SRR2136927\n",
    "SRR2136928\n",
    "SRR2136929\n",
    "SRR2136930\n",
    "SRR2136931\n",
    "SRR2136932\n",
    "SRR2136933\n",
    "SRR2136934\n",
    "SRR2136935\n",
    "SRR2136936\n",
    "SRR2136937\n",
    "SRR2136938\n",
    "SRR2136939\n",
    "SRR2136940\n",
    "SRR2136941\n",
    "SRR2136942\n",
    "SRR2136943\n",
    "SRR2136944\n",
    "SRR2136945\n",
    "SRR2136946\n",
    "SRR2136947\n",
    "SRR2136948\n",
    "SRR2136949\n",
    "SRR2136950\n",
    "SRR2136951\n",
    "SRR2136952\n",
    "SRR2136953\n",
    "SRR2136954\n",
    "SRR2136955\n",
    "SRR2136956\n",
    "SRR2136957\n",
    "SRR2136958\n",
    "SRR2136959\n",
    "SRR2136960\n",
    "SRR2136961\n",
    "SRR2136962\n",
    "SRR2136963\n",
    "SRR2136964\n",
    "SRR2136965\n",
    "SRR2136966\n",
    "SRR2136967\n",
    "SRR2136968\n",
    "SRR2136969\n",
    "SRR2136970\n",
    "SRR2136971\n",
    "SRR2136972\n",
    "SRR2136973\n",
    "SRR2136974\n",
    "SRR2136975\n",
    "SRR2136976\n",
    "SRR2136977\n",
    "SRR2136978\n",
    "SRR2136979\n",
    "SRR2136980\n",
    "SRR2136981\n",
    "SRR2136982\n",
    "SRR2136983\n",
    "SRR2136984\n",
    "SRR2136985\n",
    "SRR2136986\n",
    "SRR2136987\n",
    "SRR2136988\n",
    "SRR2136989\n",
    "SRR2136990\n",
    "SRR2136991\n",
    "SRR2136992\n",
    "SRR2136993\n",
    "SRR2136994\n",
    "SRR2136995\n",
    "SRR2136996\n",
    "SRR2136997\n",
    "SRR2136998\n",
    "SRR2136999\n",
    "SRR2137000\n",
    "SRR2137001\n",
    "SRR2137002\n",
    "SRR2137003\n",
    "SRR2137004\n",
    "SRR2137005\n",
    "SRR2137006\n",
    "SRR2137007\n",
    "SRR2137008\n",
    "SRR2137009\n",
    "SRR2137010\n",
    "SRR2137011\n",
    "SRR2137012\n",
    "SRR2137013\n",
    "SRR2137014\n",
    "SRR2137015\n",
    "SRR2137016\n",
    "SRR2137017\n",
    "SRR2137018\n",
    "SRR2137019\n",
    "SRR2137020\n",
    "SRR2137021\n",
    "SRR2137022\n",
    "SRR2137023\n",
    "SRR2137024\n",
    "SRR2137025\n",
    "SRR2137026\n",
    "SRR2137027\n",
    "SRR2137028\n",
    "SRR2137029\n",
    "SRR2137030\n",
    "SRR2137031\n",
    "SRR2137032\n",
    "SRR2137033\n",
    "SRR2137034\n",
    "SRR2137035\n",
    "SRR2137036\n",
    "SRR2137037\n",
    "SRR2137038\n",
    "SRR2137039\n",
    "SRR2137040\n",
    "SRR2137041\n",
    "SRR2137042\n",
    "SRR2137043\n",
    "SRR2137044\n",
    "SRR2137045\n",
    "SRR2137046\n",
    "SRR2137047\n",
    "SRR2137048\n",
    "SRR2137049\n",
    "SRR2137050\n",
    "SRR2137051\n",
    "SRR2137052\n",
    "SRR2137053\n",
    "SRR2137054\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Workflow"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "# Fire up container\n",
    "sudo yum install -y docker\n",
    "sudo yum install -y git\n",
    "sudo service docker start\n",
    "\n",
    "git clone https://github.com/ababaian/serratus.git; cd serratus/containers\n",
    "./build_containers.sh\n",
    "\n",
    "sudo docker run --rm --entrypoint /bin/bash -it serratus-dl:latest\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "#### CONTAINER WORKFLOW\n",
    "yum install -y vim less gzip wget unzip tar\n",
    "\n",
    "mkdir align; cd align\n",
    "\n",
    "# Install HISAT2\n",
    "wget ftp://ftp.ccb.jhu.edu/pub/infphilo/hisat2/downloads/hisat2-2.1.0-Linux_x86_64.zip\n",
    "unzip hisat2-2.1.0-Linux_x86_64.zip\n",
    "mv hisat2-2.1.0/* /usr/bin/ \n",
    "\n",
    "# Install samtools\n",
    "SAMTOOLSVERSION='1.10'\n",
    "wget -O /samtools-${SAMTOOLSVERSION}.tar.bz2 \\\n",
    "  https://github.com/samtools/samtools/releases/download/${SAMTOOLSVERSION}/samtools-${SAMTOOLSVERSION}.tar.bz2 &&\\\n",
    "  tar xvjf /samtools-${SAMTOOLSVERSION}.tar.bz2 && rm /samtools-${SAMTOOLSVERSION}.tar.bz2\n",
    "\n",
    "cd samtools-${SAMTOOLSVERSION} && make && make install\n",
    "\n",
    "cd .. && rm -rf samtools-${SAMTOOLSVERSION}\n",
    "\n",
    "#download monax genome\n",
    "wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/901/343/595/GCA_901343595.1_MONAX5/GCA_901343595.1_MONAX5_genomic.fna.gz\n",
    "#should work\n",
    "gzip -d GCA_901343595.1_MONAX5_genomic.fna.gz\n",
    "mv GCA_901343595.1_MONAX5_genomic.fna monax5.fa\n",
    "\n",
    "# Download MHV genome NC_004107.1\n",
    "# https://www.ncbi.nlm.nih.gov/nuccore/22256030\n",
    "# vim mhv.fa\n",
    "\n",
    "# Download MDV genome (murray)\n",
    "# https://raw.githubusercontent.com/wiki/ababaian/serratus/assets/murray.fa\n",
    "# vim mdv.fa\n",
    "\n",
    "# Make reference genome\n",
    "cat monax5.fa mhv.fa mdv.fa > monax5v.fa\n",
    "\n",
    "# Index reference\n",
    "hisat2-build --seed 1337 -p 36 monax5v.fa monax5v\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "#read the sras file into array\n",
    "# vim sra.txt\n",
    "mapfile -t SRARRAY < sra.txt\n",
    "\n",
    "# Upload to S3\n",
    "S3_PATH=s3://serratus-public/notebook/200702_ab\n",
    "aws s3 sync ./ \"$S3_PATH\"/\n",
    "\n",
    "# Align each SRA\n",
    "for SRA in \"${SRARRAY[@]}\"\n",
    "do\n",
    "   mkdir -p $SRA\n",
    "   cd $SRA\n",
    "\n",
    "   RGLB=$SRA\n",
    "   echo \"Aligning $SRA\"\n",
    "\n",
    "   #hisat2 command - random things for rg-id, etc\n",
    "   hisat2 --rg-id \"$SRA\" --rg LB:\"$SRA\" \\\n",
    "          --rg SM:\"woodchuck\" --rg PL:\"ILLUMINA\" \\\n",
    "          --rg PU:\"xxx\" -p 35 -x ../monax5v \\\n",
    "          --sra-acc $SRA | samtools view -bS - \\\n",
    "          > aligned_unsorted.bam\n",
    "\n",
    "   #sort bam file\n",
    "   samtools sort aligned_unsorted.bam > \"$SRA\".monax5v.bam\n",
    "   samtools index \"$SRA\".monax5v.bam\n",
    "   rm aligned_unsorted.bam\n",
    "\n",
    "   samtools flagstat \"$SRA\".monax5v.bam > \"$SRA\".flagstat\n",
    "\n",
    "   # Upload to S3\n",
    "   aws s3 sync ./ \"$S3_PATH\"/\"$SRA\"/\n",
    "   cd ..\n",
    "done\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Correct error in first version\n",
    "# re-do flagstats\n",
    "for SRA in $(ls | grep \"SRR\"); do\n",
    "  cd $SRA\n",
    "  samtools flagstat -@ 35 \"$SRA\".monax5v.bam > \"$SRA\".flagstat\n",
    "  cd ..\n",
    "done "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Correct error in first version\n",
    "# re-do flagstats\n",
    "for SRA in $(ls | grep \"SRR\"); do\n",
    "  cd $SRA\n",
    "  samtools idxstats -@ 30 \"$SRA\".monax5v.bam > \"$SRA\".idxstat\n",
    "  cd ..\n",
    "done \n",
    "\n",
    "cp */*.idxstat idx/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Uploaded to "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Results & Discussion\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
