{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# RdRp0 pan-proteome v0\n",
    "```\n",
    "Lead     : ababaian\n",
    "Issue    : \n",
    "start    : 2020 12 10\n",
    "complete : 2020 12 13\n",
    "revision : 2020 12 15\n",
    "files    : ~/serratus/notebook/201210_rdrp0/\n",
    "s3 files : s3://serratus-public/notebook/201210_rdrp0/\n",
    "```\n",
    "\n",
    "## Introduction\n",
    "\n",
    "We've been considering other ways to 'dive' into the SRA to yield meaningful, interpretable results. An idrea which is recurring is to focus on a gene-family/domain that we would like to characterize exhaustively.\n",
    "\n",
    "The prime candidate is viral RNA-dependent RNA-polymerase or `RdRp`. This is slowly-evolving and central reference gene for the identification and classification of RNA viruses.\n",
    "\n",
    "It is a daunting task to isolate all known RdRp and categorize them into a meaningful system, this is a first approximation of that goal putting together the components to do so.\n",
    "\n",
    "The ideal end-goal will be to create a hierarchly/taxonomically nested set of RdRp protein sequences at various cut-off thresholds.\n",
    "\n",
    "- rdrp1_99: all unique RdRp sequences\n",
    "- rdrp1:  Species-Approximate. 90% identity clusters\n",
    "- rdrp1_75:  Genus-Approximate. 75% identity clusters\n",
    "- rdrp1_45:  Family-Approximate. 45% identity clusters\n",
    "\n",
    "\n",
    "### Key Literature\n",
    "\n",
    "- [Wolf20: Doubling RNA viruses](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7508674/)\n",
    "- [Wolf18: RdRp evo/origin](https://pubmed.ncbi.nlm.nih.gov/30482837/)\n",
    "- [Venk18: RdRp evo/origin](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5850383/)\n",
    "- [Zhange: Expanding RNA virome (review)](https://pubmed.ncbi.nlm.nih.gov/31100994/)\n",
    "- [Wu: Structural overview of RdRp](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4490480/)\n",
    "- [Velthuis: Features of RdRp](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4207942/#Sec5)\n",
    "- [Bruenn: Core motif-set of RdRp](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC152793/)\n",
    "- [Jia: Diversity of motif-synteny](https://www.frontiersin.org/articles/10.3389/fmicb.2019.01945/full)\n",
    "\n",
    "### Objectives\n",
    "- Compile the materials neccesary for a comprehensive RdRp-ome\n",
    "- Create the `rdrp0.fa` reference pan-proteome to run a pilot serratus run and see what results would look like\n",
    "\n",
    "### END PRODUCT: `rdrp1_DELTA.fa `\n",
    "\n",
    "- s3_link: `s3://serratus-public/notebook/201226_rdrp0/rdrp1_DELTA.fa`\n",
    "- html_link: `https://serratus-public.s3.amazonaws.com/notebook/201226_rdrp0/rdrp1_DELTA.fa`\n",
    "- Sequences: `14659`\n",
    "- Lines: `128997`\n",
    "- md5sum: `c5301d494edb4a065f16d701fa5c08ce`\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sat Dec 26 22:30:20 PST 2020\n",
      "340d2fd52e816b20e5c88b7fb08e68b25066a12b\n"
     ]
    }
   ],
   "source": [
    "# Serratus commit version\n",
    "SERRATUS=\"/home/artem/serratus\"\n",
    "cd $SERRATUS\n",
    "\n",
    "# Create local run directory\n",
    "WORK=\"$SERRATUS/notebook/201210_rdrp0\"\n",
    "mkdir -p $WORK; cd $WORK\n",
    "\n",
    "# S3 notebook path\n",
    "S3_WORK='s3://serratus-public/notebook/201210_rdrp0/'\n",
    "\n",
    "# date and version\n",
    "date\n",
    "git rev-parse HEAD # commit version"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GenBank Virome\n",
    "\n",
    "The master corpus for all viral sequences\n",
    "\n",
    "See also: [Genbank FTP README](https://ftp.ncbi.nlm.nih.gov/genbank/release.notes/gb240.release.notes)\n",
    "\n",
    "### Nucleotide Sequences\n",
    "- Query: `txid10239[Organism:exp]` # all viruses\n",
    "- Date: `201205`\n",
    "- Results: `3 535 357` sequences\n",
    "- File : `ntViro_gb201005.fa`\n",
    "\n",
    "### CDS Sequences (INCOMPLETE)\n",
    "- Query: `txid2552587[Organism:exp]` # all RNA virus CDS\n",
    "- Date: `201205` # error in this version\n",
    "- Results: `2 825 230` sequences\n",
    "- File : `cdsViro_gb201005.fa`\n",
    "\n",
    "### Protein Sequences\n",
    "- Query: # GenBank Protein v240 [see README](ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/README.asn1.protein_fasta)\n",
    "- Date: `201220` # error in this version\n",
    "- Results: `5 095 919` aa sequences\n",
    "- File : `gbViro240.fa`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Download GenBank v240 Viral Protein Sequences\n",
    "# (RAN ON EC2)\n",
    "#\n",
    "\n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl1.fsa_aa.gz \n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl2.fsa_aa.gz \n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl3.fsa_aa.gz \n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl4.fsa_aa.gz \n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl5.fsa_aa.gz \n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl6.fsa_aa.gz \n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl7.fsa_aa.gz \n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl8.fsa_aa.gz \n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl9.fsa_aa.gz \n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl10.fsa_aa.gz \n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl11.fsa_aa.gz \n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl12.fsa_aa.gz \n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl13.fsa_aa.gz \n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl14.fsa_aa.gz \n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl15.fsa_aa.gz \n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl16.fsa_aa.gz \n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl17.fsa_aa.gz \n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl18.fsa_aa.gz \n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl19.fsa_aa.gz \n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl20.fsa_aa.gz \n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl21.fsa_aa.gz \n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl22.fsa_aa.gz \n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl23.fsa_aa.gz \n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl24.fsa_aa.gz \n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl25.fsa_aa.gz \n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl26.fsa_aa.gz \n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl27.fsa_aa.gz \n",
    "\n",
    "gzip -d *\n",
    "cat * > aaViro_gb240.fa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# EC2\n",
    "AA='gbViro240.fa'\n",
    "grep \">\" $AA | wc -l\n",
    "# 5095919\n",
    "md5sum $AA\n",
    "# 078505a82990197ccee5ebdbdf48ad6f  gbViro240.fa\n",
    "md5sum $AA > $AA.md5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# December 23rd Release 241 was available\n",
    "mkdir gb241; cd gb241\n",
    "\n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl1.fsa_aa.gz \n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl2.fsa_aa.gz \n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl3.fsa_aa.gz \n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl4.fsa_aa.gz \n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl5.fsa_aa.gz \n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl6.fsa_aa.gz \n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl7.fsa_aa.gz \n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl8.fsa_aa.gz \n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl9.fsa_aa.gz \n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl10.fsa_aa.gz \n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl11.fsa_aa.gz \n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl12.fsa_aa.gz \n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl13.fsa_aa.gz \n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl14.fsa_aa.gz \n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl15.fsa_aa.gz \n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl16.fsa_aa.gz \n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl17.fsa_aa.gz \n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl18.fsa_aa.gz \n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl19.fsa_aa.gz \n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl20.fsa_aa.gz \n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl21.fsa_aa.gz \n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl22.fsa_aa.gz \n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl23.fsa_aa.gz \n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl24.fsa_aa.gz \n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl25.fsa_aa.gz \n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl26.fsa_aa.gz \n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl27.fsa_aa.gz \n",
    "wget ftp://ftp.ncbi.nih.gov/ncbi-asn1/protein_fasta/gbvrl28.fsa_aa.gz \n",
    "\n",
    "gzip -d *.gz\n",
    "\n",
    "# Viral GenBank version 241\n",
    "cat * > vgb241.fa\n",
    "\n",
    "# md5: e0e2f609074182de74745eafc8acd976  vgb241.fa\n",
    "md5sum vgb241.fa > vgb241.fa.md5\n",
    "\n",
    "# faidx: 5 267 608 entries\n",
    "seqkit faidx -f vgb241.fa; mv vgb241.fa.seqkit.fai vgb241.fa.fai\n",
    "rm *.fsa_aa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3535357\n",
      "9102eceda85185cfb124023dbf129621  ntViro_gb201205.fa\n",
      "2825230\n",
      "81946a20fc18b24eb6b49541d41b8dd0  cdsViro_gb201205.fa\n",
      "2825230\n",
      "81946a20fc18b24eb6b49541d41b8dd0  aaViro_gb201212.fa\n",
      "13870\n",
      "671fb5b3f02fd41457be4d6c7a31a417  gbRdRp_201212.fa\n",
      "13870\n",
      "671fb5b3f02fd41457be4d6c7a31a417  gbRdRp_201212.fa\n"
     ]
    }
   ],
   "source": [
    "cd $WORK\n",
    "NT='ntViro_gb201205.fa'\n",
    "grep \">\" $NT | wc -l\n",
    "md5sum $NT\n",
    "md5sum $NT > $NT.md5\n",
    "\n",
    "CDS='cdsViro_gb201205.fa'\n",
    "grep \">\" $CDS | wc -l\n",
    "md5sum $CDS\n",
    "md5sum $CDS > $CDS.md5\n",
    "\n",
    "AA='aaViro_gb201212.fa'\n",
    "grep \">\" $AA | wc -l\n",
    "md5sum $AA\n",
    "md5sum $AA > $AA.md5\n",
    "\n",
    "# GB RdRp sequences (see below WOLF18)\n",
    "GB='gbRdRp_201212.fa'\n",
    "grep \">\" $GB | wc -l\n",
    "md5sum $GB\n",
    "md5sum $GB > $GB.md5\n",
    "\n",
    "# YA RdRp sequences (see below WOLF20)\n",
    "YA='gbRdRp_201212.fa'\n",
    "grep \">\" $YA | wc -l\n",
    "md5sum $YA\n",
    "md5sum $YA > $YA.md5"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## WOLF18 RdRp\n",
    "\n",
    "FTP Access: `ftp://ftp.ncbi.nlm.nih.gov/pub/wolf/_suppl/rnavir18/`\n",
    "\n",
    "Sequence data: `ftp://ftp.ncbi.nlm.nih.gov/pub/wolf/_suppl/rnavir18/RNAvirome.S2.afa`\n",
    "\n",
    "Saved as: `gb_rdrp.afa`\n",
    "\n",
    "Date Accessed: `201212`\n",
    "\n",
    "![Figure 1](/home/artem/serratus/notebook/201210_rdrp0/wolf18/figure1.png)\n",
    "\n",
    "### Level 1 - Supergroup / Branches\n",
    "\n",
    "The RdRp can be broadly classified into 5 branches which will form the lowest level of the hierarchy: `rdrp1`, `rdrp2`, `rdrp3`, `rdrp4`, `rdrp5`.\n",
    "\n",
    ">Branch 1 consists of leviviruses and their eukaryotic relatives, namely, “mitoviruses,” “narnaviruses,” and “ourmiaviruses” (the latter three terms are placed in quotation marks as our analysis contradicts the current ICTV framework, which classifies mitoviruses and narnaviruses as members of one family, Narnaviridae, and ourmiaviruses as members of a free-floating genus, Ourmiavirus).\n",
    "\n",
    "> Branch 2 (“picornavirus supergroup”) consists of a large assemblage of +RNA viruses of eukaryotes, in particular, those of the orders Picornavirales and Nidovirales; the families Caliciviridae, Potyviridae, Astroviridae, and Solemoviridae, a lineage of dsRNA viruses that includes partitiviruses and picobirnaviruses; and several other, smaller groups of +RNA and dsRNA viruses.\n",
    "\n",
    "> Branch 3 consists of a distinct subset of +RNA viruses, including the “alphavirus supergroup” along with the “flavivirus supergroup,” nodaviruses, and tombusviruses; the “statovirus,” “wèivirus,” “yànvirus,” and “zhàovirus” groups; and several additional, smaller groups.\n",
    "\n",
    "> Branch 4 consists of dsRNA viruses, including cystoviruses, reoviruses, and totiviruses and several additional families.\n",
    "\n",
    "> Branch 5 consists of −RNA viruses.\n",
    "\n",
    "Boundary defintions of Branches with relation to RdRp are taken from paper\n",
    "\n",
    "Based on: [Supplementary Data 4](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6282212/bin/mbo006184203sd4.txt)\n",
    "Saved as: `rdrp_representative_branches.tree`\n",
    "\n",
    "### Level 2 - Viral Family\n",
    "\n",
    "Based on: `https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6282212/bin/mbo006184203sd1.xls`\n",
    "Saved as: `wolf18_vlist.xlsx`\n",
    "\n",
    "Spreadsheet includes the fields\n",
    "\n",
    "- RdRp num ID: Ordinal numbering for RdRp\n",
    "- RdRp GenBank Acc: Protein accession ID\n",
    "- NCBI Tax ID: taxid from NCBI\n",
    "- virus name: virus name\n",
    "- taxonomy: taxonomic tree\n",
    "\n",
    "Taxonomy field was parsed to retrieve \"*dae* suffix for \"Family\", relatively appropriate family-name or \"unclassified\" when unavailable. Monkey work.\n",
    "\n",
    "### Level 3 - Sequence/Species\n",
    "\n",
    "Based on: `wolf18_vlist.xlsx`\n",
    "\n",
    "- Virus name field (most shallow taxonomic classification) taken for each record.\n",
    "- GenBank accession taken from each record.\n",
    "\n",
    "### Example RdRp\n",
    "\n",
    "```\n",
    "rdrp5.Hantaviridae.Bowe_virus:AGW23849.1\n",
    "rdrp5.Bunyaviridae.Azagny_virus:AEA42011.1\n",
    "...\n",
    "rdrp2.Coronaviridae.Night_heron_coronavirus_HKU19:YP_005352862.1\n",
    "rdrp2.Coronaviridae.Munia_coronavirus_HKU13_3514:YP_002308505.1\n",
    "rdrp2.Coronaviridae.Wigeon_coronavirus_HKU20:YP_005352870.1\n",
    "rdrp2.Coronaviridae.Feline_infectious_peritonitis_virus:AGZ84535.1\n",
    "rdrp2.Coronaviridae.Lucheng_Rn_rat_coronavirus:YP_009336483.1\n",
    "rdrp2.Coronaviridae.Hipposideros_bat_coronavirus_HKU10:AFU92121.1\n",
    "rdrp2.Coronaviridae.BtMs_AlphaCoV_GS2013:AIA62270.1\n",
    "rdrp2.Coronaviridae.Chaerephon_bat_coronavirus_Kenya_KY41_2006:ADX59465.1\n",
    "rdrp2.Coronaviridae.Porcine_epidemic_diarrhea_virus:AID56804.1\n",
    "rdrp2.Coronaviridae.Bat_coronavirus_CDPHE15_USA_2006:YP_008439200.1\n",
    "rdrp2.Coronaviridae.Anlong_Ms_bat_coronavirus:AID16674.1\n",
    "rdrp2.Coronaviridae.Scotophilus_bat_coronavirus_512:YP_001351683.1\n",
    "rdrp2.Coronaviridae.BtNv_AlphaCoV_SC2013:YP_009201729.1\n",
    "rdrp2.Coronaviridae.Bat_coronavirus_1B:ACA52156.1\n",
    "rdrp2.Coronaviridae.NL63_related_bat_coronavirus:YP_009328933.1\n",
    "rdrp2.Coronaviridae.NL63_related_bat_coronavirus:APD51489.1\n",
    "rdrp2.Coronaviridae.229E_related_bat_coronavirus:ALK43115.1\n",
    "rdrp2.Coronaviridae.Rhinolophus_bat_coronavirus_HKU2:ABQ57223.1\n",
    "rdrp2.Coronaviridae.Wencheng_Sm_shrew_coronavirus:AID16677.1\n",
    "rdrp2.Coronaviridae.Human_coronavirus_HKU1:ABD75543.1\n",
    "rdrp2.Coronaviridae.Betacoronavirus_Erinaceus_VMC_DEU_2012:YP_008719930.1\n",
    "rdrp2.Coronaviridae.Pipistrellus_bat_coronavirus_HKU5:YP_001039961.1\n",
    "rdrp2.Coronaviridae.Rousettus_bat_coronavirus:AOG30811.1\n",
    "rdrp2.Coronaviridae.Rousettus_bat_coronavirus:YP_009273004.1\n",
    "rdrp2.Coronaviridae.Bat_CoV_279_2005:P0C6V9.1\n",
    "rdrp2.Coronaviridae.Bat_Hp_betacoronavirus_Zhejiang2013:YP_009072438.1\n",
    "rdrp2.Coronaviridae.Bottlenose_dolphin_coronavirus_HKU22:AHB63494.1\n",
    "rdrp2.Coronaviridae.Duck_coronavirus:AKF17722.1\n",
    "rdrp2.Coronaviridae.Avian_infectious_bronchitis_virus_partridge_GD_S14_2003:AAT70770.1\n",
    "rdrp2.Coronaviridae.Infectious_bronchitis_virus:ADA83556.1\n",
    "...\n",
    "rdrp1.unclassified.Wenzhou_levi_like_virus_3:APG77299.1\n",
    "rdrp1.Leviviridae.Pseudomonas_phage_PP7:NP_042307.1\n",
    "\n",
    "```\n",
    "\n",
    "saved as: `gb_assign_group.txt`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 201213 REPEAT; accidently duplicated rdrp0.fa output here\n",
    "# 201226 REPEAT; update the viral-family classification names (monkey work)\n",
    "cd $WORK/wolf18"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">AMN92168.1|Bourbon_virus\n",
      ">YP_009352882.1|Dhori_virus\n",
      ">YP_145794.1|Thogoto_virus\n",
      ">AHB34055.1|Upolu_virus\n",
      ">ABF68025.1|Infectious_salmon_anemia_virus\n",
      ">AQM37684.1|Steelhead_trout_orthomyxovirus_1\n",
      ">APG77864.1|Beihai_orthomyxo_like_virus_2\n",
      ">APG77905.1|Hubei_orthomyxo_like_virus_5\n",
      ">YP_009246481.1|Tilapia_lake_virus\n",
      ">YP_009337891.1|Changping_earthworm_virus_2\n",
      "\n",
      ">branch.family.name:acc\n",
      ">rdrp5.unclassified.Wuhan_Insect_virus_3:AJG39263\n",
      ">rdrp5.unclassified.Mucorales_RNA_virus_1:AMK47917\n",
      ">rdrp5.unclassified.Wenling_crustacean_virus_9:YP_009329879\n",
      ">rdrp5.Bunyaviridae.Groundnut_chlorotic_fan_spot_virus:AJT59689\n",
      ">rdrp5.Tospoviridae.Soybean_vein_necrosis_virus:ADX01591\n",
      ">rdrp5.Tospoviridae.Bean_necrotic_mosaic_virus:YP_006468898\n",
      ">rdrp5.Tospoviridae.Polygonum_ringspot_tospovirus:AHZ45965\n",
      ">rdrp5.Tospoviridae.Pepper_chlorotic_spot_virus:AQX77525\n",
      ">rdrp5.Tospoviridae.Melon_yellow_spot_virus:BAG82842\n"
     ]
    }
   ],
   "source": [
    "grep \">\" gb_rdrp.afa | tail -\n",
    "echo ''\n",
    "head gb_assign_group_v2.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "KAPDSAARESLDRASEIMTGKSYNAVHTGDLSKLPNQGESPLRIVDSDLYSERSCCWVIEKEGRVVCKSTTLTRGMTGLLNTTRCSSPSELICKVLTVESLSEKIGDTSVEELLSHGRYFKCALRDQERGKPKSRAIFLSHPFFRLLSSVVETHARSVLSKVSAVYTATASAEQRAMMAAQVVESRKHVLNGDCTKYNEAIDADTLLKVWDAIGMGSIGVMLAYMVRRKCVLIKDTLVECPGGMLMGMFNATATLALQGTTDRFLSFSDDFITSFNSPAELREIEDLLFASCHNLSLKKSYISVASLEINSCTLTRDGDLATGLGCTAGVPFRGPLVTLKQTAAMLSGAVDSGVMPFHSAERLFQIKQQECAYRYNNPTYTTRNEDFLPTCLGGKTVISFQSLLTWDCHPFWYQVHPDGPDTIDQKVLSVLASK\n",
      ">YP_009337891.1\n",
      "WDDQDQSMFLRPKNRTGYGPLIFNTMKRISDMSPTRARELSEVFSVTEKERSISVLASGGTKFVPARGTSVPASTAFWDYQDQMRPIFEHYNIKYTDNSWWHIVICANIFGEYFEILPPTWDRSTLTKLFVEIFSAGLAVKQTEHNRSEGRNIVTMSISLQNFQNFVEEVAKIVNRMTGSHGTDLSSLEKRDLLRKVGLAASIELDTFLASLDKTKWNQLLQISTAMLLLAASYPNDASERRFVLLVGQIWREKCLYFPSKHSYYTGGMKTPKTIDELSRMNDEQLLNDNIRDDLMMVLRHYRKKRVIPQYIKCDLIMLMGMFNHSSTTLHIWPAYANHLDDNQTVSKIIDFCASSDDSMVRAKKILGMSALESYRTISSLWKSMGLNDSEDKSIIHDRLVKVEYNSNVFSMGQLIPNLSRDVAGTKVLYENPEKDLETMKNQLFVYINEGTLSTQDAAIILSDKYLTSLDIHDMLPFQKRHPIFLNNLTSAGLIPQCIPIWCGGTNHIPPELWGTMDDKMYWYHHHKDTGKTNLYLEFLASISTPPDVLASEMPVDFKLGL\n"
     ]
    }
   ],
   "source": [
    "# Rename the header to remove virus name\n",
    "# remove gaps from sequence (unaligned)\n",
    "sed 's/|.*//g' gb_rdrp.afa \\\n",
    "  | sed 's/-//g' - \\\n",
    "  > rdrp_1.tmp\n",
    "\n",
    "# One sequence \"Pseudomonas_phage_phiYY\" has no accession\n",
    "# YP_009618381.1\n",
    "sed -i 's/^>$/>YP_009618381.1/g' rdrp_1.tmp\n",
    "\n",
    "tail -n3 rdrp_1.tmp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "c912fa5199f573bbffcf4549336b9c4b  ../gbRdRp_201212.fa\n"
     ]
    }
   ],
   "source": [
    "# Iterate through each line / fasta name.\n",
    "# to swap out headers\n",
    "\n",
    "while read -r line; do\n",
    "  # Find headers\n",
    "  if [[ \"$line\" = \">\"* ]]; then\n",
    "    acc=$(echo $line | sed 's/\\..*//g' - | sed 's/>//g' -)\n",
    "    newheader=$(grep \"$acc\" gb_assign_group_v2.txt)\n",
    "    \n",
    "    if [[ \"$newheader\" = \"\" ]]; then\n",
    "      echo \">NA_$acc\" >> rdrp_2.tmp\n",
    "    else\n",
    "      echo $newheader >> rdrp_2.tmp\n",
    "    fi\n",
    "    \n",
    "  else\n",
    "    # print aa sequence\n",
    "    echo $line >> rdrp_2.tmp \n",
    "  fi\n",
    "done < rdrp_1.tmp\n",
    "\n",
    "\n",
    "# Manually remove first ten sequences (Group II introns outgroup)\n",
    "tail -n +21 rdrp_2.tmp > rdrp_3.tmp\n",
    "\n",
    "mv rdrp_3.tmp ../gbRdRp_201212.fa\n",
    "\n",
    "md5sum ../gbRdRp_201212.fa\n",
    "#rm *.tmp"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## WOLF20 RdRp\n",
    "\n",
    "FTP Access: `ftp://ftp.ncbi.nlm.nih.gov/pub/wolf/_suppl/yangshan/`\n",
    "\n",
    "Sequence data: `gb_rdrp.afa`\n",
    "\n",
    "Saved as: `gb_rdrp.fa`\n",
    "\n",
    "Date Accessed: `201212`\n",
    "\n",
    "![Figure 2](/home/artem/serratus/notebook/201210_rdrp0/wolf20/wolf20_figure2.jpg)\n",
    "\n",
    ">RNA virome analysis performed using complementary DNA derived from approximately 10 l of samples from Yangshan Deep-Water Harbour yielded 4,593 nearly full-length RNA virus RdRPs that formed 2,192 clusters at 75% amino acid identity which represents virus diversity at a level between species and genus. Among the RdRP sequences from GenBank (October 2018), 2,021 comparable clusters were detected. Thus, the 10 l water sample analysed here more than doubles the known diversity of RNA viruses.\n",
    "\n",
    "There are two sets of data treated independently here; the genbank RdRp and the yangshan RdRp. Only yangshen sequences will be considered as the genbank records are not comprehensive. The data is clustered globally at `75%` aa identity.\n",
    "\n",
    "### Level 1/2 - Supergroup / Branches and Clusters\n",
    "\n",
    "From [Supplementary Table 1](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7508674/bin/41564_2020_755_MOESM2_ESM.xlsx) parsed the \"clade id\" field to retrieve branch numbering and relate to the OV.x clusters. A few sequences are unassigned / deep so they will form `rdrp0`.\n",
    "\n",
    "The `OV.x` was renamed to `yaOVx` to remove period. So `OV.1` became `rdrp2.yaOV1`\n",
    "\n",
    "Saved as: `ovx.branches.txt`\n",
    "\n",
    "### Level 3 - ORF Identifier\n",
    "Sequence data is from `rdrp.ya.fa`\n",
    "\n",
    "Header parsing:\n",
    "`>ya20_JAAOEH010000011_1 JAAOEH010000011.1 5194-3782 OV.1 NODE_11_truseq orf.65`\n",
    "will become\n",
    "`>rdrp2.yaOV1.orf65\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mv: missing destination file operand after 'ya_2.tmp'\n",
      "Try 'mv --help' for more information.\n"
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "cd $WORK/wolf20\n",
    "\n",
    "# Iterate through each line / fasta name.\n",
    "# to swap out headers\n",
    "\n",
    "while read -r line; do\n",
    "  # Find headers\n",
    "  if [[ \"$line\" = \">\"* ]]; then\n",
    "    acc=$(echo $line | cut -f 4 -d ' ' - | sed 's/\\.//g' -)\n",
    "    branch=$(grep \"$acc\\.\" ovx.branches.txt)\n",
    "    orf=$(echo $line | sed 's/.* //g' - | sed 's/\\.//g' - )\n",
    "    \n",
    "    if [[ \"$branch\" = \"\" ]]; then\n",
    "      echo \">NA\" >> ya_2.tmp\n",
    "    else\n",
    "      echo \">$branch\"\"$orf\" >> ya_2.tmp\n",
    "    fi\n",
    "    \n",
    "  else\n",
    "    # print aa sequence\n",
    "    echo $line >> ya_2.tmp \n",
    "  fi\n",
    "done < rdrp.ya.fa\n",
    "\n",
    "mv ya_2.tmp ../yaRdRp_201212.fa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## rdrp0 - pilot panproteome\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "c912fa5199f573bbffcf4549336b9c4b  gbRdRp_201212.fa\n",
      "7e6781ed5902ea6034393c0dae521c62  yaRdRp_201212.fa\n",
      "ce012c07be4cfc2bc7abd2dc4c81c10a  rdrp0.fa\n"
     ]
    }
   ],
   "source": [
    "cd $WORK\n",
    "\n",
    "# Make rdrp0\n",
    "cat gbRdRp_201212.fa yaRdRp_201212.fa > rdrp0.fa\n",
    "samtools faidx rdrp0.fa\n",
    "\n",
    "md5sum gbRdRp_201212.fa\n",
    "md5sum yaRdRp_201212.fa \n",
    "md5sum rdrp0.fa\n",
    "md5sum rdrp0.fa > rdrp0.fa.md5"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Upload to S3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 9.8M\n",
      "drwxrwxr-x  5 artem artem 4.0K Dec 26 22:31 \u001b[0m\u001b[01;34m.\u001b[0m\n",
      "drwxr-xr-x 41 artem artem  12K Dec 26 22:31 \u001b[01;34m..\u001b[0m\n",
      "-rw-rw-r--  1 artem artem 2.5M Dec 26 22:31 gbRdRp_201212.fa\n",
      "-rw-rw-r--  1 artem artem  16K Dec 20 23:18 quenya_rdrp.fa\n",
      "-rw-rw-r--  1 artem artem 4.6M Dec 26 22:31 rdrp0.fa\n",
      "-rw-rw-r--  1 artem artem 524K Dec 26 22:31 rdrp0.fa.fai\n",
      "-rw-rw-r--  1 artem artem   43 Dec 26 22:31 rdrp0.fa.md5\n",
      "drwxrwxr-x  7 artem artem 4.0K Dec 19 16:39 \u001b[01;34mrev1\u001b[0m\n",
      "drwxrwxr-x  2 artem artem 4.0K Dec 26 22:31 \u001b[01;34mwolf18\u001b[0m\n",
      "drwxrwxr-x  5 artem artem 192K Dec 12 22:20 \u001b[01;34mwolf20\u001b[0m\n",
      "-rw-rw-r--  1 artem artem 2.1M Dec 12 22:14 yaRdRp_201212.fa\n"
     ]
    }
   ],
   "source": [
    "cd $WORK\n",
    "ls -alh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "upload: ./rdrp0.fa.md5 to s3://serratus-public/notebook/201210_rdrp0/rdrp0.fa.md5\n",
      "upload: wolf18/~$wolf18_vlist.xlsx to s3://serratus-public/notebook/201210_rdrp0/wolf18/~$wolf18_vlist.xlsx\n",
      "upload: wolf18/gb_assign_group_v2.txt to s3://serratus-public/notebook/201210_rdrp0/wolf18/gb_assign_group_v2.txt\n",
      "upload: ./rdrp0.fa.fai to s3://serratus-public/notebook/201210_rdrp0/rdrp0.fa.fai\n",
      "upload: ./gbRdRp_201212.fa to s3://serratus-public/notebook/201210_rdrp0/gbRdRp_201212.fa\n",
      "upload: ./rdrp0.fa to s3://serratus-public/notebook/201210_rdrp0/rdrp0.fa\n",
      "upload: wolf18/rdrp_1.tmp to s3://serratus-public/notebook/201210_rdrp0/wolf18/rdrp_1.tmp\n",
      "upload: wolf18/rdrp_2.tmp to s3://serratus-public/notebook/201210_rdrp0/wolf18/rdrp_2.tmp\n"
     ]
    }
   ],
   "source": [
    "aws s3 sync ./ $S3_WORK # done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create rdrp0 database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load serratus-align container on EC2\n",
    "# From base amazon linux 2\n",
    "sudo yum install -y docker\n",
    "sudo yum install -y git\n",
    "sudo service docker start\n",
    "\n",
    "git clone --branch diamond-dev https://github.com/ababaian/serratus.git; cd serratus\n",
    "\n",
    "# If you want to upload containers to your repository, include this.\n",
    "export DOCKERHUB_USER='serratusbio' # optional\n",
    "sudo docker login # optional\n",
    "\n",
    "# Build all containers and upload them docker hub repo (if available)\n",
    "cd containers\n",
    "./build_containers.sh   # run this in the folder 'serratus/containers'\n",
    "\n",
    "sudo docker run --rm --entrypoint /bin/bash \\\n",
    "  -it serratus-align:latest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# rdrp0 v201213 for pilot run\n",
    "mkdir rdrp0; cd rdrp0\n",
    "GENOME='rdrp0'\n",
    "\n",
    "# Download rdrp0\n",
    "aws s3 cp s3://serratus-public/notebook/201210_rdrp0/rdrp0.fa ./\n",
    "\n",
    "# Make diamond index for protref5\n",
    "diamond makedb --in $GENOME.fa -d $GENOME\n",
    "\n",
    "# Make fasta index for protref5\n",
    "samtools faidx $GENOME.fa\n",
    "mv $GENOME.fa.fai $GENOME.sumzer.tsv\n",
    "\n",
    "md5sum * > $GENOME.md5\n",
    "\n",
    "# use protref5 msa as place-holder\n",
    "aws s3 cp s3://serratus-public/seq/protref5/protref5.msa ./rdrp0.msa\n",
    "\n",
    "# 657d302bd62a9e0b588668101f581e4c  rdrp0.dmnd\n",
    "# 8479a3347bbe73224cb2eac0c2138a92  rdrp0.fa\n",
    "# 6bf2ffa27bb2b08bf0d6056b675fa348  rdrp0.sumzer.tsv\n",
    "# e094fc7db19c07ffcedf8bc42963ab80  rdrp0.msa\n",
    "\n",
    "# Upload to S3\n",
    "aws s3 sync ./ s3://serratus-public/seq/$GENOME/\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Revision 0 - Clean-up \"Virus Name\" field (deprecated)\n",
    "\n",
    "In parsing the \"Virus Name\" from `rdrp0` the \".\" character was in quite a few virus names and these result in icky downstream parsing. This is a manual removal of those periods from all rdrp0 names to make a 'clean' version\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# revision 0 folder\n",
    "mkdir rev0; cd rev0\n",
    "\n",
    "# Wolf18 Genbank Sequences\n",
    "cp ../gbRdRp_201212.fa  ./\n",
    "# Wolf20 Yangshen Sequences\n",
    "cp ../yaRdRp_201212.fa  ./ "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Update Sequence names (to make uniform)\n",
    "# This is from ongoing work from\n",
    "\n",
    "## sp. --> sp\n",
    "grep \"sp\\.\" gbRdRp_201212.fa | less -NS -\n",
    "sed -i 's/sp\\./sp/g' gbRdRp_201212.fa\n",
    "\n",
    "#      1 >rdrp1.Narnaviridae.Rhizophagus_sp._HR1_mitovirus_like_ssRNA:BAN85985.1\n",
    "#      2 >rdrp1.Narnaviridae.Rhizophagus_sp._RF1_mitovirus:BAJ23143.2\n",
    "#      3 >rdrp2.Picobirnaviridae.Picobirnavirus_sp.:AQS16638.1\n",
    "#      4 >rdrp2.Picobirnaviridae.Picobirnavirus_sp.:AOW41971.1\n",
    "#      5 >rdrp2.Picobirnaviridae.Picobirnavirus_sp.:AOW41973.1\n",
    "#      6 >rdrp2.Iflaviridae.Iflavirus_sp.:APB88805.1\n",
    "#      7 >rdrp2.unclassified.Chaetoceros_sp._RNA_virus_2:BAK40203.1\n",
    "#      8 >rdrp2.unclassified.Posavirus_sp.:APQ44560.1\n",
    "#      9 >rdrp2.unclassified.Posavirus_sp.:APQ44553.1\n",
    "#     10 >rdrp2.unclassified.Posavirus_sp.:APQ44556.1\n",
    "#     11 >rdrp2.unclassified.Posavirus_sp.:APQ44559.1\n",
    "#     12 >rdrp2.unclassified.Posavirus_sp.:APQ44517.1\n",
    "#     13 >rdrp2.unclassified.Basavirus_sp.:APQ44489.1\n",
    "#     14 >rdrp2.unclassified.Basavirus_sp.:APQ44495.1\n",
    "#     15 >rdrp2.unclassified.Posavirus_sp.:APQ44558.1\n",
    "#     16 >rdrp2.unclassified.Basavirus_sp.:APQ44499.1\n",
    "#     17 >rdrp2.unclassified.Basavirus_sp.:APQ44492.1\n",
    "#     18 >rdrp2.unclassified.Basavirus_sp.:APQ44496.1\n",
    "#     19 >rdrp2.unclassified.Basavirus_sp.:APQ44502.1\n",
    "#     20 >rdrp2.unclassified.Posavirus_sp.:APQ44531.1\n",
    "#     21 >rdrp2.unclassified.Posavirus_sp.:YP_009333148.1\n",
    "#     22 >rdrp2.unclassified.Posavirus_sp.:APQ44537.1\n",
    "#     23 >rdrp2.unclassified.Rasavirus_sp.:APQ44507.1\n",
    "#     24 >rdrp2.unclassified.Husavirus_sp.:APQ44514.1\n",
    "#     25 >rdrp2.unclassified.Rasavirus_sp.:APQ44506.1\n",
    "#     26 >rdrp2.unclassified.Rasavirus_sp.:YP_009333305.1\n",
    "#     27 >rdrp2.unclassified.Posavirus_sp.:APQ44547.1\n",
    "#     28 >rdrp2.unclassified.Basavirus_sp.:APQ44500.1\n",
    "#     29 >rdrp2.Picornaviridae.Sicinivirus_sp.:APR73491.1\n",
    "#     30 >rdrp2.Picornaviridae.Picornaviridae_sp._rodent_Ee_PicoV_NX2015:APA29022.1\n",
    "#     31 >rdrp2.Picornaviridae.Picornaviridae_sp._rodent_Mc_PicoV_Tibet2015:APA29023.1\n",
    "#     32 >rdrp2.Picornaviridae.Enterovirus_sp.:AHY21610.1\n",
    "#     33 >rdrp2.Picornaviridae.Picornaviridae_sp._rodent_CK_PicoV_Tibet2014:APA29019.1\n",
    "#     34 >rdrp2.Picornaviridae.Picornaviridae_sp._rodent_Rn_PicoV_SX2015_1:APA29018.1\n",
    "#     35 >rdrp2.Picornaviridae.Picornaviridae_sp._rodent_Ds_PicoV_IM2014:APA29017.1\n",
    "#     36 >rdrp3.Picobirnaviridae.Picobirnavirus_sp.:AOW41972.1\n",
    "#     37 >rdrp4.unclassified.unclassified_Rhizophagus_sp._RF1_medium_virus:BAJ23141.1\n",
    "#     38 >rdrp4.Chrysoviridae.Fusarium_oxysporum_f._sp._dianthi_mycovirus_1:YP_009158913.1\n",
    "#     39 >rdrp5.Bunyaviridae.Bunyavirus_sp.:AOY18806.1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Manual interventions\n",
    "\n",
    "grep \">\" gbRdRp_201212.fa | cut -f3- -d'.' | cut -f1 -d':' | grep \"\\.\" - | less -NS\n",
    "\n",
    "#      1 Chaetoceros_socialis_f._radians_RNA_virus_01\n",
    "sed -i 's/f\\./f/g' gbRdRp_201212.fa\n",
    "#      2 Norovirus_dog_GVI.1_HKU_Ca026F_2007_HKG\n",
    "sed -i 's/GVI\\.1/GVI_1/g' gbRdRp_201212.fa\n",
    "#      3 Norovirus_cat_GIV.2_CU081210E_USA_2010\n",
    "sed -i 's/GIV\\.2/GIV_2/g' gbRdRp_201212.fa\n",
    "#      4 Norovirus_Hu_GII.12_CGMH42_2010_TW\n",
    "#      5 Norovirus_Hu_GII.12_CGMH40_2010_TW\n",
    "#      6 Norovirus_GII.17\n",
    "sed -i 's/GII\\.12/GII_12/g' gbRdRp_201212.fa\n",
    "sed -i 's/GII\\.17/GII_17/g' gbRdRp_201212.fa\n",
    "#      7 Sapovirus_Hu_GI.2_BR_DF01_BRA_2009\n",
    "sed -i 's/GI\\.2/GI_2/g' gbRdRp_201212.fa\n",
    "#      8 Sapovirus_GII.8\n",
    "sed -i 's/GII\\.8/GII_8/g' gbRdRp_201212.fa\n",
    "#      9 St._Louis_encephalitis_virus\n",
    "#     10 St._Louis_encephalitis_virus\n",
    "sed -i 's/St\\._L/St_L/g' gbRdRp_201212.fa\n",
    "#     11 Fusarium_oxysporum_f._sp_dianthi_mycovirus_1\n",
    "##\n",
    "\n",
    "# Accesion field does not contain a version number\n",
    "# Do not chnage accessions at this point\n",
    "grep \">\" gbRdRp_201212.fa |  grep -v \"\\.[0-9]$\" -\n",
    "#>rdrp1.Leviviridae.Escherichia_virus_Qbeta:4R71\n",
    "#>rdrp3.Flaviviridae.Douroucouli_hepatitis_GB_virus_A:T08841\n",
    "#>rdrp3.Flaviviridae.Marmoset_hepatitis_GB_virus_A:T08839\n",
    "#>rdrp3.Alphaflexiviridae.Potato_aucuba_mosaic_virus:2012194A\n",
    "#>rdrp3.Virgaviridae.Barley_stripe_mosaic_virus:2211403A\n",
    "#>rdrp4.Reoviridae.Simian_rotavirus:2R7Q\n",
    "#>rdrp4.unclassified.White_button_mushroom_virus_1:T00494\n",
    "#>rdrp5.Peribunyaviridae.La_Crosse_virus:5AMR_A\n",
    "#>rdrp5.Orthomyxoviridae.Influenza_B_virus_(B_Memphis_13_2003):4WRT_B\n",
    "#>rdrp5.Orthomyxoviridae.Influenza_A_virus_(A_little_yellow_shouldered_bat_Guatemala_060_2010(H17N10)):4WSB_B\n",
    "\n",
    "\n",
    "md5sum *\n",
    "# 7d2c1858d5d842ada3e8103c43f29e27  gbRdRp_201212.fa\n",
    "# 7e6781ed5902ea6034393c0dae521c62  yaRdRp_201212.fa\n",
    "\n",
    "# Overwrite previous gb versions\n",
    "mv gbRdRp_201212.fa ../\n",
    "mv yaRdRp_201212.fa ../"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Revision 1 - Cluster known + unclassified sequences\n",
    "\n",
    "Some lessons learned so far... there are ALOT of new viruses to be discovered! From the ~9.5K datasets run with the `rdrp0` pilot, there are ~4,700 high score (>50) and high divergence (55-85% aa id) hits at the family level. Random sampling from vertebrate/virome SRA queries means overlap is likely to be minimal here but a conservative estimate would be 1,000 distinct RdRp, about 25% of the known biodiversity of sequences.\n",
    "\n",
    "### Objectives\n",
    "\n",
    "- There is a large amount of \"unclassified\" sequences in the data within branches, these should be grouped relative to one another such that if two unclassified sequences exist in one branch, they do not \"collide\" in the read summaries.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tue Dec 15 20:57:00 PST 2020\n",
      "6ac78a036910813c0f5fb2e7ef0b88599e683959\n"
     ]
    }
   ],
   "source": [
    "# Serratus commit version\n",
    "SERRATUS=\"/home/artem/serratus\"\n",
    "cd $SERRATUS\n",
    "\n",
    "# Create local run directory\n",
    "WORK=\"$SERRATUS/notebook/201210_rdrp0\"\n",
    "mkdir -p $WORK; cd $WORK\n",
    "\n",
    "# S3 notebook path\n",
    "S3_WORK='s3://serratus-public/notebook/201210_rdrp0/'\n",
    "\n",
    "# date and version\n",
    "date\n",
    "git rev-parse HEAD # commit version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Work on EC2 Instance\n",
    "\n",
    "# Local usearch install\n",
    "#The clustered database was made with usearch:\n",
    "wget https://drive5.com/downloads/usearch11.0.667_i86linux32.gz\n",
    "gzip -dc usearch11.0.667_i86linux32.gz > usearch\n",
    "chmod 755 usearch; sudo mv usearch /usr/bin/usearch\n",
    "rm https://drive5.com/downloads/usearch11.0.667_i86linux32.gz\n",
    "\n",
    "# Install seqkit\n",
    "wget https://github.com/shenwei356/seqkit/releases/download/v0.12.0/seqkit_linux_amd64.tar.gz\n",
    "  tar -xvf seqkit*\n",
    "  sudo mv seqkit /usr/local/bin/\n",
    "  rm seqkit_linux*\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize workspace\n",
    "\n",
    "# Download rdrp working directory\n",
    "mkdir rdrp0; cd rdrp0\n",
    "aws s3 sync s3://serratus-public/notebook/201210_rdrp0/ ./\n",
    "\n",
    "# revision 1 folder\n",
    "mkdir rev1; cd rev1\n",
    "\n",
    "# Wolf18 == Genbank Sequences from 2018\n",
    "cp ../gbRdRp_201212.fa  ./\n",
    "# Wolf20 Yangshen Sequences\n",
    "cp ../yaRdRp_201212.fa  ./ \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# in gbRdRp (from wolf 18); seperate out taxonomic and unclassified sequences\n",
    "grep \">\" gbRdRp_201212.fa | wc -l\n",
    "# 4617\n",
    "\n",
    "# -v inverse match\n",
    "seqkit grep -v -r -p 'unclassified' gbRdRp_201212.fa > gb_tax.fa\n",
    "grep \">\" gb_tax.fa | wc -l\n",
    "# 2802\n",
    "\n",
    "seqkit grep -r -p 'unclassified' gbRdRp_201212.fa > gb_unc.fa\n",
    "grep \">\" gb_unc.fa | wc -l\n",
    "# 1815\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sort and Uclust unclassified sequences\n",
    "\n",
    "usearch -sortbylength gb_tax.fa \\\n",
    "   -fastaout gb_tax.sort.fa\n",
    "\n",
    "usearch -sortbylength gb_unc.fa \\\n",
    "   -fastaout gb_unc.sort.fa\n",
    "\n",
    "cat gb_tax.sort.fa gb_unc.sort.fa > gb_cat.sort.fa\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Cluster sequences at 75% identity\n",
    "\n",
    "# Prune TAX\n",
    "usearch -cluster_smallmem gb_tax.sort.fa \\\n",
    "   -id 0.75 \\\n",
    "   -maxaccepts 4 \\\n",
    "   -maxrejects 64 \\\n",
    "   -maxhits 1 \\\n",
    "   -uc gb_tax.id75.uc \\\n",
    "   -centroids gb_tax.id75.fa\n",
    "\n",
    "#      Seqs  2802\n",
    "#  Clusters  1738\n",
    "#  Max size  29\n",
    "#  Avg size  1.6\n",
    "#  Min size  1\n",
    "#Singletons  1298, 46.3% of seqs, 74.7% of clusters\n",
    "\n",
    "\n",
    "\n",
    "# Prune UNC\n",
    "usearch -cluster_smallmem gb_unc.sort.fa \\\n",
    "   -id 0.75 \\\n",
    "   -maxaccepts 4 \\\n",
    "   -maxrejects 64 \\\n",
    "   -maxhits 1 \\\n",
    "   -uc gb_unc.id75.uc \\\n",
    "   -centroids gb_unc.id75.fa\n",
    "\n",
    "#      Seqs  1815\n",
    "#  Clusters  1708\n",
    "#  Max size  6\n",
    "#  Avg size  1.1\n",
    "#  Min size  1\n",
    "#Singletons  1620, 89.3% of seqs, 94.8% of clusters\n",
    "\n",
    "\n",
    "\n",
    "# Prune CAT\n",
    "usearch -cluster_smallmem gb_cat.sort.fa \\\n",
    "   -sortedby other \\\n",
    "   -id 0.75 \\\n",
    "   -maxaccepts 4 \\\n",
    "   -maxrejects 64 \\\n",
    "   -maxhits 1 \\\n",
    "   -uc gb_cat.id75.uc \\\n",
    "   -centroids gb_cat.id75.fa\n",
    "   \n",
    "#      Seqs  4617\n",
    "#  Clusters  3426\n",
    "#  Max size  29\n",
    "#  Avg size  1.3\n",
    "#  Min size  1\n",
    "#Singletons  2889, 62.6% of seqs, 84.3% of clusters\n",
    "\n",
    "## At 75%, no \"unclassifed\" sequence groups with a taxonomic identifier\n",
    "\n",
    "mkdir id75\n",
    "mv *id75.* id75/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Repeat process at 55%; base of diamond detection\n",
    "\n",
    "# Prune TAX\n",
    "usearch -cluster_smallmem gb_tax.sort.fa \\\n",
    "   -id 0.55 \\\n",
    "   -maxaccepts 4 \\\n",
    "   -maxrejects 64 \\\n",
    "   -maxhits 1 \\\n",
    "   -uc gb_tax.id55.uc \\\n",
    "   -centroids gb_tax.id55.fa\n",
    "\n",
    "#      Seqs  2802\n",
    "#  Clusters  781\n",
    "#  Max size  143\n",
    "#  Avg size  3.6\n",
    "#  Min size  1\n",
    "#Singletons  433, 15.5% of seqs, 55.4% of clusters\n",
    "\n",
    "\n",
    "\n",
    "# Prune UNC\n",
    "usearch -cluster_smallmem gb_unc.sort.fa \\\n",
    "   -id 0.55 \\\n",
    "   -maxaccepts 4 \\\n",
    "   -maxrejects 64 \\\n",
    "   -maxhits 1 \\\n",
    "   -uc gb_unc.id55.uc \\\n",
    "   -centroids gb_unc.id55.fa\n",
    "\n",
    "#      Seqs  1815\n",
    "#  Clusters  1428\n",
    "#  Max size  11\n",
    "#  Avg size  1.3\n",
    "#  Min size  1\n",
    "#Singletons  1190, 65.6% of seqs, 83.3% of clusters\n",
    "\n",
    "\n",
    "\n",
    "# Prune CAT\n",
    "usearch -cluster_smallmem gb_cat.sort.fa \\\n",
    "   -sortedby other \\\n",
    "   -id 0.55 \\\n",
    "   -maxaccepts 4 \\\n",
    "   -maxrejects 64 \\\n",
    "   -maxhits 1 \\\n",
    "   -uc gb_cat.id55.uc \\\n",
    "   -centroids gb_cat.id55.fa\n",
    "\n",
    "#      Seqs  4617\n",
    "#  Clusters  2143\n",
    "#  Max size  143\n",
    "#  Avg size  2.2\n",
    "#  Min size  1\n",
    "#Singletons  1548, 33.5% of seqs, 72.2% of clusters\n",
    "\n",
    "\n",
    "## At 55%, no \"unclassifed\" sequence groups with a taxonomic identifier\n",
    "\n",
    "mkdir id55\n",
    "mv *.id55.* id55/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Repeat process at 45%; base of diamond detection\n",
    "\n",
    "# Prune TAX\n",
    "usearch -cluster_smallmem gb_tax.sort.fa \\\n",
    "   -id 0.45 \\\n",
    "   -maxaccepts 4 \\\n",
    "   -maxrejects 64 \\\n",
    "   -maxhits 1 \\\n",
    "   -uc gb_tax.id45.uc \\\n",
    "   -centroids gb_tax.id45.fa\n",
    "\n",
    "#      Seqs  2802\n",
    "#  Clusters  566\n",
    "#  Max size  156\n",
    "#  Avg size  5.0\n",
    "#  Min size  1\n",
    "#Singletons  280, 10.0% of seqs, 49.5% of clusters\n",
    "\n",
    "\n",
    "\n",
    "# Prune UNC\n",
    "usearch -cluster_smallmem gb_unc.sort.fa \\\n",
    "   -id 0.45 \\\n",
    "   -maxaccepts 4 \\\n",
    "   -maxrejects 64 \\\n",
    "   -maxhits 1 \\\n",
    "   -uc gb_unc.id45.uc \\\n",
    "   -centroids gb_unc.id45.fa\n",
    "\n",
    "#      Seqs  1815\n",
    "#  Clusters  1176\n",
    "#  Max size  17\n",
    "#  Avg size  1.5\n",
    "#  Min size  1\n",
    "#Singletons  886, 48.8% of seqs, 75.3% of clusters\n",
    "\n",
    "# Prune CAT\n",
    "usearch -cluster_smallmem gb_cat.sort.fa \\\n",
    "   -sortedby other \\\n",
    "   -id 0.45 \\\n",
    "   -maxaccepts 4 \\\n",
    "   -maxrejects 64 \\\n",
    "   -maxhits 1 \\\n",
    "   -uc gb_cat.id45.uc \\\n",
    "   -centroids gb_cat.id45.fa\n",
    "\n",
    "#      Seqs  4617\n",
    "#  Clusters  1651\n",
    "#  Max size  156\n",
    "#  Avg size  2.8\n",
    "#  Min size  1\n",
    "#Singletons  1073, 23.2% of seqs, 65.0% of clusters\n",
    "\n",
    "\n",
    "## At 45%, no \"unclassifed\" sequence groups with a taxonomic identifier.\n",
    "\n",
    "mkdir id45\n",
    "mv *.id45.* id45/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Repeat process at 35%; base of diamond detection (NA)\n",
    "\n",
    "# Prune TAX\n",
    "usearch -cluster_smallmem gb_tax.sort.fa \\\n",
    "   -id 0.35 \\\n",
    "   -maxaccepts 4 \\\n",
    "   -maxrejects 64 \\\n",
    "   -maxhits 1 \\\n",
    "   -uc gb_tax.id35.uc \\\n",
    "   -centroids gb_tax.id35.fa\n",
    "\n",
    "# Prune UNC\n",
    "usearch -cluster_smallmem gb_unc.sort.fa \\\n",
    "   -id 0.35 \\\n",
    "   -maxaccepts 4 \\\n",
    "   -maxrejects 64 \\\n",
    "   -maxhits 1 \\\n",
    "   -uc gb_unc.id35.uc \\\n",
    "   -centroids gb_unc.id35.fa\n",
    "\n",
    "# Prune CAT\n",
    "usearch -cluster_smallmem gb_cat.sort.fa \\\n",
    "   -sortedby other \\\n",
    "   -id 0.35 \\\n",
    "   -maxaccepts 4 \\\n",
    "   -maxrejects 64 \\\n",
    "   -maxhits 1 \\\n",
    "   -uc gb_cat.id35.uc \\\n",
    "   -centroids gb_cat.id35.fa\n",
    "\n",
    "## At 35%, no \"unclassifed\" sequence groups with a taxonomic identifier.\n",
    "\n",
    "mkdir id35\n",
    "mv *.id35.* id35/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mkdir -p all\n",
    "mv *.fa all/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Conclusions\n",
    "\n",
    "The clustering here is actually not as extensive as I originally had thought it would be. It looks like "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Assigning Sequence Clusters\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Identity to use for clustering\n",
    "ID='45'\n",
    "\n",
    "# Extract centroid sequence names\n",
    "grep \">\" id$ID/gb_cat.id$ID.fa \\\n",
    "  | sed 's/>//g' \\\n",
    "  > centroid$ID.name.tmp\n",
    "\n",
    "\n",
    "# Convert uc format to TSV\n",
    "# CENTROID hit1 hit2 hitN ...\n",
    "rm clust.tmp; touch clust.tmp\n",
    "\n",
    "# Read through centroid names\n",
    "while read -r line; do\n",
    "  # parse to just accession number\n",
    "  acc=$(echo $line | sed 's/.*://g')\n",
    "  \n",
    "  # Search accession in uclust file\n",
    "  # extract seqname and make it one line\n",
    "  grep \"$acc\" id$ID/gb_cat.id$ID.uc \\\n",
    "    | cut -f 9 \\\n",
    "    | tr '\\n' '\\t' \\\n",
    "    >> clust.tmp\n",
    "    \n",
    "  echo -e \"\\n\" >> clust.tmp\n",
    "\n",
    "done < centroid$ID.name.tmp\n",
    "\n",
    "# remove empty rows\n",
    "cat clust.tmp | sed '/^$/d' > cluster$ID.tsv\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Assign \"family_id\" to sequences (unclassified)\n",
    "\n",
    "# Field 1: Centroid Name\n",
    "# Field 2+: Sequence name\n",
    "cut -f 1 cluster$ID.tsv \\\n",
    "  | sed 's/\\./\\t/g' - \\\n",
    "  | sed 's/:/\\t/g' - \\\n",
    "  > centroid$ID.info\n",
    "\n",
    "# Centroid Information tsv\n",
    "# 1: Branch\n",
    "# 2: Family\n",
    "# 3: Rep. Viral name\n",
    "# 4: Rep. Viral accession\n",
    "# 5: Rep. Viral accession version\n",
    "# 6: Re-annotated Viral Family Name\n",
    "\n",
    "# Iterate through centroid file\n",
    "# assign each \"unclassified family\" a unique ordinal number\n",
    "# 1055 --> use 4 leading zeros\n",
    "\n",
    "# ID\n",
    "N=1\n",
    "rm family2.tmp\n",
    "\n",
    "while read -r line; do\n",
    "  # Read Family name\n",
    "  #echo $line\n",
    "  \n",
    "  family=$(echo $line | cut -f 2 -d' ' -)  \n",
    "\n",
    "  if [[ \"$family\" = \"unclassified\" ]]; then\n",
    "\n",
    "    uncN=$(printf \"%04d\" $N)\n",
    "    echo \"unc$uncN\" >> family2.tmp\n",
    "    \n",
    "    # increment up\n",
    "    N=$((N+1))\n",
    "    \n",
    "  else\n",
    "    echo $family >> family2.tmp\n",
    "  \n",
    "  fi\n",
    "\n",
    "done < centroid$ID.info\n",
    "\n",
    "# Add 6th column of new family names\n",
    "cp centroid$ID.info centroid.tmp\n",
    "paste family2.tmp centroid.tmp  > centroid$ID.info\n",
    "rm *.tmp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Assign Centroid Family-Name to Each Sequence\n",
    "cat all/gbRdRp_201212.fa all/yaRdRp_201212.fa > rdrp0_r1.fa\n",
    "\n",
    "#\n",
    "cut -f 1  centroid$ID.info > fam.name.tmp\n",
    "cut -f 2- cluster$ID.tsv   > cluster.members.tmp\n",
    "paste fam.name.tmp cluster.members.tmp > assign.family.tmp\n",
    "# col 1 == new name\n",
    "# col N == old sequence name\n",
    "\n",
    "while read -r line; do\n",
    "  newfam=$(echo $line | cut -f1 -d' ' -)\n",
    "  echo $newfame\n",
    "  \n",
    "  echo $line \\\n",
    "    | cut -f 2- -d' ' - \\\n",
    "    > members.tmp\n",
    "    \n",
    "  cat members.tmp \\\n",
    "    | tr \" \" \"\\n\" \\\n",
    "    > members2.tmp\n",
    "    \n",
    "    while read -r line2; do\n",
    "      # rdrp5.Sunviridae.Sunshine_Coast_virus:YP_009094051.1\n",
    "      branch=$(echo $line2 | cut -f1 -d'.')\n",
    "      oldfam=$(echo $line2 | cut -f2 -d'.')\n",
    "      virname=$(echo $line2 | cut -f3 -d'.' | cut -f1 -d':')\n",
    "      acc=$(echo $line2 | cut -f2 -d':')\n",
    "      \n",
    "      echo $branch $oldfam $virname $acc\n",
    "      echo $branch $newfam $virname $acc\n",
    "      \n",
    "      # Inline rename\n",
    "      matchline=$( echo $(grep -n $acc rdrp0_r1.fa  | cut -f1 -d':' -)s)\n",
    "      \n",
    "      sed -i \"$matchline/.*/>$branch.$newfam.$virname:$acc/\" rdrp0_r1.fa\n",
    "      \n",
    "    done < members2.tmp\n",
    "    \n",
    "    echo ''\n",
    "\n",
    "done < assign.family.tmp\n",
    "\n",
    "grep \">\" rdrp0_r1.fa | sed 's/>//g' > rdrp0_r1.fai\n",
    "\n",
    "rm *.tmp\n",
    "rm gbRdRp.fa yaRdRp.fa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aws s3 sync ./ s3://serratus-public/notebook/201210_rdrp0/rev1/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## CHECKPOINT\n",
    "\n",
    "As of `201226` all code above this checkpoint is copied to a new directory\n",
    "\n",
    "CHECKPOINT: `s3://serratus-public/notebook/201226_rdrp0/`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aws s3 sync ./ s3://serratus-public/notebook/201226_rdrp0/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Revision 2 - RdRp from GenBank NT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mkdir rev2; cd rev2\n",
    "cp ../rdrp0.fa ./"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Install diamond\n",
    "# DIAMONDVERSION='0.9.35'\n",
    "# wget --quiet https://github.com/bbuchfink/diamond/releases/download/v\"$DIAMONDVERSION\"/diamond-linux64.tar.gz\n",
    "# tar -xvf diamond-linux64.tar.gz\n",
    "# rm diamond-linux64.tar.gz\n",
    "# sudo mv    diamond /usr/local/bin/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Install diamond\n",
    "DIAMONDVERSION='2.0.6-dev'\n",
    "cd ~/\n",
    "\n",
    "# Libraries for diamond\n",
    "sudo yum -y install git gcc gcc-c++ glibc-devel \\\n",
    "  cmake patch automake zlib-devel\n",
    "\n",
    "# grab latest with fix from Benjamin\n",
    "git clone https://github.com/bbuchfink/diamond.git\n",
    "cd diamond\n",
    "\n",
    "mkdir bin; cd bin\n",
    "cmake ..\n",
    "make -j4\n",
    "sudo make install\n",
    "\n",
    "# build\n",
    "# cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_BUILD_MARCH=nehalem ..\n",
    "# make && make install"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "GENOME='rdrp0'\n",
    "cp ../rev1/rdrp0_r1.fa ./rdrp0.fa\n",
    "\n",
    "diamond makedb --in $GENOME.fa -d $GENOME"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Quenya sensitivity test\n",
    "# rdrp segments from q-viruses\n",
    "aws s3 cp s3://serratus-public/tmp/quenya5.fasta ./qRdRP.fa\n",
    "aws s3 cp s3://serratus-public/tmp/dicistro_all.fasta ./dRdRP.fa\n",
    "# 22 input Q nucleotide sequences\n",
    "\n",
    "# Dicistro\n",
    "IN='dRdRP.fa'\n",
    "GENOME='rdrp0'\n",
    "OUTNAME='d_v_rdrp0'\n",
    "\n",
    "# Diamond blastx alignment\n",
    "\n",
    "time cat $IN |\\\n",
    "diamond blastx \\\n",
    "  -d \"$GENOME\".dmnd \\\n",
    "  -p 4 \\\n",
    "  -k 1 \\\n",
    "  -f 6 qseqid qstart qend qlen qstrand \\\n",
    "       sseqid sstart send slen \\\n",
    "       pident evalue cigar \\\n",
    "       qseq qseq_translated \\\n",
    "  > \"$OUTNAME\".us.pro\n",
    "\n",
    "# First trial:\n",
    "# [ec2-user@ip-172-31-65-53 r2]$ grep \">\" dRdRP.fa | wc -l\n",
    "# 2871\n",
    "# [ec2-user@ip-172-31-65-53 r2]$ wc -l d_v_rdrp0.pro \n",
    "# 854 d_v_rdrp0.pro\n",
    "\n",
    "# Quenya\n",
    "IN='qRdRP.fa'\n",
    "GENOME='rdrp0'\n",
    "OUTNAME='q_v_rdrp0'\n",
    "\n",
    "# Diamond blastx alignment\n",
    "\n",
    "time cat $IN |\\\n",
    "diamond blastx \\\n",
    "  -d \"$GENOME\".dmnd \\\n",
    "  -p 4 \\\n",
    "  -k 1 \\\n",
    "  -f 6 qseqid qstart qend qlen qstrand \\\n",
    "       sseqid sstart send slen \\\n",
    "       pident evalue cigar \\\n",
    "       qseq qseq_translated \\\n",
    "  > \"$OUTNAME\".pro\n",
    "  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Link to all genbank CDS\n",
    "ln -s ../ntViro_gb201205.fa ./\n",
    "\n",
    "# Input file\n",
    "IN='ntViro_gb201205.fa'\n",
    "#IN='tmp.fa'\n",
    "\n",
    "GENOME='rdrp0'\n",
    "\n",
    "# Output name\n",
    "OUTNAME='gbViro_rdrp'\n",
    "\n",
    "# Diamond blastx alignment\n",
    "time cat $IN |\\\n",
    "diamond blastx \\\n",
    "  -d \"$GENOME\".dmnd \\\n",
    "  --unal 0 \\\n",
    "  -k 1 \\\n",
    "  -p 4 \\\n",
    "  -b 1 \\\n",
    "  -f 6 qseqid qstart qend qlen qstrand \\\n",
    "       sseqid sstart send slen \\\n",
    "       pident evalue cigar \\\n",
    "       qseq qseq_translated \\\n",
    "  > \"$OUTNAME\".pro\n",
    "  \n",
    "# real    1175m3.649s\n",
    "# user    4650m34.609s\n",
    "# sys     1m27.468s\n",
    "\n",
    "# wc -l\n",
    "# 390965 gbViro_rdrp.pro\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aws s3 cp gbViro_rdrp.pro s3://serratus-public/notebook/201210_rdrp0/rev2/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Notes:\n",
    "#\n",
    "# qseq\n",
    "# qseq_translated \n",
    "# full_qseq_mate\n",
    "\n",
    "# tmp.fa timing tests:\n",
    "# tail -n +20000000 ntViro_gb201205.fa | head -n 500000 > tmp.fa\n",
    "\n",
    "# default\n",
    "# real    5m22.409s\n",
    "# user    21m6.984s\n",
    "# sys     0m0.410s\n",
    "\n",
    "# --sensitive\n",
    "# real    11m19.287s\n",
    "# user    44m18.393s\n",
    "# sys     0m0.595s\n",
    "\n",
    "# output size is equal\n",
    "\n",
    "# NOTE: sseq returns protein sequence from the database, not the query\n",
    "# Changed field 15 sseq to qseq\n",
    "# use qseq instead"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "OUTNAME='gbViro_rdrp'\n",
    "\n",
    "# qseqid qstart qend qlen qstrand \\\n",
    "# sseqid sstart send slen \\\n",
    "# pident evalue cigar \\\n",
    "# qseq qseq_translated \\\n",
    "\n",
    "# AA of result hits\n",
    "cut -f1,13 $OUTNAME.pro |  sed 's/^/>/g' | sed 's/\\t/\\n/g' > $OUTNAME.aa_hit.fa\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# UCLUST\n",
    "OUTNAME='gbViro_rdrp'\n",
    "INPUT=\"$OUTNAME.aa_hit.fa\"\n",
    "OUTPUT=\"$OUTNAME.aa_hit.id75.fa\"\n",
    "\n",
    "usearch -sortbylength $INPUT \\\n",
    "   -fastaout tmp.sort.fa\n",
    "\n",
    "mv tmp.sort.fa $INPUT\n",
    "\n",
    "# Cluster at 75%\n",
    "usearch -cluster_smallmem $INPUT \\\n",
    "   -id 0.75 \\\n",
    "   -maxaccepts 4 \\\n",
    "   -maxrejects 64 \\\n",
    "   -maxhits 1 \\\n",
    "   -uc $OUTPUT.uc \\\n",
    "   -centroids $OUTPUT\n",
    "    \n",
    "#      Seqs  390965 (391.0k)\n",
    "#  Clusters  9591\n",
    "#  Max size  74741 (74.7k)\n",
    "#  Avg size  40.8\n",
    "#  Min size  1\n",
    "# Singletons  4462, 1.1% of seqs, 46.5% of clusters\n",
    "\n",
    "\n",
    "# UCLUST\n",
    "OUTPUT=\"$OUTNAME.aa_hit.id90.fa\"\n",
    "\n",
    "# Prune UNC\n",
    "usearch -cluster_smallmem $INPUT \\\n",
    "   -id 0.90 \\\n",
    "   -maxaccepts 4 \\\n",
    "   -maxrejects 64 \\\n",
    "   -maxhits 1 \\\n",
    "   -uc $OUTPUT.uc \\\n",
    "   -centroids $OUTPUT\n",
    "\n",
    "#       Seqs  390965 (391.0k)\n",
    "#   Clusters  15487 (15.5k)\n",
    "#  Max size  72136 (72.1k)\n",
    "#  Avg size  25.2\n",
    "#  Min size  1\n",
    "#Singletons  7716, 2.0% of seqs, 49.8% of clusters\n",
    "\n",
    "usearch -sortbylength $OUTNAME.aa_hit.id90.fa \\\n",
    "   -fastaout $OUTNAME.aa_hit.id90.sort.fa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Within the output there are lots of matches where the protein is yielding \"XXXXX\"\n",
    "\n",
    "```\n",
    "     12 >NC_031221.1\n",
    "     13 MTSSKSCVSVNEDEEYYKNIKGKIPSKASKKTIRYLMKRADPSGKFKDVDLSNLVKSKLLDTAKLEEINRLGFSIVKNNY\n",
    "     14 NTISNIITNYKDYTTNKVFDELYKVIKKMYNLSGTPDEILAGLKNKEELDIGFITIMENVFLYTDTLYFTLFPKNQRTAV\n",
    "     15 DREIYEGTLAAKLGLYFIERIYKEYAGDDESEAISMPGESKWMAIGRKRKTCLDFLVNHCSNQADVEAKEEIGAVELNLD\n",
    "     16 DQLEYPDLLKRDGEWSEVLGRKSRSSSISSQHSXXXXXXXXXXXXXXXXXXXXXXXXXXKVTKEAENTLKESQSLNATKN\n",
    "     17 TFDVLRNLIWSEDVEQEXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXLEINADMSKFSAKD\n",
    "     18 NLYKFMIFVILDPNLYKNEKYFILKFFCIYLRKRLIIPDQVFGTILDQTTSNDQCVFKKMTNNFETNHIQVSHNWLQGNL\n",
    "     19 NYISSCWHSMVMDVFKRVFTSAMKHLNTTVLVEPLVHSDDNSTSIAFISYNKDVNTVSLGAFAMDSIRRTLAKGSLILNT\n",
    "     20 KKTNVSTFHKEFVSLHDVNSEPVSIYGRFLYPVVGDCSYLGPYEDLSARLSTIQVALKHGCPPGVAHLGVALGVQMTYRT\n",
    "     21 YSMLPGQMHDPLPALDLQGESRFNIPVELGGYPDLDLWLFGSLGIEAHDLKKLAQIAEFYDLGGYKISLGWHSFIEDLTE\n",
    "     22 HNSQSIHRPYINGRIQYVKQSLWMDFYIMYMKNYAVGKDIQPDCDTGLSHDMKQR\n",
    "\n",
    "```\n",
    "\n",
    "Yet from GenBank:\n",
    "\n",
    "\n",
    "```\n",
    "                     ....................MTSSKSCVSVNEDEEYYKNIKGKIPSKASKKTIRYLMK\n",
    "                     RADPSGKFKDVDLSNLVKSKLLDTAKLEEINRLGFSIVKNNYNTISNIITNYKDYTTN\n",
    "                     KVFDELYKVIKKMYNLSGTPDEILAGLKNKEELDIGFITIMENVFLYTDTLYFTLFPK\n",
    "                     NQRTAVDREIYEGTLAAKLGLYFIERIYKEYAGDDESEAISMPGESKWMAIGRKRKTC\n",
    "                     LDFLVNHCSNQADVEAKEEIGAVELNLDDQLEYPDLLKRDGEWSEVLGRKSRSSSISS\n",
    "                     QHSDRGSPKSSKTQSKKGSSTGSKKGKKKKVTKEAENTLKESQSLNATKNTFDVLRNL\n",
    "                     IWSEDVEQEEMESVEKERQEKERKKEKVREEMEALKNEARTKPMILKVMNHKILSHGI\n",
    "                     LEINADMSKFSAKDNLYKFMIFVILDPNLYKNEKYFILKFFCIYLRKRLIIPDQVFGT\n",
    "                     ILDQTTSNDQCVFKKMTNNFETNHIQVSHNWLQGNLNYISSCWHSMVMDVFKRVFTSA\n",
    "                     MKHLNTTVLVEPLVHSDDNSTSIAFISYNKDVNTVSLGAFAMDSIRRTLAKGSLILNT\n",
    "                     KKTNVSTFHKEFVSLHDVNSEPVSIYGRFLYPVVGDCSYLGPYEDLSARLSTIQVALK\n",
    "                     HGCPPGVAHLGVALGVQMTYRTYSMLPGQMHDPLPALDLQGESRFNIPVELGGYPDLD\n",
    "                     LWLFGSLGIEAHDLKKLAQIAEFYDLGGYKISLGWHSFIEDLTEHNSQSIHRPYINGR\n",
    "                     IQYVKQSLWMDFYIMYMKNYAVGKDIQPDCDTGLSHDMKQR...\n",
    "```\n",
    "\n",
    "Matching line from the .pro output of diamond\n",
    "\n",
    "```\n",
    "NC_031221.1     2362    4686    7323    +       rdrp5.Peribunyaviridae.Shuangao_Insect_Virus_1:YP_009300681.1   1       775     775     100.0   0.0e+00775M\n",
    "\n",
    "ATGACAAGCTCCAAATCTTGTGTCAGCGTTAATGAAGACGAAGAATATTATAAAAATATAAAAGGGAAAATACCTAGCAAGGCAAGCAAAAAGACTATAAGATACCTGATGAAAAGGGCAGACCCATCTGGAAAATTCAAAGATGTAGATCTTAGCAACTTAGTAAAAAGTAAACTATTGGACACTGCTAAACTAGAGGAGATAAATAGGCTCGGTTTCTCTATTGTAAAGAACAATTATAACACAATATCCAATATAATCACAAATTACAAAGATTATACAACAAATAAAGTCTTTGATGAGTTGTACAAAGTTATTAAAAAGATGTACAACCTGTCTGGAACTCCTGATGAGATATTAGCAGGCTTAAAAAACAAAGAAGAACTAGACATTGGATTCATAACAATAATGGAGAATGTTTTTTTGTATACAGACACACTTTATTTTACTTTGTTCCCCAAGAACCAGAGGACAGCAGTAGACAGAGAGATATATGAAGGCACCTTGGCTGCAAAACTTGGTCTCTATTTTATAGAGAGAATATATAAGGAATATGCTGGAGATGACGAATCTGAAGCTATAAGCATGCCTGGGGAATCAAAATGGATGGCCATTGGTAGGAAAAGGAAAACCTGTTTAGACTTTCTTGTAAACCATTGCTCCAACCAGGCAGATGTGGAGGCAAAGGAAGAGATAGGTGCTGTGGAGTTAAATTTGGATGACCAACTAGAATACCCTGATTTATTGAAAAGAGATGGTGAATGGTCAGAAGTCTTGGGTAGAAAAAGCAGAAGCTCAAGTATATCAAGTCAACACTCCGATAGAGGCTCACCAAAAAGCTCCAAAACACAATCAAAAAAAGGATCATCGACAGGGAGTAAGAAAGGGAAGAAGAAAAAAGTAACAAAAGAAGCAGAAAACACCTTGAAAGAAAGTCAAAGTCTTAATGCTACTAAGAACACCTTTGATGTTTTAAGAAATTTGATCTGGTCTGAAGATGTAGAACAAGAAGAAATGGAGAGTGTAGAGAAGGAGAGACAGGAGAAAGAGAGAAAGAAAGAGAAAGTAAGAGAAGAAATGGAAGCATTAAAGAATGAAGCAAGAACAAAGCCTATGATTCTTAAAGTGATGAACCATAAAATTTTATCCCATGGAATATTAGAGATAAATGCAGACATGTCAAAGTTTAGTGCAAAGGATAATCTTTACAAGTTCATGATCTTTGTTATACTTGATCCAAATTTGTACAAAAACGAAAAGTATTTCATACTCAAATTCTTCTGTATTTATCTAAGGAAGCGATTGATTATACCAGATCAAGTATTTGGAACTATTCTTGACCAGACTACCAGCAATGATCAATGTGTATTTAAAAAGATGACAAACAACTTTGAGACAAATCACATTCAAGTTTCTCATAATTGGTTACAAGGGAATTTGAATTACATTTCCTCATGTTGGCATTCAATGGTAATGGATGTTTTCAAAAGAGTTTTTACATCTGCAATGAAACACTTAAACACAACAGTACTAGTAGAACCTTTGGTGCATTCTGACGATAACAGTACTAGCATTGCTTTTATCTCGTATAACAAAGATGTCAACACAGTTTCTTTAGGAGCATTTGCAATGGATAGTATTAGACGGACACTTGCAAAAGGTTCTTTGATACTCAACACAAAGAAAACAAATGTGTCTACTTTCCACAAAGAATTTGTCTCACTACATGACGTAAATTCAGAACCTGTTAGCATTTACGGTAGATTCCTTTATCCTGTTGTTGGAGATTGTTCATATTTAGGGCCTTATGAAGACTTGAGTGCGAGATTGTCGACTATACAAGTTGCTTTAAAACATGGTTGCCCACCAGGAGTTGCACATCTGGGTGTTGCTTTGGGAGTTCAAATGACATACAGAACTTACAGCATGTTACCAGGACAGATGCATGACCCATTGCCTGCATTAGACCTACAAGGAGAATCTCGATTCAATATTCCAGTTGAGTTGGGAGGATACCCTGATTTAGATCTATGGCTTTTTGGCAGTTTAGGTATAGAAGCACATGATCTGAAGAAATTAGCACAGATTGCAGAATTCTATGATCTAGGAGGCTATAAAATATCGTTAGGATGGCATAGTTTCATAGAAGATCTTACAGAACACAACTCCCAGTCTATTCATAGGCCATACATAAATGGAAGAATACAATATGTGAAGCAATCTTTATGGATGGATTTTTACATTATGTATATGAAAAATTATGCAGTTGGTAAAGACATACAACCAGACTGTGACACAGGGTTATCTCATGATATGAAGCAACGA\n",
    "\n",
    "MTSSKSCVSVNEDEEYYKNIKGKIPSKASKKTIRYLMKRADPSGKFKDVDLSNLVKSKLLDTAKLEEINRLGFSIVKNNYNTISNIITNYKDYTTNKVFDELYKVIKKMYNLSGTPDEILAGLKNKEELDIGFITIMENVFLYTDTLYFTLFPKNQRTAVDREIYEGTLAAKLGLYFIERIYKEYAGDDESEAISMPGESKWMAIGRKRKTCLDFLVNHCSNQADVEAKEEIGAVELNLDDQLEYPDLLKRDGEWSEVLGRKSRSSSISSQHSXXXXXXXXXXXXXXXXXXXXXXXXXXKVTKEAENTLKESQSLNATKNTFDVLRNLIWSEDVEQEXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXLEINADMSKFSAKDNLYKFMIFVILDPNLYKNEKYFILKFFCIYLRKRLIIPDQVFGTILDQTTSNDQCVFKKMTNNFETNHIQVSHNWLQGNLNYISSCWHSMVMDVFKRVFTSAMKHLNTTVLVEPLVHSDDNSTSIAFISYNKDVNTVSLGAFAMDSIRRTLAKGSLILNTKKTNVSTFHKEFVSLHDVNSEPVSIYGRFLYPVVGDCSYLGPYEDLSARLSTIQVALKHGCPPGVAHLGVALGVQMTYRTYSMLPGQMHDPLPALDLQGESRFNIPVELGGYPDLDLWLFGSLGIEAHDLKKLAQIAEFYDLGGYKISLGWHSFIEDLTEHNSQSIHRPYINGRIQYVKQSLWMDFYIMYMKNYAVGKDIQPDCDTGLSHDMKQR\n",
    "```\n",
    "\n",
    "The `qseq_tr` field appears to have a mask applied wihch removes sequences such as\n",
    "\n",
    "`DRGSPKSSKTQSKKGSSTGSKKGKKK`\n",
    "and\n",
    "`EMESVEKERQEKERKKEKVREEMEALKNEARTKPMILKVMNHKILSHGI`\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Revision 3 - RdRp from GenBank AA sequences (DEPRECATED SEE rev3B)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Install diamond\n",
    "DIAMONDVERSION='2.0.6-dev'\n",
    "cd ~/\n",
    "\n",
    "# Libraries for diamond\n",
    "sudo yum -y install git gcc gcc-c++ glibc-devel \\\n",
    "  cmake patch automake zlib-devel\n",
    "\n",
    "# grab latest with fix from Benjamin\n",
    "git clone https://github.com/bbuchfink/diamond.git\n",
    "cd diamond\n",
    "\n",
    "mkdir bin; cd bin\n",
    "cmake ..\n",
    "make -j4\n",
    "sudo make install\n",
    "\n",
    "# build\n",
    "# cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_BUILD_MARCH=nehalem ..\n",
    "# make && make install\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Work on EC2 Instance\n",
    "\n",
    "# Local usearch install\n",
    "#The clustered database was made with usearch:\n",
    "wget https://drive5.com/downloads/usearch11.0.667_i86linux32.gz\n",
    "gzip -dc usearch11.0.667_i86linux32.gz > usearch\n",
    "chmod 755 usearch; sudo mv usearch /usr/bin/usearch\n",
    "rm https://drive5.com/downloads/usearch11.0.667_i86linux32.gz\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Install seqkit\n",
    "wget https://github.com/shenwei356/seqkit/releases/download/v0.12.0/seqkit_linux_amd64.tar.gz\n",
    "  tar -xvf seqkit*\n",
    "  sudo mv seqkit /usr/local/bin/\n",
    "  rm seqkit_linux*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Install PathRacer\n",
    "mkdir -p ~/pr; cd ~/pr\n",
    "\n",
    "# Libraries\n",
    "sudo yum -y install gcc gcc-c++ glibc-devel \\\n",
    "  cmake patch automake zlib-devel bzip2-devel \\\n",
    "  openssl-devel\n",
    "\n",
    "# Update cmake to 3.18\n",
    "wget https://cmake.org/files/v3.18/cmake-3.18.0.tar.gz\n",
    "tar -xvzf cmake-3.18.0.tar.gz\n",
    "cd cmake-3.18.0\n",
    "./bootstrap\n",
    "make\n",
    "sudo make install\n",
    "cd ..\n",
    "\n",
    "# Install dev SPAdes\n",
    "wget http://cab.spbu.ru/files/pathracer/SPAdes-3.15.0-pathracer-dev.tar.gz\n",
    "tar -xvf SPAdes-3.15.0-pathracer-dev.tar.gz\n",
    "\n",
    "cd SPA*\n",
    "./spades_compile.sh\n",
    "\n",
    "sudo cp bin/* /usr/bin/\n",
    "sudo cp build_spades/bin/* /usr/bin/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Start set-up"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# rdrp0 v201213 for pilot run\n",
    "mkdir rdrp0; cd rdrp0\n",
    "\n",
    "# Download rdrp0\n",
    "aws s3 sync s3://serratus-public/notebook/201226_rdrp0/ ./\n",
    "\n",
    "# Revision 3 folder\n",
    "mkdir -p rev3; cd rev3\n",
    "GENOME='rdrp0'\n",
    "cp ../rev1/rdrp0_r1.fa ./rdrp0.fa\n",
    "\n",
    "diamond makedb --in $GENOME.fa -d $GENOME"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Link to all genbank CDS\n",
    "ln -s ../aaViro_gb240.fa ./\n",
    "\n",
    "# Input file\n",
    "IN='aaViro_gb240.fa'\n",
    "# Output name\n",
    "OUTNAME='gbViro_rdrp'\n",
    "# Genome ID\n",
    "GENOME='rdrp0'\n",
    "\n",
    "## test\n",
    "# head -n 100000 $IN > tmp.fa\n",
    "# IN='tmp.fa'\n",
    "# OUTNAME='tmp.pro'\n",
    "\n",
    "# Diamond blastx alignment\n",
    "time cat $IN |\\\n",
    "diamond blastp \\\n",
    "  -d \"$GENOME\".dmnd \\\n",
    "  --unal 0 \\\n",
    "  --masking 0 \\\n",
    "  -k 1 \\\n",
    "  -p 4 \\\n",
    "  -b 1 \\\n",
    "  -f 6 qseqid qstart qend qlen qstrand \\\n",
    "       sseqid sstart send slen \\\n",
    "       pident evalue cigar \\\n",
    "       qseq \\\n",
    "  > \"$OUTNAME\".pro\n",
    "\n",
    "# On C5.9xlarge\n",
    "# real    5m52.304s\n",
    "# user    22m23.987s\n",
    "# sys     0m11.048s\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# AA of result hits\n",
    "cut -f1,13 $OUTNAME.pro |  sed 's/^/>/g' | sed 's/\\t/\\n/g' > $OUTNAME.aa_hit.fa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# UCLUST\n",
    "OUTNAME='gbViro_rdrp'\n",
    "INPUT=\"$OUTNAME.aa_hit.fa\"\n",
    "\n",
    "usearch -sortbylength $INPUT \\\n",
    "   -fastaout tmp.sort.fa\n",
    "mv tmp.sort.fa $INPUT\n",
    "\n",
    "# Cluster at 45%\n",
    "OUTPUT=\"$OUTNAME.aa_hit.id45.fa\"\n",
    "usearch -cluster_smallmem $INPUT \\\n",
    "   -id 0.45 \\\n",
    "   -maxaccepts 4 \\\n",
    "   -maxrejects 64 \\\n",
    "   -maxhits 1 \\\n",
    "   -uc $OUTPUT.uc \\\n",
    "   -centroids $OUTPUT\n",
    "\n",
    "#      Seqs  354010 (354.0k)\n",
    "#  Clusters  2937\n",
    "#  Max size  85482 (85.5k)\n",
    "#  Avg size  120.5\n",
    "#  Min size  1\n",
    "#Singletons  1434, 0.4% of seqs, 48.8% of clusters\n",
    "\n",
    "# Cluster at 75%\n",
    "OUTPUT=\"$OUTNAME.aa_hit.id75.fa\"\n",
    "usearch -cluster_smallmem $INPUT \\\n",
    "   -id 0.75 \\\n",
    "   -maxaccepts 4 \\\n",
    "   -maxrejects 64 \\\n",
    "   -maxhits 1 \\\n",
    "   -uc $OUTPUT.uc \\\n",
    "   -centroids $OUTPUT\n",
    "\n",
    "#      Seqs  354010 (354.0k)\n",
    "#  Clusters  6911\n",
    "#  Max size  74942 (74.9k)\n",
    "#  Avg size  51.2\n",
    "#  Min size  1\n",
    "#Singletons  4116, 1.2% of seqs, 59.6% of clusters\n",
    "\n",
    "\n",
    "# Cluster at 90%\n",
    "OUTPUT=\"$OUTNAME.aa_hit.id90.fa\"\n",
    "usearch -cluster_smallmem $INPUT \\\n",
    "   -id 0.90 \\\n",
    "   -maxaccepts 4 \\\n",
    "   -maxrejects 64 \\\n",
    "   -maxhits 1 \\\n",
    "   -uc $OUTPUT.uc \\\n",
    "   -centroids $OUTPUT\n",
    "   \n",
    "#         Seqs  354010 (354.0k)\n",
    "#  Clusters  11712 (11.7k)\n",
    "#  Max size  72990 (73.0k)\n",
    "#  Avg size  30.2\n",
    "#  Min size  1\n",
    "#Singletons  6854, 1.9% of seqs, 58.5% of clusters\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract full-names from genbank\n",
    "ln -s ../aaViro_gb240.fa ./\n",
    "\n",
    "# Create index\n",
    "seqkit faidx -f aaViro_gb240.fa\n",
    "\n",
    "\n",
    "# Sort by length (large-small)\n",
    "usearch -sortbylength gbViro_rdrp.aa_hit.id90.fa \\\n",
    "   -fastaout id90.tmp.fa\n",
    "seqkit faidx id90.tmp.fa\n",
    "grep \">\" id90.tmp.fa > id90.header.tmp.fa\n",
    "\n",
    "while read -r line; do\n",
    "      acc=$(echo $line | cut -f1 -d' ' - | sed 's/>//g' - )\n",
    "      match=$(grep \"$acc\" aaViro_gb240.fa.seqkit.fai)\n",
    "      echo $acc $match >> output.tmp\n",
    "done < id90.header.tmp.fa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Extending Partial Matches\n",
    "\n",
    "Partial Matches. The diamond rdrp0 hits being returned for some highly diverged viruses are partial-length,\n",
    "\n",
    "```\n",
    "acc\tdescription\trdrp_len\tname\tacc_len\tfull_acc_match\n",
    "AXB38890.1\tpolyprotein, partial \t225\tPotato virus Y\t289\t0\n",
    "AAZ08330.1\tpolyprotein, partial \t225\tEustrephus virus Y\t479\t0\n",
    "ADM89618.1\tpolyprotein, partial \t225\tTrillium crinkled leaf virus\t455\t0\n",
    "QIR30294.1\tRNA-dependent RNA polymerase \t225\tPlasmopara viticola lesion associated narnavirus 15\t778\t0\n",
    "SBU65124.1\tRNA dependent RNA polymerase, partial \t225\tParsley mottle mimic virus\t225\t1\n",
    "AAX07288.1\tpolyprotein, partial \t225\tPapaya ringspot virus\t500\t0\n",
    "QIR30321.1\tRNA-dependent RNA polymerase \t225\tPlasmopara viticola lesion associated narnavirus 42\t777\t0\n",
    "AAZ08326.1\tpolyprotein, partial \t225\tSiratro 1 virus Y\t490\t0\n",
    "AVA17452.1\tRNA-dependent RNA polymerase \t225\tGigaspora margarita mitovirus 4\t1008\t0\n",
    "```\n",
    "\n",
    "For instance, QIR30294.1 Plasmopara viticola lesion associated narnavirus is 778 aa long, yet the match is 225 aa.\n",
    "\n",
    "\n",
    "GenBank Complete Query Record:\n",
    "```\n",
    ">QIR30294.1 RNA-dependent RNA polymerase [Plasmopara viticola lesion associated narnavirus 15]\n",
    "MRVRLPAYPFVTVKGVLPKSSPKPVAMLAGRRRPQRKVPSRRMPVHTKKLWEATWCALIS\n",
    "CGVGTRRGAWEIRRWVSLASDRMGWEEVARSLKAVCSELRSSALEGRRARLPKAHHFPAR\n",
    "LLSFLDRRLTVKGKLAFSRLARALPAASEKMKKEAVSTHARNLSARHVTPKAYLESIEEH\n",
    "VRYTLRGAFQQTSRYSVPSSSAAVVEADRKSGGYNKVLADLARAGWRGVAAGYTRSGVPT\n",
    "DTQYWKETSKLAHHFEHSTQQRIRSDKFNMYPTTVSAERNMGFATALMLRKAVGVRVVHH\n",
    "ASVIAELGMKARVITVPPASVFAQGDLLRQVLWPALVSRVPQILPYAPHTEEAILQSIGQ\n",
    "WVPGMVYLSADLTRATDGFGHDAILAVIRGLRKAGLPAFLCAELESNLGVGPDVHYVRYS\n",
    "KSQLTQQTWQEFGKRFGVNEDGETVDVPKVRGSLMGTPCSFSILSILNHWMSELLGRRRI\n",
    "ICGDDLAALTHPDNVSTYARRACAIGSELHQGKSFRSKIGFVFCEAYALTNGGGLRSFRP\n",
    "ASLKEFVRDGNGVMCQHGVDATSFNRLARCARTLYKRQRVIATKKRRYPELPAALGGLGH\n",
    "PCKGRLRIPAAGRAALFELYLCENTAHDGPHDPTRYVSTLTYPAVPASRREFRERVSQVR\n",
    "SWLDDKRIDEPQPGDKFATNREISAYASMCANLTYLAGGGRFRKSRPQEIKVSRQRWPKP\n",
    "IDGCRSGVLSSRTRINQVLDWDRRARSELGTYLDAPLQTHIRRRICAHREGDLPGDDR\n",
    "```\n",
    "Diamond Pro Output\n",
    "```\n",
    "QIR30294.1      301     525     778     +       rdrp1.Narnaviridae.Zhejiang_mosquito_virus_3:YP_009333331       126     359     513     32.9    7.3e-1738M1D13M2D14M3D28M1I12M1I3M3D2M1D19M10I38M11D46M        ASVIAELGMKARVITVPPASVFAQGDLLRQVLWPALVSRVPQILPYAPHTEEAILQSIGQWVPGMVYLSADLTRATDGFGHDAILAVIRGLRKAGLPAFLCAELESNLGVGPDVHYVRYSKSQLTQQTWQEFGKRFGVNEDGETVDVPKVRGSLMGTPCSFSILSILNHWMSELLGRRRIICGDDLAALTHPDNVSTYARRACAIGSELHQGKSFRSKIGFVFCE\n",
    "```\n",
    "\n",
    "Diamond Hit Sequence:\n",
    "```\n",
    ">QIR30294.1\n",
    "ASVIAELGMKARVITVPPASVFAQGDLLRQVLWPALVSRVPQILPYAPHTEEAILQSIGQ\n",
    "WVPGMVYLSADLTRATDGFGHDAILAVIRGLRKAGLPAFLCAELESNLGVGPDVHYVRYS\n",
    "KSQLTQQTWQEFGKRFGVNEDGETVDVPKVRGSLMGTPCSFSILSILNHWMSELLGRRRI\n",
    "ICGDDLAALTHPDNVSTYARRACAIGSELHQGKSFRSKIGFVFCE\n",
    "```\n",
    "\n",
    "rdrp0 hit record\n",
    "```\n",
    ">rdrp1.Narnaviridae.Zhejiang_mosquito_virus_3:YP_009333331\n",
    "LRAWAEIWTRERLGAGGRTLTNPAAHFSRSASATVSALKGGQLTELRQLPAVAEHYALIQ\n",
    "ELLGDQLDPVTGEPFEWDPYTFDADGGAIGLLTSASEADDLQILADCALRTATEHAADNT\n",
    "PIPMTATVISELGMKARVVTKPPAWAVVAGDACRKTVWPLLEGDRRIDLSGVRPTAEVLD\n",
    "AFHDNLAHSLVGARSTQFYSADLTAATDLMPFDVSRAMWNGLCDGLGATATAPLRKLGLY\n",
    "LLGPVQVSYPDLSALPASSKLYVAGERVECLSERGCMMGLPVSWTVLNLYNLAMADLACT\n",
    "PEGSPVLVNVAPAIARGDDLVAAIPAEEATRYEDLIAATGGEANRLKSFRSADAFVLAER\n",
    "TFEVGVLRRPNVELKQRGYVTRRTYRTAPLLAQFDSAELHGGTGDTRRLGDAPEVVAIRM\n",
    "ACDVPIRSLLGGGPRTVGANPVPDYVSIPPAAAACLAEFEGTRLYRSVAEGLMSVHRGLV\n",
    "ADLRRSAIPLFYPRELGGAGFPHPKGFAAAVAS\n",
    "```\n",
    "\n",
    "GenBank Record\n",
    "```\n",
    "Plasmopara viticola lesion associated narnavirus 15 isolate DMG-E_DN24203 RNA-dependent RNA polymerase gene, complete cds\n",
    "\n",
    ">MN539832.1 Plasmopara viticola lesion associated narnavirus 15 isolate DMG-E_DN24203 RNA-dependent RNA polymerase gene, complete cds\n",
    "CCTTGACAGAATCATGCGAGTGCGCCTTCCAGCGTACCCGTTTGTGACGGTTAAGGGTGTTCTTCCTAAG\n",
    "AGCTCTCCTAAGCCTGTAGCCATGCTAGCAGGTCGAAGGAGACCTCAAAGGAAGGTCCCTAGTCGTCGGA\n",
    "TGCCGGTGCACACCAAGAAGCTTTGGGAAGCTACTTGGTGTGCTTTGATTTCTTGTGGTGTTGGAACACG\n",
    "CCGTGGTGCGTGGGAGATACGTAGGTGGGTATCCCTTGCGAGCGATCGCATGGGATGGGAGGAAGTCGCA\n",
    "CGTTCCCTCAAGGCAGTCTGCAGTGAGCTCCGCAGTTCTGCTCTTGAGGGCCGACGAGCACGTCTTCCTA\n",
    "AGGCCCACCATTTCCCAGCACGGCTCCTTTCTTTCCTCGATCGTCGTCTCACTGTGAAAGGTAAGCTGGC\n",
    "CTTTAGTAGGCTGGCTCGCGCTCTCCCAGCGGCAAGTGAGAAGATGAAGAAAGAAGCCGTGTCCACGCAT\n",
    "GCACGGAACCTCTCGGCGCGACATGTTACTCCCAAGGCTTACCTTGAGAGCATTGAGGAACATGTCCGCT\n",
    "ACACGCTGAGAGGTGCGTTCCAACAAACGTCTCGTTACTCTGTGCCTTCTTCGTCTGCGGCTGTGGTTGA\n",
    "AGCGGATCGCAAGAGCGGAGGTTACAACAAAGTATTGGCAGATCTTGCCCGCGCAGGTTGGCGCGGTGTG\n",
    "GCAGCAGGTTACACTAGGAGCGGTGTACCCACTGACACACAATACTGGAAGGAAACCTCCAAACTTGCCC\n",
    "ACCACTTCGAGCACAGCACCCAGCAGCGAATACGGAGTGATAAGTTCAACATGTACCCAACAACTGTCAG\n",
    "TGCTGAGCGAAACATGGGTTTTGCAACCGCGCTTATGCTTCGCAAAGCAGTTGGTGTACGTGTTGTCCAT\n",
    "CACGCATCCGTAATCGCCGAGCTGGGGATGAAGGCACGAGTCATTACCGTCCCTCCTGCCAGTGTGTTTG\n",
    "CCCAGGGCGACTTGCTTCGGCAAGTCCTCTGGCCTGCCCTTGTTAGCAGAGTGCCTCAGATCCTTCCATA\n",
    "TGCTCCGCATACGGAAGAGGCTATTCTGCAAAGCATAGGACAGTGGGTGCCTGGTATGGTCTACCTTTCG\n",
    "GCAGATCTTACTAGGGCGACTGACGGTTTCGGACATGATGCCATCTTGGCAGTCATCAGGGGCCTGAGAA\n",
    "AGGCGGGTCTCCCCGCGTTTCTCTGTGCCGAACTCGAGAGTAATCTCGGAGTTGGCCCTGATGTGCATTA\n",
    "TGTCCGTTACAGCAAGTCGCAGCTTACCCAGCAAACATGGCAGGAGTTTGGAAAACGTTTCGGTGTGAAT\n",
    "GAGGATGGGGAGACCGTCGATGTCCCGAAGGTTAGGGGTTCCCTTATGGGTACTCCTTGTTCCTTCTCGA\n",
    "TATTGTCGATCCTCAATCACTGGATGAGCGAGCTTCTTGGTCGGCGAAGAATCATCTGCGGAGATGATCT\n",
    "TGCTGCGCTCACTCATCCCGATAACGTTTCTACCTACGCGCGCAGAGCCTGCGCAATAGGTAGCGAACTC\n",
    "CATCAAGGAAAGTCTTTCAGGTCGAAGATAGGTTTCGTGTTCTGCGAAGCCTACGCCCTTACCAATGGGG\n",
    "GTGGTCTTCGGTCCTTTAGACCCGCTTCCTTGAAGGAGTTCGTCCGTGACGGTAATGGGGTCATGTGTCA\n",
    "ACATGGCGTGGACGCAACTTCGTTCAACCGCTTGGCTCGTTGTGCTCGTACTCTGTACAAGCGTCAACGA\n",
    "GTGATCGCAACGAAGAAGCGTAGGTACCCTGAACTTCCGGCTGCGCTCGGAGGATTGGGGCATCCCTGTA\n",
    "AGGGGAGGCTCAGAATCCCGGCGGCTGGTCGTGCAGCATTGTTCGAGTTGTACCTGTGTGAGAACACCGC\n",
    "GCATGATGGGCCACATGACCCAACGAGATACGTTTCGACTCTTACTTACCCGGCGGTACCCGCGTCTAGA\n",
    "CGCGAGTTCCGTGAACGAGTATCCCAGGTTCGCTCTTGGCTCGATGACAAGAGGATTGATGAACCGCAAC\n",
    "CTGGGGACAAGTTCGCCACCAACAGGGAGATTAGTGCATATGCATCGATGTGTGCTAATCTCACCTATCT\n",
    "GGCAGGTGGTGGCAGGTTCCGTAAGAGTCGACCACAAGAAATCAAAGTTTCGAGACAGAGGTGGCCCAAG\n",
    "CCGATCGACGGATGTCGATCGGGGGTCTTGTCCAGTCGGACGAGGATTAACCAGGTCCTCGACTGGGACA\n",
    "GGAGAGCTCGTAGCGAGCTCGGCACCTATCTCGATGCACCGCTTCAGACGCACATCCGGCGTAGGATATG\n",
    "CGCCCACCGGGAGGGTGACCTCCCGGGAGATGACAGGTAGACTAGGG\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Global Alignments (NW)\n",
    "\n",
    "Diamond Local vs. complete rdrp0\n",
    "```\n",
    "Sbjct  1    LRAWAEIWTRERLGAGGRTLTNPAAHFSRSASATVSALKGGQLTELRQLPAVAEHYALIQ  60\n",
    "\n",
    "Sbjct  61   ELLGDQLDPVTGEPFEWDPYTFDADGGAIGLLTSASEADDLQILADCALRTATEHAADNT  120\n",
    "\n",
    "Query  1         ASVIAELGMKARVITVPPASVFAQGDLLRQVLWPALVS-RVPQILPYAPHTE--E  52\n",
    "                 A+VI+ELGMKARV+T PPA     GD  R+ +WP L   R   +    P  E  +\n",
    "Sbjct  121  PIPMTATVISELGMKARVVTKPPAWAVVAGDACRKTVWPLLEGDRRIDLSGVRPTAEVLD  180\n",
    "\n",
    "Query  53   AILQSIGQWVPGM---VYLSADLTRATDGFGHDAILAVIRGLRKAGLPAFLCAELESNLG  109\n",
    "            A   ++   + G     + SADLT ATD    D   A+  GL   GL A   A L   LG\n",
    "Sbjct  181  AFHDNLAHSLVGARSTQFYSADLTAATDLMPFDVSRAMWNGLCD-GLGATATAPLRK-LG  238\n",
    "\n",
    "Query  110  VGPDVHYVRYSKSQLTQQTWQEFGKRFGVNEDGETVDVPKVRGSLMGTPCSFSILSILNH  169\n",
    "            +     Y+     Q++            +   GE V+    RG +MG P S+++L++ N \n",
    "Sbjct  239  L-----YL-LGPVQVSYPDLSALPASSKLYVAGERVECLSERGCMMGLPVSWTVLNLYNL  292\n",
    "\n",
    "Query  170  WMSELLGRRR-----------IICGDDLAALTHPDNVSTYARRACAIGSELHQGKSFRSK  218\n",
    "             M++L                I  GDDL A    +  + Y     A G E ++ KSFRS \n",
    "Sbjct  293  AMADLACTPEGSPVLVNVAPAIARGDDLVAAIPAEEATRYEDLIAATGGEANRLKSFRSA  352\n",
    "\n",
    "Query  219  IGFVFCE                                                       225\n",
    "              FV  E                                                     \n",
    "Sbjct  353  DAFVLAERTFEVGVLRRPNVELKQRGYVTRRTYRTAPLLAQFDSAELHGGTGDTRRLGDA  412\n",
    "\n",
    "Sbjct  413  PEVVAIRMACDVPIRSLLGGGPRTVGANPVPDYVSIPPAAAACLAEFEGTRLYRSVAEGL  472\n",
    "\n",
    "Sbjct  473  MSVHRGLVADLRRSAIPLFYPRELGGAGFPHPKGFAAAVAS  513\n",
    "```\n",
    "\n",
    "Total Record vs. complete rdrp0\n",
    "```\n",
    "Query  1    MRVRLPAYPFVTVKGVLPKSSPKPVAMLAGRRRPQRKVPSRRMPVHTKKLWEATWCALIS  60\n",
    "            +R                                                W   W     \n",
    "Sbjct  1    LRA-----------------------------------------------WAEIW-----  8\n",
    "\n",
    "Query  61   CGVGTRRGAWEIRRWVSLASDRMGWEEVARSLKAVCSELRSSALEGRRARLPKAHHFPAR  120\n",
    "                TR              +R+G                     GR    P AH     \n",
    "Sbjct  9    ----TR--------------ERLG-------------------AGGRTLTNPAAH-----  26\n",
    "\n",
    "Query  121  LLSFLDRRLTVKGKLAFSRLARALPAASEKMKKEAVSTHARNLSARHVTPKAYLESIEEH  180\n",
    "                            FSR A A  +A               L    +T    L ++ EH\n",
    "Sbjct  27   ----------------FSRSASATVSA---------------LKGGQLTELRQLPAVAEH  55\n",
    "\n",
    "Query  181  VRYTLRGAFQQTSRYSVPSSSAAVVEADRKSGGYNKVLADLARAGWRGVAAGYTRSGVPT  240\n",
    "              Y L    Q+     +   +    E D  +   +            G A G   S    \n",
    "Sbjct  56   --YAL---IQELLGDQLDPVTGEPFEWDPYTFDAD------------GGAIGLLTSASEA  98\n",
    "\n",
    "Query  241  DTQYWKETSKLAHHFEHSTQQRIRSDKFNMYPTTVSAERNMGFATALMLRKAVGVRVVHH  300\n",
    "            D         L    EH+             P T                          \n",
    "Sbjct  99   DDLQILADCALRTATEHAADN-------TPIPMT--------------------------  125\n",
    "\n",
    "Query  301  ASVIAELGMKARVITVPPASVFAQGDLLRQVLWPALVS-RVPQILPYAPHTE--EAILQS  357\n",
    "            A+VI+ELGMKARV+T PPA     GD  R+ +WP L   R   +    P  E  +A   +\n",
    "Sbjct  126  ATVISELGMKARVVTKPPAWAVVAGDACRKTVWPLLEGDRRIDLSGVRPTAEVLDAFHDN  185\n",
    "\n",
    "Query  358  IGQWVPGM---VYLSADLTRATDGFGHDAILAVIRGLRKAGLPAFLCAELESNLGVGPDV  414\n",
    "            +   + G     + SADLT ATD    D   A+  GL   GL A   A L   LG+    \n",
    "Sbjct  186  LAHSLVGARSTQFYSADLTAATDLMPFDVSRAMWNGLCD-GLGATATAPLRK-LGL----  239\n",
    "\n",
    "Query  415  HYVRYSKSQLTQQTWQEFGKRFGVNEDGETVDVPKVRGSLMGTPCSFSILSILNHWMSEL  474\n",
    "             Y+     Q++            +   GE V+    RG +MG P S+++L++ N  M++L\n",
    "Sbjct  240  -YL-LGPVQVSYPDLSALPASSKLYVAGERVECLSERGCMMGLPVSWTVLNLYNLAMADL  297\n",
    "\n",
    "Query  475  LGRRR-----------IICGDDLAALTHPDNVSTYARRACAIGSELHQGKSFRSKIGFVF  523\n",
    "                            I  GDDL A    +  + Y     A G E ++ KSFRS   FV \n",
    "Sbjct  298  ACTPEGSPVLVNVAPAIARGDDLVAAIPAEEATRYEDLIAATGGEANRLKSFRSADAFVL  357\n",
    "\n",
    "Query  524  CEAYALTNGGGLRSFRPASLKEFVRDGNGVMCQHGVDATSFNRLARCARTLYKRQRVIAT  583\n",
    "             E          R+F             GV+ +  V+                +QR   T\n",
    "Sbjct  358  AE----------RTFEV-----------GVLRRPNVEL---------------KQRGYVT  381\n",
    "\n",
    "Query  584  KKRRYPELPAALGGLGHPCKGRLRIPAAGRAALFELYLCENTAHDGPHDPTRYVSTLTYP  643\n",
    "            + R Y   P                      A F+        H G  D  R       P\n",
    "Sbjct  382  R-RTYRTAPLL--------------------AQFD----SAELHGGTGDTRRLGDA---P  413\n",
    "\n",
    "Query  644  AVPASRREFRERVSQVRSWLDDKRIDEPQPGDKFATNREISAYASMCANLTYLAGGGRFR  703\n",
    "             V A R         +RS L       P+          +S   +  A L    G   +R\n",
    "Sbjct  414  EVVAIRMACDV---PIRSLLGGG----PRTVGANPVPDYVSIPPAAAACLAEFEGTRLYR  466\n",
    "\n",
    "Query  704  KSRPQEIKVSRQRWPKPIDGCRSGVLSSRTRINQVLDWDRRARSELGTYLDAPLQTHIRR  763\n",
    "                  + V R            G+++   R    L + R    ELG     P       \n",
    "Sbjct  467  SVAEGLMSVHR------------GLVADLRRSAIPLFYPR----ELGG-AGFP-------  502\n",
    "\n",
    "Query  764  RICAHREGDLPGDDR  778\n",
    "                H +G       \n",
    "Sbjct  503  ----HPKGFAAAVAS  513\n",
    "```\n",
    "\n",
    "And semi-global (from EMBOSS Needle)\n",
    "\n",
    "```\n",
    "QIR30294.1      MRVRLPAYPFVTVKGVLPKSSPKPVAMLAGRRRPQRKVPSRRMPVHTKKLWEATWCALIS\n",
    "YP_009333331    ------------------------------------------------------------\n",
    "                                                                            \n",
    "QIR30294.1      CGVGTRRGAWEIRRWVSL-ASDRMGWEEVARSLKAVCSELRSSALEGRRARLPKAHHFPA\n",
    "YP_009333331    -----------LRAWAEIWTRERLG-------------------AGGRTLTNPAAH----\n",
    "                                                                            \n",
    "QIR30294.1      RLLSFLDRRLTVKGKLAFSRLARALPAASEKMKKEAVSTHARNLSARHVTPKAYLESIEE\n",
    "YP_009333331    -----------------FSRSASATVSA---------------LKGGQLTELRQLPAVAE\n",
    "                                                                            \n",
    "QIR30294.1      HVRYTL------------RGAFQQTSRYSVPSSSAAV------VEADRKSGGYNKVLADL\n",
    "YP_009333331    H--YALIQELLGDQLDPVTGEPFEWDPYTFDADGGAIGLLTSASEADDL-----QILADC\n",
    "                                                                            \n",
    "QIR30294.1      ARAGWRGVAAGYTRSGVPTDTQYWKETSKLAHHFEHSTQQRIRSDKFNMYPTTVSAERNM\n",
    "YP_009333331    A---------------LRTATEHAADNTPI--------------------PMT-------\n",
    "                                                                            \n",
    "QIR30294.1      GFATALMLRKAVGVRVVHHASVIAELGMKARVITVPPASVFAQGDLLRQVLWPALV-SRV\n",
    "YP_009333331    -------------------ATVISELGMKARVVTKPPAWAVVAGDACRKTVWPLLEGDRR\n",
    "                                                                            \n",
    "QIR30294.1      PQILPYAPHTE--EAILQSIGQWVPG---MVYLSADLTRATDGFGHDAILAVIRGLRKAG\n",
    "YP_009333331    IDLSGVRPTAEVLDAFHDNLAHSLVGARSTQFYSADLTAATDLMPFDVSRAMWNGLCD-G\n",
    "                                                                            \n",
    "QIR30294.1      LPAFLCAELESNLG---VGP-DVHYVRYSKSQLTQQTWQEFGKRFGVNEDGETVDVPKVR\n",
    "YP_009333331    LGATATAPLR-KLGLYLLGPVQVSYPDLSALPASSKLYVA----------GERVECLSER\n",
    "                                                                            \n",
    "QIR30294.1      GSLMGTPCSFSILSILNHWMSELLGRRR-----------IICGDDLAALTHPDNVSTYAR\n",
    "YP_009333331    GCMMGLPVSWTVLNLYNLAMADLACTPEGSPVLVNVAPAIARGDDLVAAIPAEEATRYED\n",
    "                                                                            \n",
    "QIR30294.1      RACAIGSELHQGKSFRSKIGFVFCEAYALTNGGGLRSFRPASLKEFVRDGNGVMCQHG--\n",
    "YP_009333331    LIAATGGEANRLKSFRSADAFVLAE----------RTFEVGVL----RRPNVELKQRGYV\n",
    "                                                                            \n",
    "QIR30294.1      -------------VDATSFNRLARCARTLYKRQRVIATKKRRYPELPAALGG----LG-H\n",
    "YP_009333331    TRRTYRTAPLLAQFDSAELHGGTGDTRRLGDAPEVVAIRMACDVPIRSLLGGGPRTVGAN\n",
    "                                                                            \n",
    "QIR30294.1      PCKGRLRIPAAGRAALFE-----LY--LCEN--TAHDG-PHDPTRYVSTLTYP-------\n",
    "YP_009333331    PVPDYVSIPPAAAACLAEFEGTRLYRSVAEGLMSVHRGLVADLRRSAIPLFYPRELGGAG\n",
    "                                                                            \n",
    "QIR30294.1      --------AVPASRREFRERVSQVRSWLDDKRIDEPQPGDKFATNREISAYASMCANLTY\n",
    "YP_009333331    FPHPKGFAAAVAS-----------------------------------------------\n",
    "                                                                            \n",
    "QIR30294.1      LAGGGRFRKSRPQEIKVSRQRWPKPIDGCRSGVLSSRTRINQVLDWDRRARSELGTYLDA\n",
    "YP_009333331    ------------------------------------------------------------\n",
    "                                                                            \n",
    "QIR30294.1      PLQTHIRRRICAHREGDLPGDDR\n",
    "YP_009333331    -----------------------\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Pathracer Semi-Global Alignment Testing"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "RdRP_1 \tPF00680.20 \tCL0027 \tRNA dependent RNA polymerase \t\n",
    "1832 \t2196 \t8.7e-19 \t4.9e-23"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compare rdrp0 hit versus pr\n",
    "# using QIR30294.1\n",
    "\n",
    "seqkit grep -r -p \"QIR30294.1\" aaViro_gb240.fa > test/narna_fl.fa\n",
    "\n",
    "grep \"QIR30294.1\" gbViro_rdrp.pro > test/narna.dmnd.pro\n",
    "seqkit grep -r -p \"QIR30294.1\" gbViro_rdrp.aa_hit.fa > test/narna.dmnd.fa\n",
    "\n",
    "\n",
    "# Download rdrp1234q.hmm from Rayan\n",
    "wget https://serratus-rayan.s3.amazonaws.com/tmp/RdRP_1234q.hmm ./\n",
    "\n",
    "# Search sequence with PathRacer\n",
    "pathracer-seq-fs RdRP_1234q.hmm narna_nt.fa --global --output pr_out  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Merge GenBank hits into rdrp0 / Assign Taxonomy\n",
    "\n",
    "Shortest sequences in rdrp0\n",
    "\n",
    "```\n",
    "rdrp2.unc0371.Hubei_sobemo_like_virus_16:YP_009330046   280     4764591 80      81\n",
    "rdrp2.Luteoviridae.Spinach_yellows_virus:CRL92755       275     4764926 80      81\n",
    "rdrp2.Luteoviridae.Lettuce_mild_yellows_virus:CRL92746  273     4765261 80      81\n",
    "rdrp3.Betaflexiviridae.Phlomis_mottle_virus:CAP46903    273     4765592 80      81\n",
    "rdrp2.Luteoviridae.Hubei_polero_like_virus_1:YP_009330062       270     4765928 80      81\n",
    "rdrp2.Picobirnaviridae.Genet_fecal_picobirnavirus:AIB06803      270     4766262 80      81\n",
    "rdrp2.Luteoviridae.Lettuce_yellows_virus:CRL92742       266     4766587 80      81\n",
    "rdrp2.unc0373.Hubei_sobemo_like_virus_14:YP_009330128   265     4766912 80      81\n",
    "rdrp2.unc0370.Hubei_sobemo_like_virus_17:APG75899       265     4767232 80      81\n",
    "rdrp2.unc0066.Wuhan_arthropod_virus_4:YP_009342307      261     4767553 80      81\n",
    "rdrp2.unc0368.Hubei_myriapoda_virus_9:YP_009345131      255     4767870 80      81\n",
    "rdrp2.unc0378.Beihai_sobemo_like_virus_16:YP_009336705  254     4768185 80      81\n",
    "rdrp2.unc0225.Prestney_Burn_virus:AMO03211      253     4768487 80      81\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mkdir -p merge; cd merge\n",
    "cp ../gbViro_rdrp.aa_hit.fa ./\n",
    "cp ../rdrp0.fa ./\n",
    "\n",
    "RDRP=\"rdrp0\"\n",
    "GB=\"gbViro_rdrp.aa_hit\"\n",
    "\n",
    "# Minimum length\n",
    "MIN='200'\n",
    "\n",
    "# Sort and filter by 200a\n",
    "usearch -sortbylength $RDRP.fa \\\n",
    "  -minseqlength $MIN \\\n",
    "  -fastaout $RDRP.sort.fa\n",
    "\n",
    "usearch -sortbylength $GB.fa \\\n",
    "  -minseqlength $MIN \\\n",
    "  -fastaout $GB.sort.fa\n",
    "\n",
    "\n",
    "cat $RDRP.sort.fa $GB.sort.fa > merge.fa.tmp\n",
    "\n",
    "# Cluster\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# UCLUST\n",
    "INPUT='merge.fa.tmp'\n",
    "OUTNAME='gb240_r0'\n",
    "\n",
    "# Cluster/Merge\n",
    "\n",
    "#45%\n",
    "OUTPUT=\"$OUTNAME.id45\"\n",
    "usearch -cluster_smallmem $INPUT \\\n",
    "   -id 0.45 \\\n",
    "   -sortedby other \\\n",
    "   -maxaccepts 4 \\\n",
    "   -maxrejects 64 \\\n",
    "   -maxhits 1 \\\n",
    "   -uc $OUTPUT.uc \\\n",
    "   -centroids $OUTPUT.fa\n",
    "\n",
    "#      Seqs  229046 (229.0k)\n",
    "#  Clusters  4437\n",
    "#  Max size  85638 (85.6k)\n",
    "#  Avg size  51.6\n",
    "#  Min size  1\n",
    "# Singletons  1922, 0.8% of seqs, 43.3% of clusters\n",
    "\n",
    "\n",
    "#90%\n",
    "OUTPUT=\"$OUTNAME.id90\"\n",
    "usearch -cluster_smallmem $INPUT \\\n",
    "   -id 0.90 \\\n",
    "   -sortedby other \\\n",
    "   -maxaccepts 4 \\\n",
    "   -maxrejects 64 \\\n",
    "   -maxhits 1 \\\n",
    "   -uc $OUTPUT.uc \\\n",
    "   -centroids $OUTPUT.fa\n",
    "\n",
    "#       Seqs  229046 (229.0k)\n",
    "#  Clusters  12477 (12.5k)\n",
    "#  Max size  72655 (72.7k)\n",
    "#  Avg size  18.4\n",
    "#  Min size  1\n",
    "# Singletons  7038, 3.1% of seqs, 56.4% of clusters\n",
    "\n",
    "# Number of non-wolf18 sequences that made it through\n",
    "# grep -r \">[^rdrp]\" gb240_r0.id45.fa | wc -l \n",
    "# 776 / 4437\n",
    "#\n",
    "# grep -r \">[^rdrp]\" gb240_r0.id90.fa | wc -l\n",
    "# 3741 / 12477\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Revision 3B USEARCH Semi-Global Extension\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Revision 3B folder\n",
    "mkdir -p rev3B; cd rev3B\n",
    "\n",
    "# Use rdrp0_revision1 as start-point\n",
    "GENOME='rdrp0'\n",
    "cp ../rev1/rdrp0_r1.fa ./rdrp0.fa\n",
    "diamond makedb --in $GENOME.fa -d $GENOME"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Link to GenBank v241 (Dec2020 Release)\n",
    "ln -s ../vgb241/vgb241.fa ./\n",
    "\n",
    "mkdir 0_rdrp_search; cd 0_rdrp_search\n",
    "\n",
    "# Input file\n",
    "IN='vgb241.fa'\n",
    "# Output name\n",
    "OUTNAME='gb241_r0'\n",
    "# Genome ID\n",
    "GENOME='rdrp0'\n",
    "\n",
    "# Diamond blastx alignment\n",
    "time cat $IN |\\\n",
    "diamond blastp \\\n",
    "  -d \"$GENOME\".dmnd \\\n",
    "  --unal 0 \\\n",
    "  --masking 0 \\\n",
    "  --ultra-sensitive \\\n",
    "  -k 1 \\\n",
    "  -p 30 \\\n",
    "  -b 2 \\\n",
    "  -f 6 qseqid qstart qend qlen qstrand \\\n",
    "       sseqid sstart send slen \\\n",
    "       pident evalue bitscore cigar \\\n",
    "       qseq \\\n",
    "  > \"$OUTNAME\".raw.pro\n",
    "\n",
    "# On C5.9xlarge\n",
    "# real    31m22.996s\n",
    "# user    890m28.269s\n",
    "# sys     1m16.513s\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter for High-Confidence alignments only\n",
    "\n",
    "# Output name\n",
    "OUTNAME='gb241_r0'\n",
    "\n",
    "# Threshold for \"cut-offs\"\n",
    "EVALUE=6  # -log(E-value) > $EVAL cut-off\n",
    "RCOV=50 # rdrp-coverage > $RCOV cut-off\n",
    "\n",
    "while read -r line; do\n",
    "  \n",
    "  # Return -log(e-value); zero == 999\n",
    "  e_value=$(echo $line | cut -f 11 -d' ' - \\\n",
    "    | sed 's/[0-9\\.]*e.//g' - \\\n",
    "    | sed 's/^0//g' -)\n",
    "  if (( $e_value == 0 )); then\n",
    "    e_value='999'\n",
    "  fi\n",
    "  \n",
    "  # rdrp-match coordinates\n",
    "  r_start=$(echo $line | cut -f 7 -d' ' - )\n",
    "  r_end=$(echo $line | cut -f 8 -d' ' - )\n",
    "  r_len=$(echo $line | cut -f 9 -d' ' - )\n",
    "  \n",
    "  # convert to percent\n",
    "  r_cov=$( echo \"$r_end - $r_start + 1\" | bc )\n",
    "  r_covpct=$( echo \"scale=0; 100 * $r_cov / $r_len\" | bc )\n",
    "  \n",
    "  #echo $r_start $r_end $r_len $r_covpct $e_value\n",
    " \n",
    "  if (( $r_len > $RCOV )) && (( $e_value > $EVALUE )); then\n",
    "    #echo HIT\n",
    "    echo $line | sed 's/ /\\t/g' >> \"$OUTNAME\".local.pro\n",
    "  else\n",
    "    #echo MISS\n",
    "    echo $line | sed 's/ /\\t/g' >> \"$OUTNAME\".REJECT.pro\n",
    "  fi\n",
    "  # I guess they never miss huh\n",
    "done < \"$OUTNAME\".raw.pro\n",
    "\n",
    "mkdir raw;\n",
    "mv \"$OUTNAME\".raw.pro raw/\n",
    "mv\"$OUTNAME\".REJECT.pro raw/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Output name\n",
    "OUTNAME='gb241_r0'\n",
    "\n",
    "# Make fasta of output\n",
    "cut -f1,14 $OUTNAME.local.pro \\\n",
    "  |  sed 's/^/>/g' \\\n",
    "  | sed 's/\\t/\\n/g' \\\n",
    "  > $OUTNAME.local.fa\n",
    "\n",
    "usearch -sortbylength $OUTNAME.local.fa \\\n",
    "   -fastaout tmp.sort.fa\n",
    "   \n",
    "mv tmp.sort.fa $OUTNAME.local.fa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# UCLUST Local Alignments\n",
    "OUTNAME='gb241_r0'\n",
    "INPUT=\"gb241_r0.local.fa\"\n",
    "\n",
    "function uclust () {\n",
    "\n",
    "  INPUT=$1\n",
    "  OUTNAME=$2\n",
    "  ID=$3\n",
    "  \n",
    "  # uclust\n",
    "  OUTPUT=\"$OUTNAME.local.id$ID\"\n",
    "  usearch -cluster_smallmem $INPUT \\\n",
    "     -id 0.$ID \\\n",
    "     -maxaccepts 4 \\\n",
    "     -maxrejects 64 \\\n",
    "     -maxhits 1 \\\n",
    "     -uc $OUTPUT.uc \\\n",
    "     -centroids $OUTPUT.fa\n",
    "\n",
    "  mkdir id$ID; mv *.id$ID.* id$ID/\n",
    "\n",
    "}\n",
    "\n",
    "# Cluster at 45%\n",
    "uclust $INPUT $OUTNAME 45\n",
    "\n",
    "#      Seqs  367343 (367.3k)\n",
    "#  Clusters  2912\n",
    "#  Max size  86855 (86.9k)\n",
    "#  Avg size  126.1\n",
    "#  Min size  1\n",
    "#Singletons  1365, 0.4% of seqs, 46.9% of clusters\n",
    "\n",
    "\n",
    "# Cluster at 75%\n",
    "uclust $INPUT $OUTNAME 75\n",
    "\n",
    "#      Seqs  367343 (367.3k)\n",
    "#  Clusters  7052\n",
    "#  Max size  75171 (75.2k)\n",
    "#  Avg size  52.1\n",
    "#  Min size  1\n",
    "#Singletons  4177, 1.1% of seqs, 59.2% of clusters\n",
    "\n",
    "\n",
    "\n",
    "# Cluster at 90%\n",
    "uclust $INPUT $OUTNAME 90\n",
    "\n",
    "#      Seqs  367343 (367.3k)\n",
    "#  Clusters  12002 (12.0k)\n",
    "#  Max size  73247 (73.2k)\n",
    "#  Avg size  30.6\n",
    "#  Min size  1\n",
    "#Singletons  6946, 1.9% of seqs, 57.9% of clusters\n",
    "\n",
    "\n",
    "\n",
    "# Cluster at 95%\n",
    "uclust $INPUT $OUTNAME 95\n",
    "\n",
    "#      Seqs  367343 (367.3k)\n",
    "#  Clusters  18733 (18.7k)\n",
    "#  Max size  41281 (41.3k)\n",
    "#  Avg size  19.6\n",
    "#  Min size  1\n",
    "#Singletons  10944 (10.9k), 3.0% of seqs, 58.4% of clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract Full Header from GenBank Local Matches {DEPRECATED}\n",
    "ln -s ../../vgb241/vgb241.fa.fai\n",
    "\n",
    "function reheader {\n",
    "  QUERYFA=$1    # Diamond Local Matches\n",
    "  SUBJECTFAI=$2 # GenBank FAI index\n",
    "  OUTPUT=$3     # Output name\n",
    "\n",
    "  # Sort Matching Diamond Accessions\n",
    "  usearch -sortbylength $QUERYFA \\\n",
    "    -fastaout tmp.sort\n",
    "\n",
    "  grep \">\" tmp.sort > tmp.qheader\n",
    "\n",
    "  rm -f output.tmp\n",
    "  while read -r line; do\n",
    "      acc=$(echo $line | cut -f1 -d' ' - | sed 's/>//g' - )\n",
    "      match=$(grep \"$acc\" $SUBJECTFAI)\n",
    "      echo $acc $match >> output.tmp\n",
    "  done < tmp.qheader\n",
    "\n",
    "  mv output.tmp > $OUTPUT\n",
    "\n",
    "  rm -f tmp.sort tmp.qheader\n",
    "}\n",
    "\n",
    "OUTNAME='gb241_r0'\n",
    "ID='90'\n",
    "QUERYFA=\"id$ID/$OUTNAME.local.id$ID.fa\"\n",
    "SUBJECTFAI='vgb241.fa.fai'\n",
    "OUTPUT='gb241_r0.local.id90.reheader'\n",
    "\n",
    "reheader $QUERYFA $SUBJECTFAI $OUTPUT\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1. Cluster all putative-RdRp in GenBank into rdrp0 at 90%\n",
    "#    Isolate centroids that are from GenBank as putative novel-RdRp\n",
    "#    Note: these are \"local\" matches, and require extension\n",
    "\n",
    "mkdir 1_merge_rdrp; cd 1_merge_rdrp\n",
    "\n",
    "RDRP=\"rdrp0.fa\"\n",
    "GBLOCAL=\"gb241_r0.local.fa\"\n",
    "OUTNAME=\"rdrp0_gb241\"\n",
    "\n",
    "# rdrp0 == wolf18/wolf20 reference rdrp-set\n",
    "usearch -sortbylength ../$RDRP \\\n",
    "   -fastaout ./$RDRP\n",
    "   \n",
    "# Local GenBank matches to rdrp0 (via diamond)\n",
    "usearch -sortbylength ../$GBLOCAL \\\n",
    "   -fastaout ./$GBLOCAL\n",
    "   \n",
    "# Order the files such that rdrp0 takes priority over genbank\n",
    "cat $RDRP $GBLOCAL > cat_rdrp.fa\n",
    "\n",
    "function uclust () {\n",
    "\n",
    "  INPUT=$1\n",
    "  OUTNAME=$2\n",
    "  ID=$3\n",
    "  \n",
    "  # uclust\n",
    "  OUTPUT=\"$OUTNAME.id$ID\"\n",
    "  usearch -cluster_smallmem $INPUT \\\n",
    "     -sortedby other \\\n",
    "     -id 0.$ID \\\n",
    "     -maxaccepts 4 \\\n",
    "     -maxrejects 64 \\\n",
    "     -maxhits 1 \\\n",
    "     -uc $OUTPUT.uc \\\n",
    "     -centroids $OUTPUT.fa\n",
    "\n",
    "  mkdir id$ID; mv *.id$ID.* id$ID/\n",
    "\n",
    "}\n",
    "\n",
    "# Cluster rdrp0 and genbank_rdrp0_matches at 90%\n",
    "uclust cat_rdrp.fa $OUTNAME 90\n",
    "\n",
    "# Retrieve non-rdrp0 sequences (GenBank-derived RdRp)\n",
    "grep \">\" id90/$OUTNAME.id90.fa \\\n",
    "  | grep -v \">rdrp\" - \\\n",
    "  | sed 's/>//g' - \\\n",
    "  > new.rdrp.acc\n",
    "\n",
    "# 7633 sequences\n",
    "cd .."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 2. For all New GenBank-derived RdRpR\n",
    "#    retrieve the full-length sequence/ORF from GenBank\n",
    "mkdir 2_retrieve_orf; cd 2_retrieve_orf\n",
    "\n",
    "cp ../1_merge_rdrp/new.rdrp.acc ./new.rdrp.acc\n",
    "\n",
    "# Viral GenBank (all)\n",
    "VGB='vgb241.fa'\n",
    "ORFFA='gb241.orf.fa'\n",
    "seqkit grep -f new.rdrp.acc ../$VGB > $ORFFA\n",
    "\n",
    "cd .."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 3. For all New GenBank ORF (containing RdRp)\n",
    "#    Semi-global align the ORF to the original rdrp0\n",
    "#    to \"trim\" the sequences to the wolf18 boundary\n",
    "mkdir 3_trim_orf; cd 3_trim_orf\n",
    "\n",
    "# Perform semi-global extension of potential hits\n",
    "wget https://drive5.com/tmp/usearch12_trim ./\n",
    "chmod 755 usearch12_trim\n",
    "\n",
    "RDRP='rdrp0.fa'\n",
    "ln -s ../$RDRP ./\n",
    "ORFFA='gb241.orf.fa'\n",
    "ln -s ../2_retrieve_orf/$ORFFA ./\n",
    "\n",
    "OUTFA=\"gb241.rdrp.fa\"\n",
    "\n",
    "# NW Semi-Global Extension\n",
    "./usearch12_trim -usearch_global $ORFFA \\\n",
    "    -id 0.01 \\\n",
    "    -fulldp \\\n",
    "    -maxaccepts 8 \\\n",
    "    -maxrejects 32 \\\n",
    "    -top_hit_only \\\n",
    "    -db rdrp0.fa \\\n",
    "    -userfields query+target+id+qtrimlo+qtrimhi \\\n",
    "    -userout results.tsv \\\n",
    "    -trimout trimmed_output.fa\n",
    "    \n",
    "usearch -sortbylength trimmed_output.fa \\\n",
    "   -fastaout $OUTFA\n",
    "mv results.tsv $OUTFA.trims\n",
    "rm trimmed_output.fa\n",
    "\n",
    "# 1. query label\n",
    "# 2 reference label of top hit\n",
    "# 3. %id of semi-global alignment\n",
    "# 4. one-based start coordinate of alignment in query\n",
    "# 5. one-based end   coordinate of alignment in query\n",
    "\n",
    "### POTENTIAL OVEREXTENSIONS\n",
    "# >BAB18266.1 polyprotein [Chiba virus]\n",
    "# >BAC79393.1 polyprotein [Turnip mosaic virus]\n",
    "# >BAA32667.1 polyprotein [Hepatitis C virus (isolate VN004)]\n",
    "# >CAI65399.1 polyprotein [Potato virus Y strain NTN]\n",
    "# >CAI65399.1 polyprotein [Potato virus Y strain NTN]\n",
    "# >CAD24793.1 polyprotein [Peru tomato mosaic virus]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 4. For all New GenBank-Rdrp (trimmed)\n",
    "#    Remove Known False Positive hits (RNA capsid/envelopes)\n",
    "#    Remove RdRp below 160aa long\n",
    "#    Add-back high quality (annotated) partial RdRp\n",
    "mkdir 4_clean_up; cd 4_clean_up\n",
    "\n",
    "ORFFA='gb241.rdrp.fa'\n",
    "ln -s ../3_trim_orf/$ORFFA ./raw.$ORFFA\n",
    "\n",
    "# MANUAL CLEAN-UP\n",
    "\n",
    "\n",
    "#### DECOY SEQUENCES ==========================\n",
    "### HCV Envelope 2 Glycoprotein (113 sequences)\n",
    "grep \"[Hh]epa\" raw.$ORFFA \\\n",
    "  | grep -e \"[Ee]nv\" - \\\n",
    "  | grep -v \"[Rr][Dd][Rr][Pp]\" - \\\n",
    "  | sed 's/>//g' - \\\n",
    "  > DY.hcv_env2.acc\n",
    "  \n",
    "grep \"[Hh]epa\" raw.$ORFFA \\\n",
    "  | grep -e \"E2 protein\" - \\\n",
    "  | grep -v \"[Rr][Dd][Rr][Pp]\" - \\\n",
    "  | sed 's/>//g' - \\\n",
    "  >> DY.hcv_env2.acc  \n",
    "  \n",
    "# i.e. ABG55345.1 envelope glycoprotein, partial [Hepacivirus C]:1-192\n",
    "\n",
    "### Partial Capsid Protein (42 sequences)\n",
    "grep -e \"[Cc]apsid\" raw.$ORFFA \\\n",
    "  | grep -v \"[Rr][Dd][Rr][Pp]\" - \\\n",
    "  | sed 's/>//g' - \\\n",
    "  > DY.capsids.acc\n",
    "  \n",
    "# i.e. ADP88743.1 capsid protein [Rabbit calicivirus]:1-532\n",
    "  \n",
    "### Coat Proteins (70 sequences)\n",
    "grep -e \"[Cc]oat\" raw.$ORFFA \\\n",
    "  | grep -v \"[Rr][Dd][Rr][Pp]\" - \\\n",
    "  | sed 's/>//g' - \\\n",
    "  > DY.coats.acc\n",
    "\n",
    "# i.e. AAK17013.1 coat protein, partial [Papaya ringspot virus]:1-307\n",
    "\n",
    "cat DY* | cut -f1 - > DECOY.acc\n",
    "\n",
    "#### BLACKLISTED SEQUENCES ==========================\n",
    "### Deplete HCV Partial Sequences (1397 sequences)\n",
    "grep -e \"[Hh]epacivirus C\" raw.$ORFFA \\\n",
    "  | grep -e \"[Pp]artial\" - \\\n",
    "  | sed 's/>//g' - \\\n",
    "  > BL.hcv_partial.acc\n",
    "  \n",
    "grep -e \"[Hh]epatitis C\" raw.$ORFFA \\\n",
    "  | grep -e \"[Pp]artial\" - \\\n",
    "  | sed 's/>//g' - \\\n",
    "  >> BL.hcv_partial.acc\n",
    "\n",
    "cat BL.hcv_partial.acc > BLACKLIST.acc\n",
    "\n",
    "\n",
    "#### QC SEQUENCE MATCHES ==========================\n",
    "\n",
    "# Trim output to at least 160 aa post-extension\n",
    "seqkit seq --min-len 160 raw.$ORFFA \\\n",
    "  | seqkit sort -l - > gt160.$ORFFA\n",
    "\n",
    "seqkit seq --max-len 160 raw.$ORFFA \\\n",
    "  | seqkit sort -l - > lt160.$ORFFA\n",
    "\n",
    "# Remove sequences containing 4 or more consecutive \"XXXX\"\n",
    "# (low quality sequences). Quality-control pass\n",
    "seqkit grep -s -v -p \"XXXX\" gt160.$ORFFA \\\n",
    "  | seqkit sort -l - > gt160.QC_PASS.$ORFFA\n",
    "\n",
    "# Removes (50 sequences)\n",
    "seqkit grep -s -p \"XXXX\" gt160.$ORFFA \\\n",
    "  | seqkit sort -l - > gt160.x4.$ORFFA\n",
    "\n",
    "## Recover Partial (non HCV) Sequences\n",
    "# Extract Known partial RdRp (911 sequences)\n",
    "seqkit grep -r -n -p \"RNA-dependent RNA polymerase\" lt160.$ORFFA \\\n",
    "  | seqkit grep -r -n -v -p \"epacivirus C\" - \\\n",
    "  | seqkit grep -r -n -v -p \"epatitis C\" - \\\n",
    "  > partial.$ORFFA\n",
    "\n",
    "# Whitelist PREDICT accessions from trimmed records (20 sequences)\n",
    "seqkit grep -r -n -p \"PREDICT\" lt160.$ORFFA >> partial.$ORFFA\n",
    "seqkit rmdup partial.$ORFFA > tmp.partial.$ORFFA\n",
    "mv tmp.partial.$ORFFA partial.QC_PASS.$ORFFA\n",
    "\n",
    "# DEPLETE Decoy and Blacklisted sequences \n",
    "# Add Long + Partial RdRp, remove false positives\n",
    "cut -f1 -d' ' DECOY.acc      > deplete.tmp\n",
    "cut -f1 -d' ' BLACKLIST.acc >> deplete.tmp\n",
    "\n",
    "cat gt160.QC_PASS.$ORFFA partial.QC_PASS.$ORFFA \\\n",
    "  | seqkit grep -r -v -n -f deplete.tmp \\\n",
    "  > gb241.rdrp.fa\n",
    "  \n",
    "rm deplete.tmp\n",
    "\n",
    "# f799cc6d3525340fecea6edd15d3adea  gb241.rdrp.fa\n",
    "cp gb241.rdrp.fa ../gb241.rdrp.fa\n",
    "\n",
    "cd .."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 5. Create rdrp1\n",
    "#    Combine independent sets of rdrp into one master set\n",
    "#    Bootstrap the annotation from rdrp0 to assign to new sets\n",
    "mkdir 5_make_rdrpN; cd 5_make_rdrpN\n",
    "\n",
    "RDRP='rdrp0.fa'\n",
    "VGB='gb241.rdrp.fa'\n",
    "\n",
    "usearch -sortbylength ../$RDRP \\\n",
    "   -fastaout $RDRP\n",
    "\n",
    "# Seperate out sorted wolf18 and wolf20 sets\n",
    "seqkit grep -n -r -v -p \"yaOV\" $RDRP > wolf18.fa\n",
    "seqkit grep -n -r -p \"yaOV\" $RDRP > wolf20.fa\n",
    "   \n",
    "usearch -sortbylength ../$VGB \\\n",
    "   -fastaout $VGB\n",
    "\n",
    "# Order of priority for taxonomic classification\n",
    "cat wolf18.fa $VGB wolf20.fa > tmpcat.fa\n",
    "\n",
    "# Create initial 45% cluster set\n",
    "function uclust () {\n",
    "\n",
    "  INPUT=$1\n",
    "  OUTNAME=$2\n",
    "  ID=$3\n",
    "  \n",
    "  # uclust\n",
    "  OUTPUT=\"$OUTNAME.id$ID\"\n",
    "  usearch -cluster_smallmem $INPUT \\\n",
    "     -sortedby other \\\n",
    "     -id 0.$ID \\\n",
    "     -maxaccepts 4 \\\n",
    "     -maxrejects 64 \\\n",
    "     -maxhits 1 \\\n",
    "     -uc $OUTPUT.uc \\\n",
    "     -centroids $OUTPUT.fa\n",
    "\n",
    "  mkdir id$ID; mv *.id$ID.* id$ID/\n",
    "\n",
    "}\n",
    "\n",
    "# Cluster rdrp0 and genbank_rdrp0_matches at 45%\n",
    "uclust tmpcat.fa initial.cluster 45\n",
    "\n",
    "#      Seqs  14613 (14.6k)\n",
    "#  Clusters  4723\n",
    "#  Max size  315\n",
    "#  Avg size  3.1\n",
    "#  Min size  1\n",
    "#Singletons  2982, 20.4% of seqs, 63.1% of clusters\n",
    "\n",
    "# For every Viral GenBank RdRp; which cluster does it belong?\n",
    "grep \"^[HC]\" id45/initial.cluster.id45.uc \\\n",
    "  | cut -f 1,9,10 - \\\n",
    "  | sort -k1 - > uctable45.tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Manual Additions\n",
    "\n",
    "# 1. Regex-fu to import UC file (uctabl45.tsv) into excel\n",
    "# 2. For accessions without a lift-over annotation, retrieve full taxonomy-line\n",
    "#    use: https://www.ncbi.nlm.nih.gov/sites/batchentrez\n",
    "# 3. For GenBank Records of unknown \"Branch\" use local alignment data from rdrp0\n",
    "#    paste accession-list into \"assign.branch\"\n",
    "\n",
    "while read -r line; do\n",
    "  grep \"$line\" gb241_r0.local.pro >> tmp.pro\n",
    "done < assign.branch\n",
    "\n",
    "# Clean-up\n",
    "seqkit rmdup rdrp1_ALPHA_r2.fa \\\n",
    "  | seqkit seq --min-len 35 - \\\n",
    "  | seqkit sort -r -l - \\\n",
    "  | seqkit rmdup -n - \\\n",
    "  > rdrp1_ALPHA_r2.sort.fa\n",
    "seqkit faidx -f rdrp1_ALPHA_r2.sort.fa\n",
    "\n",
    "# Re-import header swap\n",
    "aws s3 cp s3://serratus-public/tmp/header_update.tsv ./\n",
    "# AAC25017.1 replicase protein, partial [Wild cucumber mosaic virus]:1-214 -->\n",
    "# kiti.Tymoviridae.wild_cucumber_mosaic_virus:AAC25017\n",
    "# AAF00520.1 polyprotein, partial [Pleione virus Y]:1-240 -->\n",
    "# pisu.Potyviridae.pleione_virus_y:AAF00520\n",
    "# ...\n",
    "\n",
    "\n",
    "function reheader {\n",
    "  INFA=$1  # Input Fasta File\n",
    "  TSV=$2   # 2-field TSV == 1:Old_Fasta_header 2:New_Fasta_Header\n",
    "  OUTFA=$3 # Output Fasta File\n",
    "  \n",
    "  cp $INFA tmp.fa\n",
    "  cut -f1 $TSV > header.old.tmp\n",
    "\n",
    "  \n",
    "  # Find exact string match of Old_Header INFA\n",
    "  # return line number of header in file\n",
    "  grep -nF -f header.old tmp.fa \\\n",
    "    | cut -f1 -d\":\" - > header.old.N.tmp\n",
    "\n",
    "  cut -f2 $TSV \\\n",
    "    | paste header.old.N.tmp - \\\n",
    "    > header.N.new.tmp\n",
    "    \n",
    "  # Replace old-header inline with new header\n",
    "  while read -r line; do\n",
    "    # add \"s\" to N for sed\n",
    "    N=$(echo $line | sed 's/ /s /g' - | cut -f1 -d' ' -)\n",
    "    NEW=$(echo $line | cut -f2 -d' ' -)\n",
    "    echo $N $NEW\n",
    "    sed -i \"$N/.*/>$NEW/\" tmp.fa\n",
    "  done < header.N.new.tmp\n",
    "  \n",
    "  mv tmp.fa $OUTFA\n",
    "    \n",
    "}\n",
    "\n",
    "reheader rdrp1_ALPHA_r2.sort.fa header_update.tsv rdrp1_BETA.fa\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Revision 3C - Re-search GenBank with rdrp1_BETA\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### GLOBAL REFINEMENT SCRIPT\n",
    "\n",
    "### PARAMETERS ===================================\n",
    "# RDRP Database\n",
    "GENOME='rdrp1_BETA'\n",
    "GPATH='../rev3B'\n",
    "RDRP=\"$GENOME.fa\"\n",
    "\n",
    "# Query Records (GenBank)\n",
    "QUERY='vgb241'\n",
    "QPATH='../vgb241'\n",
    "\n",
    "# Output Suffix\n",
    "OUTNAME='gb241_r1B'\n",
    "\n",
    "# Diamond QC-Filtering (Initial)\n",
    "EVALUE=6  # -log(E-value) > $EVAL cut-off\n",
    "RCOV=50 # rdrp-coverage > $RCOV cut-off\n",
    "\n",
    "### INITIALIZE ===================================\n",
    "# Revision 3C folder\n",
    "mkdir -p rev3C; cd rev3C\n",
    "\n",
    "# Use rdrp0_revision1 as start-point\n",
    "ln -s $GPATH/$GENOME.fa ./\n",
    "diamond makedb --in $GENOME.fa -d $GENOME\n",
    "\n",
    "# Query\n",
    "ln -s $QPATH/$QUERY.fa ./\n",
    "ln -s $QPATH/$QUERY.fa.fai ./\n",
    "\n",
    "### 0 DIAMOND SEARCH ==============================\n",
    "mkdir 0_rdrp_search; cd 0_rdrp_search\n",
    "\n",
    "# Diamond blastp alignment\n",
    "time cat ../$QUERY.fa |\\\n",
    "diamond blastp \\\n",
    "  -d ../$GENOME.dmnd \\\n",
    "  --unal 0 \\\n",
    "  --masking 0 \\\n",
    "  --ultra-sensitive \\\n",
    "  -k 1 \\\n",
    "  -p 30 \\\n",
    "  -b 2 \\\n",
    "  -f 6 qseqid qstart qend qlen qstrand \\\n",
    "       sseqid sstart send slen \\\n",
    "       pident evalue bitscore cigar \\\n",
    "       qseq \\\n",
    "  > $OUTNAME.raw.pro\n",
    "\n",
    "\n",
    "# real    39m30.035s\n",
    "# user    1120m51.264s\n",
    "# sys     1m46.920s\n",
    "\n",
    "# QC Filter diamond output based on\n",
    "# E-value threshold and and rdrp-Coverage\n",
    "\n",
    "while read -r line; do\n",
    "  # Return -log(e-value); zero == 999\n",
    "  e_value=$(echo $line | cut -f 11 -d' ' - \\\n",
    "    | sed 's/[0-9\\.]*e.//g' - \\\n",
    "    | sed 's/^0//g' -)\n",
    "  if (( $e_value == 0 )); then\n",
    "    e_value='999'\n",
    "  fi\n",
    "  \n",
    "  # rdrp-match coordinates\n",
    "  r_start=$(echo $line | cut -f 7 -d' ' - )\n",
    "  r_end=$(echo $line | cut -f 8 -d' ' - )\n",
    "  r_len=$(echo $line | cut -f 9 -d' ' - )\n",
    "  \n",
    "  # convert to percent\n",
    "  r_cov=$( echo \"$r_end - $r_start + 1\" | bc )\n",
    "  r_covpct=$( echo \"scale=0; 100 * $r_cov / $r_len\" | bc )\n",
    "  \n",
    "  #echo $r_start $r_end $r_len $r_covpct $e_value\n",
    " \n",
    "  if (( $r_len > $RCOV )) && (( $e_value > $EVALUE )); then\n",
    "    #echo HIT\n",
    "    echo $line | sed 's/ /\\t/g' >> \"$OUTNAME\".local.pro\n",
    "  else\n",
    "    #echo MISS\n",
    "    echo $line | sed 's/ /\\t/g' >> \"$OUTNAME\".REJECT.pro\n",
    "  fi\n",
    "  # I guess they never miss huh\n",
    "done < \"$OUTNAME\".raw.pro\n",
    "\n",
    "mkdir raw;\n",
    "mv \"$OUTNAME\".raw.pro raw/\n",
    "mv\"$OUTNAME\".REJECT.pro raw/\n",
    "\n",
    "# Make fasta of output\n",
    "cut -f1,14 $OUTNAME.local.pro \\\n",
    "  |  sed 's/^/>/g' \\\n",
    "  | sed 's/\\t/\\n/g' \\\n",
    "  > $OUTNAME.local.fa\n",
    "\n",
    "usearch -sortbylength $OUTNAME.local.fa \\\n",
    "   -fastaout tmp.sort.fa\n",
    "   \n",
    "mv tmp.sort.fa $OUTNAME.local.fa\n",
    "\n",
    "cd ..\n",
    "\n",
    "### 1 CLUSTER RESULTS ============================\n",
    "# 1. Cluster all putative-RdRp in QUERY into rdrp at 90%\n",
    "#    Isolate centroids that are from QUERY as putative novel-RdRp\n",
    "#    Note: these are \"local\" matches, and require extension\n",
    "\n",
    "mkdir 1_merge_rdrp; cd 1_merge_rdrp\n",
    "GBLOCAL=\"$OUTNAME.local.fa\"\n",
    "\n",
    "# Reference set of rdrp\n",
    "usearch -sortbylength ../$RDRP \\\n",
    "   -fastaout ./$RDRP\n",
    "   \n",
    "# Local query matches to rdrp\n",
    "usearch -sortbylength ../0_rdrp_search/$GBLOCAL \\\n",
    "   -fastaout ./$GBLOCAL\n",
    "   \n",
    "# Order the files such that rdrp0 takes priority over genbank\n",
    "cat $RDRP $GBLOCAL > cat_rdrp.fa\n",
    "\n",
    "function uclust () {\n",
    "\n",
    "  INPUT=$1\n",
    "  OUTNAME=$2\n",
    "  ID=$3\n",
    "  \n",
    "  # uclust\n",
    "  OUTPUT=\"$OUTNAME.id$ID\"\n",
    "  usearch -cluster_smallmem $INPUT \\\n",
    "     -sortedby other \\\n",
    "     -id 0.$ID \\\n",
    "     -maxaccepts 4 \\\n",
    "     -maxrejects 64 \\\n",
    "     -maxhits 1 \\\n",
    "     -uc $OUTPUT.uc \\\n",
    "     -centroids $OUTPUT.fa\n",
    "\n",
    "  mkdir id$ID; mv *.id$ID.* id$ID/\n",
    "\n",
    "}\n",
    "\n",
    "# Cluster rdrp0 and genbank_rdrp0_matches at 90%\n",
    "uclust cat_rdrp.fa $OUTNAME 90\n",
    "\n",
    "# Retrieve non-rdrp0 sequences (GenBank-derived RdRp)\n",
    "grep \">\" id90/$OUTNAME.id90.fa \\\n",
    "  | grep -v -e \">[lkpdnr]\" \\\n",
    "  | sed 's/>//g' - \\\n",
    "  > new.rdrp.acc\n",
    "\n",
    "cd ..\n",
    "\n",
    "### 2 RETREIVE ORF ============================\n",
    "# 2. For all New GenBank-derived RdRpR\n",
    "#    retrieve the full-length sequence/ORF from GenBank\n",
    "mkdir 2_retrieve_orf; cd 2_retrieve_orf\n",
    "\n",
    "cp ../1_merge_rdrp/new.rdrp.acc ./new.rdrp.acc\n",
    "\n",
    "# Full Length ORF Fasta file\n",
    "ORFFA=\"$OUTNAME.orf.fa\"\n",
    "seqkit grep -f new.rdrp.acc ../$QUERY.fa > $ORFFA\n",
    "\n",
    "cd ..\n",
    "\n",
    "### 3 NW Extension + Trim =====================\n",
    "# 3. For all  ORF (containing rdrp)\n",
    "#    Semi-global align the ORF to the original rdrp\n",
    "#    to \"trim\" the sequences to the boundary\n",
    "mkdir 3_trim_orf; cd 3_trim_orf\n",
    "\n",
    "# Perform semi-global extension of potential hits\n",
    "aws s3 cp s3://serratus-public/bin/usearch12_trim ./\n",
    "chmod 755 usearch12_trim\n",
    "\n",
    "ln -s ../$RDRP ./\n",
    "ln -s ../2_retrieve_orf/$ORFFA ./\n",
    "\n",
    "RDRPFA=\"raw.$OUTNAME.rdrp.fa\"\n",
    "\n",
    "# NW Semi-Global Extension\n",
    "./usearch12_trim -usearch_global $ORFFA \\\n",
    "    -id 0.01 \\\n",
    "    -fulldp \\\n",
    "    -maxaccepts 8 \\\n",
    "    -maxrejects 32 \\\n",
    "    -top_hit_only \\\n",
    "    -db $RDRP \\\n",
    "    -userfields query+target+id+qtrimlo+qtrimhi \\\n",
    "    -userout results.tsv \\\n",
    "    -trimout trimmed_output.fa\n",
    "    \n",
    "usearch -sortbylength trimmed_output.fa \\\n",
    "   -fastaout $RDRPFA\n",
    "\n",
    "mv results.tsv $RDRPFA.trims\n",
    "rm trimmed_output.fa\n",
    "\n",
    "### 4 Semi-Automated Clean-up =====================\n",
    "# 4. For all New Query-Rdrp (trimmed)\n",
    "#    Remove Known False Positive hits (RNA capsid/envelopes)\n",
    "#    Remove RdRp below 160aa long\n",
    "#    Add-back high quality (annotated) partial RdRp\n",
    "mkdir 4_clean_up; cd 4_clean_up\n",
    "\n",
    "ln -s ../3_trim_orf/$RDRPFA ./$RDRPFA\n",
    "\n",
    "# Output removes raw- prefix\n",
    "RDRPFA=\"$OUTNAME.rdrp.fa\"\n",
    "\n",
    "# MANUAL CLEAN-UP\n",
    "\n",
    "\n",
    "#### DECOY SEQUENCES\n",
    "### HCV Envelope 2 Glycoprotein (113 sequences)\n",
    "grep \"[Hh]epa\" raw.$RDRPFA \\\n",
    "  | grep -e \"[Ee]nv\" - \\\n",
    "  | grep -v \"[Rr][Dd][Rr][Pp]\" - \\\n",
    "  | sed 's/>//g' - \\\n",
    "  > DY.hcv_env2.acc\n",
    "  \n",
    "grep \"[Hh]epa\" raw.$RDRPFA \\\n",
    "  | grep -e \"E2 protein\" - \\\n",
    "  | grep -v \"[Rr][Dd][Rr][Pp]\" - \\\n",
    "  | sed 's/>//g' - \\\n",
    "  >> DY.hcv_env2.acc  \n",
    "  \n",
    "# i.e. ABG55345.1 envelope glycoprotein, partial [Hepacivirus C]:1-192\n",
    "\n",
    "### Partial Capsid Protein (42 sequences)\n",
    "grep -e \"[Cc]apsid\" raw.$RDRPFA \\\n",
    "  | grep -v \"[Rr][Dd][Rr][Pp]\" - \\\n",
    "  | sed 's/>//g' - \\\n",
    "  > DY.capsids.acc\n",
    "  \n",
    "# i.e. ADP88743.1 capsid protein [Rabbit calicivirus]:1-532\n",
    "  \n",
    "### Coat Proteins (70 sequences)\n",
    "grep -e \"[Cc]oat\" raw.$RDRPFA \\\n",
    "  | grep -v \"[Rr][Dd][Rr][Pp]\" - \\\n",
    "  | sed 's/>//g' - \\\n",
    "  > DY.coats.acc\n",
    "\n",
    "# i.e. AAK17013.1 coat protein, partial [Papaya ringspot virus]:1-307\n",
    "\n",
    "cat DY* | cut -f1 - > DECOY.acc\n",
    "\n",
    "#### BLACKLISTED SEQUENCES ==========================\n",
    "### Deplete HCV Partial Sequences (1397 sequences)\n",
    "grep -e \"[Hh]epacivirus C\" raw.$RDRPFA \\\n",
    "  | grep -e \"[Pp]artial\" - \\\n",
    "  | sed 's/>//g' - \\\n",
    "  > BL.hcv_partial.acc\n",
    "  \n",
    "grep -e \"[Hh]epatitis C\" raw.$RDRPFA \\\n",
    "  | grep -e \"[Pp]artial\" - \\\n",
    "  | sed 's/>//g' - \\\n",
    "  >> BL.hcv_partial.acc\n",
    "\n",
    "cat BL.hcv_partial.acc > BLACKLIST.acc\n",
    "\n",
    "\n",
    "#### QC SEQUENCE MATCHES \n",
    "\n",
    "# Trim output to at least 160 aa post-extension\n",
    "seqkit seq --min-len 160 raw.$RDRPFA \\\n",
    "  | seqkit sort -l - > gt160.$RDRPFA\n",
    "\n",
    "seqkit seq --max-len 160 raw.$RDRPFA \\\n",
    "  | seqkit sort -l - > lt160.$RDRPFA\n",
    "\n",
    "# Remove sequences containing 4 or more consecutive \"XXXX\"\n",
    "# (low quality sequences). Quality-control pass\n",
    "seqkit grep -s -v -p \"XXXX\" gt160.$RDRPFA \\\n",
    "  | seqkit sort -l - > gt160.QC_PASS.$RDRPFA\n",
    "\n",
    "# Removes (50 sequences)\n",
    "seqkit grep -s -p \"XXXX\" gt160.$RDRPFA \\\n",
    "  | seqkit sort -l - > gt160.x4.$RDRPFA\n",
    "\n",
    "## Recover Partial (non HCV) Sequences\n",
    "# Extract Known partial RdRp (911 sequences)\n",
    "seqkit grep -r -n -p \"RNA-dependent RNA polymerase\" lt160.$RDRPFA \\\n",
    "  | seqkit grep -r -n -v -p \"epacivirus C\" - \\\n",
    "  | seqkit grep -r -n -v -p \"epatitis C\" - \\\n",
    "  > partial.$RDRPFA\n",
    "\n",
    "# Whitelist PREDICT accessions from trimmed records (20 sequences)\n",
    "seqkit grep -r -n -p \"PREDICT\" lt160.$RDRPFA >> partial.$RDRPFA\n",
    "seqkit rmdup partial.$RDRPFA > tmp.partial.$RDRPFA\n",
    "mv tmp.partial.$RDRPFA partial.QC_PASS.$RDRPFA\n",
    "\n",
    "# DEPLETE Decoy and Blacklisted sequences \n",
    "# Add Long + Partial RdRp, remove false positives\n",
    "cut -f1 -d' ' DECOY.acc      > deplete.tmp\n",
    "cut -f1 -d' ' BLACKLIST.acc >> deplete.tmp\n",
    "\n",
    "cat gt160.QC_PASS.$RDRPFA partial.QC_PASS.$RDRPFA \\\n",
    "  | seqkit grep -r -v -n -f deplete.tmp \\\n",
    "  | seqkit rmdup -n \\\n",
    "  > final.$RDRPFA\n",
    "  \n",
    "rm deplete.tmp\n",
    "\n",
    "# f799cc6d3525340fecea6edd15d3adea  gb241.rdrp.fa\n",
    "\n",
    "cd ..\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### 5 Manual Clean-up =====================\n",
    "# 5. Create rdrpN+1\n",
    "#    Combine independent sets of rdrp into one master set\n",
    "#    Bootstrap the annotation from rdrp0 to assign to new sets\n",
    "mkdir 5_make_rdrpN; cd 5_make_rdrpN\n",
    "ln -s ../4_clean_up/final.gb241_r1B.rdrp.fa ./\n",
    "ln -s ../1_merge_rdrp/id90/gb241_r1B.id90.uc ./\n",
    "\n",
    "# Make \"manual.additions\" of accessions to add to rdrp1\n",
    "# see below\n",
    "seqkit grep -f manual.additions final.gb241_r1B.rdrp.fa > manual.additions.fa\n",
    "grep -f manual.additions \n",
    "\n",
    "## BOOKMARK"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Manual inspection of hits\n",
    "\n",
    "- Flanking sequences from imperfect rdrp-extensions result in more capsid/helicase/flanking proteins coming up in this round.\n",
    "- 30 matches annotated as RdRp were additionally retrieved, adding these back to `rdrp1`. These are pretty much all \"out there\" wrt research subjects so we're scrapping the bottom of the barrel for GenBank for now.\n",
    "\n",
    "```\n",
    "ASM94010.1 putative RNA-dependent RNA polymerase, partial [Caledonia beadlet anemone chu-like virus 1]\t75.9\n",
    "ASV45868.1 RNA-dependent RNA polymerase protein [Lobeira virus]\t70.4\n",
    "AXG65493.1 RNA-dependent RNA polymerase [Kundal virus]\t69.1\n",
    "QED22729.1 RNA-dependent RNA polymerase, partial [Botryosphaeria dothidea narnavirus 1]\t48.6\n",
    "QIR30265.1 RNA-dependent RNA polymerase [Plasmopara viticola lesion associated mitovirus 42]\t42.9\n",
    "QOW97233.1 RNA-dependent RNA polymerase [Partiti-like adriusvirus]\t18.4\n",
    "ABA46925.1 putative RNA dependent RNA polymerase, partial [Cassava frogskin virus]\t18.2\n",
    "AHB12495.1 RNA-dependent RNA polymerase, partial [Cassava frogskin associated virus]\t17.9\n",
    "AVM87434.1 RNA-dependent RNA polymerase [Wenling jack mackerels birnavirus]\t17.7\n",
    "ASQ44445.1 RNA-dependent RNA polymerase, partial [Human rotavirus A]\t17.5\n",
    "BBA12670.1 RNA-dependent RNA polymerase, partial [Sclerotium Rolfsii dsRNA virus]\t17.4\n",
    "QBZ92966.1 RNA-directed RNA polymerase, partial [Potato virus S]\t17.2\n",
    "AWY10996.1 replicase-associated protein, partial [Sclerotinia sclerotiorum tymo-like RNA virus 5]\t17\n",
    "QCQ84350.1 RNA-dependent RNA polymerase [Lates calcarifer birnavirus]\t17\n",
    "AOX47590.1 RNA-dependent RNA polymerase, partial [Ceratobasidium mycovirus-like]\t16.6\n",
    "CAD30691.1 RNA-dependent RNA polymerase [Blotched snakehead virus]\t16.3\n",
    "BAP99821.1 predicted structural protein [Chaetoceros tenuissimus RNA virus type-II]\t16\n",
    "BCH36656.1 RNA-dependent RNA polymerase [Magnaporthe oryzae narnavirus 1]\t14.2\n",
    "QAY29233.1 putative RNA-dependent RNA-polymerase, partial [Agassiz Rock virus]\t33.6\n",
    "ASU43981.1 RdRp [Australian Anopheles totivirus]\t71.1\n",
    "QED43009.1 RdRp, partial [Phakopsora virgavirus A]\t59\n",
    "ANC52159.1 RdRp [Sclerotinia sclerotiorum mycoreovirus 4]\t48.5\n",
    "QNQ74064.1 RdRp [Plasmopara viticola lesion associated orfanplasmovirus 2]\t27.6\n",
    "QNQ74065.1 RdRp [Plasmopara viticola lesion associated orfanplasmovirus 3]\t26.5\n",
    "QNQ74063.1 RdRp [Plasmopara viticola lesion associated orfanplasmovirus 1]\t26.1\n",
    "QNQ74067.1 RdRp [Plasmopara viticola lesion associated orfanplasmovirus 5]\t26.1\n",
    "QNQ74066.1 RdRp [Plasmopara viticola lesion associated orfanplasmovirus 4]\t24.7\n",
    "QED42926.1 putative RdRp [Leucocoprinus gammaflexivirus B]\t22.4\n",
    "QBP78746.1 RdRp, partial [Botrytis virus F]\t20.2\n",
    "QHD64760.1 RdRp, partial [Plasmopara viticola lesion associated picorna-like 4]\t16.7\n",
    "QAY29251.1 putative RNA-dependent RNA-polymerase [Dumyat virus]\t55.4\n",
    "```\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "usearch -sortbylength ../$RDRP \\\n",
    "   -fastaout $RDRP\n",
    "   \n",
    "usearch -sortbylength manual.additions.fa \\\n",
    "   -fastaout manual.additions.sort.fa\n",
    "\n",
    "# Order of priority for taxonomic classification\n",
    "cat $RDRP manual.additions.sort.fa > tmpcat.fa\n",
    "\n",
    "# Create initial 45% cluster set\n",
    "function uclust () {\n",
    "\n",
    "  INPUT=$1\n",
    "  OUTNAME=$2\n",
    "  ID=$3\n",
    "  \n",
    "  # uclust\n",
    "  OUTPUT=\"$OUTNAME.id$ID\"\n",
    "  usearch -cluster_smallmem $INPUT \\\n",
    "     -sortedby other \\\n",
    "     -id 0.$ID \\\n",
    "     -maxaccepts 4 \\\n",
    "     -maxrejects 64 \\\n",
    "     -maxhits 1 \\\n",
    "     -uc $OUTPUT.uc \\\n",
    "     -centroids $OUTPUT.fa\n",
    "\n",
    "  mkdir id$ID; mv *.id$ID.* id$ID/\n",
    "\n",
    "}\n",
    "\n",
    "# Cluster rdrp0 and genbank_rdrp0_matches at 45%\n",
    "uclust tmpcat.fa initial.cluster 45\n",
    "\n",
    "#      Seqs  14613 (14.6k)\n",
    "#  Clusters  4723\n",
    "#  Max size  315\n",
    "#  Avg size  3.1\n",
    "#  Min size  1\n",
    "#Singletons  2982, 20.4% of seqs, 63.1% of clusters\n",
    "\n",
    "# For every Viral GenBank RdRp; which cluster does it belong?\n",
    "grep \"^[HC]\" id45/initial.cluster.id45.uc \\\n",
    "  | cut -f 1,9,10 - \\\n",
    "  | sort -k1 - > uctable45.tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Manually adjust headers based on UC assignment/monkey work\n",
    "#  in excel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Re-import header swap\n",
    "# vim header_update.tsv\n",
    "\n",
    "function reheader {\n",
    "  INFA=$1  # Input Fasta File\n",
    "  TSV=$2   # 2-field TSV == 1:Old_Fasta_header 2:New_Fasta_Header\n",
    "  OUTFA=$3 # Output Fasta File\n",
    "  \n",
    "  cp $INFA tmp.fa\n",
    "  cut -f1 -d' ' $TSV > header.old.tmp\n",
    "\n",
    "  # Find exact string match of Old_Header INFA\n",
    "  # return line number of header in file\n",
    "  grep -nF -f header.old.tmp tmp.fa \\\n",
    "    | cut -f1 -d\":\" - > header.old.N.tmp\n",
    "\n",
    "  cut -f2 $TSV \\\n",
    "    | paste header.old.N.tmp - \\\n",
    "    > header.N.new.tmp\n",
    "    \n",
    "  # Replace old-header inline with new header\n",
    "  while read -r line; do\n",
    "    # add \"s\" to N for sed\n",
    "    N=$(echo $line | sed 's/ /s /g' - | cut -f1 -d' ' -)\n",
    "    NEW=$(echo $line | cut -f2 -d' ' -)\n",
    "    echo $N $NEW\n",
    "    sed -i \"$N/.*/>$NEW/\" tmp.fa\n",
    "  done < header.N.new.tmp\n",
    "  \n",
    "  mv tmp.fa $OUTFA\n",
    "    \n",
    "}\n",
    "\n",
    "reheader tmpcat.fa header_update.tsv rdrp1_GAMMA.fa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Manual Addition of sequences\n",
    "cat rdrp1_GAMMA.fa manual.seq.additions.fa \\\n",
    "  | seqkit sort -n - \\\n",
    "  | seqkit rmdup - \\\n",
    "  > rdrp1_DELTA.fa \n",
    "  \n",
    "cp rdrp1_DELTA.fa ../../\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "File: `manual.seq.additions.fa`\n",
    "\n",
    "```\n",
    ">rdrp.Serrataviridae.d_ana_virus:SS0000001\n",
    "MTFEYRSATPTKGTFAGLSSEVIWTKKGGWRTLLTERPKKFRLENLKYVGGLPYYVRQKEQTRRVNEVVHELIRKHFPDFDANAYGKSRPTFEKIENEAKRINPAPTWELDEEARAFADYTLYNYILPLGPFPLATDDEIINELLPKSPGLWYEDLHKNATKRTTCGVAVKQARGLLDGTLKDFEDSDPLFKFVGKTEILSAVKIQEKRNRNYAVPPLYLYILEIIFFHHLSNAYKEHDKGYTLTLQHGGLYELFAQMDKYGTVISDDKTGFDLRQQWIVQKSVARICAYTIDMSDKHHMMYMHLAHQFCAKNILMPDGSVFYTPYHHASGRYVTTMKGSLFHRWEQAYVFYVVMQQQTPDRATEGLGREALRLLFERMNTRIASDDVLQGVPNDPIYEPWLDQEKRTSVWETLVPIKPGSALQTHSSTGHFYLGWTVENGKVKHNSRTKLLAKLLYSSIKDKQGVITGLVHTSPYDTPMCDFLRELANKWGVEFCGQRIAQLLWRRRLDTAPFLPSPPTPEGVIMMHPVDNVHKIRNSVTMARKKQVKQAAKRKEERKIERAVERRQAVHPLRHLMTPKLHGNTYKYLETLVDPYNTPAGASTPDKVVRKSARYKSYVKTTMTTGTGGFGFCLFNPYSAAWGDQFAVATSNGTYAGTLTDPNSATAGVNNYKSTAPLNWADASANGYSCRVVGAGIRILNNTALLDMGGSVTGIREPGNQYLTGYTFADILNYNQCHVIRPEAGKWIHVEWAPTGNIGQETGDQFEFDDENPASAASMIGGQIAIVATSAGPAANAQSYDVEAVVIYEMLGEKLPLQDVHVDPLGQAACLEVIAELDGSASDKWITQVGKVAKRTGKALREISAAAMPAALMLGLA\n",
    ">var.Deltavirus.mmondv:SS0000002\n",
    "MENPKQGNSKGREETLRQWVKGRERKEELEEELRRLNKKIKKLEERNPWLGNILGMVRKRGGEEQSSPSKRPRGESMEVDGGSSGGPQGPRFSEEEKRDHRRRKALENKKKQLEGHGKKLPQEEEEELRRLTGPDEARERRLGDHGPGDVNVGGGGPRGAPGGGFVSNLQGVPESPYSRRGEGLDIRGKQEFP\n",
    ">var.Deltavirus.ovirdV:SS0000003\n",
    "MDTPGNKANQRGREDVLRDWVEGRRRKDELEKELQRLSRKLKRLEEKNPWLGNVLGMIRKGGGGEGASPAKRPRSDPMDVDPGPSGSQQRPRFTDQERRDHRRRKALENKKKQLSGGGKDLSPEEEEELRRLTGPDEERERRVAGPPVGDVNPFGGPPRGAPGGGFVPNLQGVPESPFARKGDGLDTRGDQ\n",
    ">var.Deltavirus.pmacdv:SS0000004\n",
    "METPGKKKTPKPRQEILEEWADMNRKKRELEKELQRTLRKKKKLEEENPWLGNVLGIVRQRAGGTDAPQAKKRRLGEEMEVDGGPGPSSAPRAPFTKKEREDHRRRCALENKKKQLEEQGKKMSEEEEAERRRLAEEDERRKKRAEGGGDGDVNPPEGTPRGAFGGGFVPGLQGVPESPFSRRGDGLSLRGEGEYP\n",
    ">var.Deltavirus.tgutdV:SS0000005\n",
    "MDATPKKKGPKSREEVLEEWCDLRKRKRELEKELQRVNKKKKKLEDENGWLGNVLGMVRKGGGGDSSSPRKRKRGEGGDMDVDPPGDGGSSRPGPYSDEERAEHRRRCMLENKKKQLEAQGKQLSGEEEAERRRLASRDEDRKRRRQEAEAGGVNPSAGQPRGAPGGGFVPDSQGVPESPFTRRGDGLSTRGEGMFP\n",
    ">var.Deltavirus.bgladv:SS0000006\n",
    "MAALTAPKKKTGVDRFATLKRWTDTSDKIKELEEALRKERDRKRKLEDKHPWLGNVKGIIKEPGAKKPESGKASERMDTEGSEEPSRKRKRKEEYSEEQRKDYREMKSLLNKAAQLQKAGKKLSKEETTKLCGLSGKTGEGIPVELLTALTYDGADVNPLQIIGSARAGENPHDLNLDGVPEGSRTIPPTLFPRGPR\n",
    ">var.Deltavirus.ichidv:SS0000007\n",
    "MESSFAPAAETTAGKSSKMERGEVLNKWVLLREDIKDLEAKLAKARGRKRKLERENPWLGNILGIIKERPPGDGEVAPKKRKEKEASEVQERPTKRKRSKKDLYSEEERERYKRLKQLLNKAKQLGSKGRKLTPEEQGELMGLAAATGEQLPEDVVVGLSGEVNPLLVSGAPAAGINPFGLNLTGVPVVEGDPMQLDIRGDRST\n",
    ">var.Deltavirus.aansdv:SS0000008\n",
    "MENKEESQKRKRGREETLQRWVDDRRRKKELEEELEKIRKRIKDRERKNPWLGNLKGILKQETGGGQEKKKETTEPMDTGESPKKKKKRDYEGLSFTPEEKKQHKRKCDLENKKKQLNSKGKQLSAQEEDELRKLQKLDEERLLKKRAEREQRSGGVNLFGSASPSTSGGGNAPTTQGLRLPWQK\n",
    ">var.Deltavirus.wldv:SS0000009\n",
    "MSEASTQRSVILVEWVSCMEEIRRATRTLEKNKRKMKKLEESNPWLGNVKGLIKKEEGEVRPKTAEAEPTRKRKRGEEEMTTDPFSDRAAKRSFTPEEKQAYQKMSQELNRLLQRAKSGKQNSLEETERVVREAAVLKRYLTNDQRELLGTAAEVEGTQLTAWQMEATRNIL\n",
    ">var.Deltavirus.hdv8:AJ584849\n",
    "MSQSDAKRERRGGREDVLSKWVEARKDLEDLEKRIRKTRRNIKRLEDENPWLGNILGIIRKGKAGEGAPPAKRARTDQMEVDSGPGKKSRKGGFTDEERRDHRRRKALENKKKQLSSGGKKLSEEEEGELRRLIVEDEKRERRAAGPQDGGVNPPGGSPRGAPGGGFVPRMLGVPESPFTRHGDGLDIRGDQQFP\n",
    ">var.Deltavirus.hdv6:AJ584847\n",
    "MGPAEQKRKRGGREEILEKWVELRKNREDLERKLRKTQKGLKKLEDDNPWLGNIIGIIRKGKDGEGAPPTKKARTDQMEVDSGPRGKPHKSGFTDEERRDHRRRKALENKKKQLAAGGKNLSREEEEELGRLTGEDEQRKRRTAGPRVGDVNPPGGDPRGAPGGGFVPTMLHVPESPFSRTGEGLDVRGTQQFP\n",
    ">var.Deltavirus.hdv5:AJ584848\n",
    "MSQSESKKARRGGREEILEKWVQARKDSEDLEKELRKTKRTIKKLEDENPWLGNILGIIRKGKDGEGAPPAKRARTDQMEVDSGPRKRTRAGDFTDQERKDHRRRKALENKKNQLSAGGKSLSREEEEELRKLTEEDERRERRVAGPRVGDVNPPEGPPRGAPGGGFVPQMLDVPESPFSRTGSGLDVRGNQQFP\n",
    ">var.Deltavirus.hdv2:X60193\n",
    "MSQSETRRGRRGTREETLEKWITARKKAEELEKDLRKTRKTIKKLEEENPWLGNIVGIIRKGKDGEGAPPAKRPRTDQMEVDSGPGKRPHKSGFTDKEREDHRRRKALENKKKQLSAGGKILSKEEEEELRRLTDEDEERKRRVAGPRVGDVNPSRGGPRGAPGGGFVPQMAGVPESPFSRTGEGLDIRGTQGFP\n",
    ">var.Deltavirus.hdv4:AF018077\n",
    "MSQPDSRRPRRGREEQLGKWIDARRRKEELERDLRKVNKTIKRLEEDNPWLGNVRGIIRKDKDGEGAPPAKRARTDQMEVDSGPRKRKHPGGFTEQERRDHRRRKALENKKKQLSSGGKNLSREEEEELRRLTEEDERRERRVAGPRVGDVNPLDGGPRGAPGGGFVPSMHDIPESPFTRRGDGLDVRGAQEFP\n",
    ">var.Deltavirus.hdv1:M21012\n",
    "MSRSESRKNRGGREEILEQWVAGRKKLEELERDLRKTKKKLKKIEDENPWLGNIKGILGKKDKDGEGAPPAKRARTDQMEVDSGPRKRPLRGGFTDKERQDHRRRKALENKKKQLSAGGKNLSKEEEEELRRLTEEDERRERRVAGPPVGGVNPLEGGSRGAPGGGFVPNLQGVPESPFSRTGEGLDIRGNQGFP\n",
    ">var.Deltavirus.hdv1:NC_001653\n",
    "MSRPEGRKNRGGREEVLEQWVSGRKKLEELERDLRKVKKKIKKLEDEHPWLGNIKGILGKKDKDGEGAPPAKRARTDQMEVDSGPRKRPSRGGFTDKERQDHRRRKALENKRKQLSAGGKNLSKEEEEELRRLTEEDERRERRIAGPQVGGVNPLEGGTRGAPGGGFVPSMQGVPESPFTRTGEGLDIRGSQGFP\n",
    ">var.Deltavirus.hdv7:AJ584844\n",
    "MSHADTKRSRKGREETLSKWLKAREDAEELERRLRKTKKTIKKLEDDNPWLGNIKGIIGKVGTGEGAPPAKRPRTDRMEVDSGPGKKSNKGGFTDEERRAHRRRKALENKKKQLSAGGKSLSKEEEEELRKLTADDEQRARRIAGPRVGDVNPPEGPPRGAPGGGFVPQLLGVPESPFSRTGEGLDIRGDRQFP\n",
    ">var.Deltavirus.hdv3:L22063\n",
    "MSQTVARLTSKEREEILEQWVEERKNRRKLEKDLRRANKKIKKLEDENPWLGNVVGLLRRKKDEDGAPPAKRPRQETMEVDSGPGRKPKARGFTDQERRDHRRRKALENKKKQLAGGGKHLSQEEEEELRRLARDDDERERRTAGPRPGGVNPMDGPPRGAPGGGFVPSLQGVPESPFSRTGEGIDIRGTQQFP\n",
    ">var.Deltavirus.drdvb:MT649206\n",
    "METPGKKKPPKPRQETLEEWCELGKRKRELEKELQRITRKRKRLEEENPWLGNVLGITRQKSGGSETPSGKKKRREEEMEVDGAPGSGAAPRTPFTKKEREDHRRRCALENKKKQLEQQGKKLSEEDEAERRRLAEEDERRKRRLEERGDGDVNPPEGPPRGAPGGGFVPGLQGVPESPYSRTGEGLSKRGEGYFP\n",
    ">var.Deltavirus.drdvb:MT649208\n",
    "METPGKKKPPKPRQETLEEWCELGKRKRELEKELQRVTRKKKRLEEENPWLGNVLGITRQKSGGSETPSGKKKRREEEMEVDGAPGSGAAPRTPFTKKEREDHRRRCALENKKKQLEQQGKKLSEEDEAERRRLAEEDERRKRRLEERGDGDVNPPEGTPRGAPGGGFVPGLQGVPESPYSRTGEGLSKRGEGYYP\n",
    ">var.Deltavirus.ratdv:MK598003\n",
    "METPGKKRSPKPRQEILEEWADLNRKKRELEKELQRTQRKKKRLEEENPWLGNVLGIVRQKAGGSDAPQAKKRRLGEEMEVDGGPGPSSAPRAPFTKKEREDHRRRCALENKKKQLEEQGKKLSEEEEAERRRLAEEDERRKKRAEGGGDGDVNPPEGTPRGAFGGGFVPGLQGVPESPFSRRGDGLSLRGEGEYP\n",
    ">var.Deltavirus.drdva:MT649207\n",
    "MESTDPKKNPKGREDTLREWVKGRERKEELERELGRLTRKLKKLEERNPWLGNVLGMVRKRDGGEGGTPSKRPREDPMDVDPGAGGLHHRPRFSEKEKQDHRRRKALENKKKQLSGGGKSLSQEEEEELRRLTREDEERERRVAGPPGGDVNPDAGGPRGAPGGGFVPNMQGVPESPFARTGEGLDPRGDRHFP\n",
    ">var.Deltavirus.snakedv:NC_040729\n",
    "METPSKKQIPTPSREDILEQWVELGKRKKELEKELQKVTKKKRKLEEQHGFLGNVLGIVRGKEQKPAATPQKKRKAEESMDVDGGSRLPPKEIKKRIFTEEERAEHRRRGQLENKKKQLEGRGKQLPEEEQKLLAELTRKDEERKQRFHYGGAGEVNPLEGQSRGAFGGGFVPSTQGVPESPFHRTGTGLDVRGDKMFP\n",
    ">var.Deltavirus.toaddv:MK962760\n",
    "MESLNPSRVKAVEEWAVTRREIAALEEKLAKKRKRVKELEKDNLFLGNIKGMCSKGTKKKASRSPTRRKEPQPGEPGTSGTAKKRKLDSSSKPKSSSSGGPKIPVSPAPKISKGTSSASLDEENFARTLAARAGDPRFREDLERLLSLEDTTRRGRLDSESSESTFSAPGKS\n",
    ">var.Deltavirus.aviandv:NC_040845\n",
    "MENKEESQKKKRGREETLQKWVDDRKRKRELEEELEKLRKRIKDRERKNPWLGNLKGMLKQDSGGGQEKKKEAAEPMDTGESPKKKKKKDYEGLSFTPEEKQRHKRKCDLENKKKQLNAKGKQLTSQEEDELRNLQEEDKKRLLKKREERERRSGGVNLFGSASPSTSGGGNASSTQGLRLPWQK\n",
    ">var.Deltavirus.fishdv:MN031240\n",
    "MADEQARFGLAESYLGGYRRQFQLELELIETKRKLREMEKENPWLKMVKGQTRKGSQGIQTILGKREKVAEAGPSGKKQRKADLSGEERARKRNMSLFKQQENQVRMKLAAKKPVSEELAKALLSSSNLVERQLDDDLVKDLTERGFLSPDQTGPYLLRWECKTRAEREGRLSPLSDMGE\n",
    ">var.Deltavirus.newtdv:MN031239\n",
    "MASTSEARVPPVKKGKTPSRERVLSKWVASRDKLAALLREEAAERAKILALERDNPWLGTIKGIIRKSAESAIPSGPLAGPSAPSTGGDPPPKKRKRTSEKLEASPDQKAIWRAGKDAVNRAKQAMSKGKEVSSELIRQMLLAHDKLGMEVPEDVQAYVMTEVNRSAQTAPASDFGTREEEMAMVPVQMGAGPPVLSIRGFESKDGYNNGQ\n",
    ">var.Deltavirus.termitedv:MK962759\n",
    "MELRGESETTLDRWLKLRESVRRRSLRLKESQEQLRLLEQQQPWLGSVKGTLRKKDGSSFDIPEGPKTTKAAPQEETQTAPPEEKVGPPRRERERKAEKRPAAPLMEKSRREAEAKRKKLQPPHTPAASSTPRRVHHLGDLSEIPVSLLKKALEEAEGRKPGMETGESSEEGAEGSESDFWKMN\n",
    ">pisu.Nidovirales.pacific_salmon_nidovirus:QEG08237\n",
    "DKSIRSVDVNITKGDGSRVSFIGTCKKSNTIQMDTGHGRSILKMSLKSEDAIERKFFPSYANTGVVCDHQFHDFVCENDTLRCIIRHHCTEYTLSDLVHSFRTADFAKMETIMLTLPQYTQTDFIPFWYDPVVNTNVYADIMSRLAPTLLQCLTHTTVFVDAIVANGHCVVLTLDNQNLNGNWLDFGDFTDLQIPGVGVAATNSYYSYLLPLLGATQCFSSVFDPSFNEFGYDFTEQKLLIYSTYFRWERVGIYHPNCVDCVDEECLLFCSNLNVLLSMYIPFNYLGPLCSIEYFNGMATPFCAGLSSVELGIVHNSCDLPSDPILKACLIFGTPPIHVVSSTPTADTRLNIPCVGSGKLMPGKLLKPAMIDTEFYEFVKDALLYEGSPITIEHFFYLQPRDCAVTDFDYYRFNRPTVLDPLQFRFVYNVVKHYFKSYSAGCLKSEFVIINNPHKSSGFPFNQLGDASDVYDFLGHERVDELYAYTKRSVLPTLTKIITKDAISAKTRARTVAGVGILSTATNRCAHQHMLKDIAAQRGRTVVIGTPKFYGGWHEMLSRFNDNRDRKLFGWDYPKCDRSMPNLLRMSAAWMFLNKHPCCSLLDNKYRLANELSQVLSETCVVNGAFYRKPGGTSSGDATTAYANSAFNLFQVITSAVNRCMSTDRFDYRSFNHELYSNIYLLDTPDRGFVEKFYNFMDDTCPLMILSDDGVAGPLVDHPLAIDVSDFRATLFYQNNVYLDDSKCWTEADVLKGPHEFCSQHTILHEGIFYPVPNPSRILAACMFVNSMEKADPKCLVERLVSLAIDAYPLVHHSIEVYRRVFPQILAYIKQVVSSRDMDIYDMFGEFTDFCSGVDITSETFYARLYKCKSSLQ\n",
    ">pisu.Nidovirales.ambystoma_mexicanum_nidovirus:SS0000010\n",
    "MKFNCLQLDCDGERILLKLLNKDEQVNELAAIPYTASNYFCKAVYAKICDRDCLLRFNCTEYTLADLAYAFKTADYIMLETIAKTLPAFVDFKFSAGWYDPIEHPESYDAFMTLLRPTLIQCLVATSEVYDVLDANYFVGTLSPDNQNLNGHWLDFGDFTKCAASVVMVDTYYSYIMPLLGAARVFTDVFDHTNYKGCYDYSDEKRRLFLKYFKGWSGIYHPNCHQCLDEMCVLHCANYNILLSMYINPKFLGPITGIFNYCEYTVPVCIGHSSVELGIVYNNDSLPEDDILRLMLIYGTPPIHVAASTPYVNNSTSLYTAGFIKHMSTRLDVPAYIDTDFYEHCRDILLHPGSTINLKHFYFLVNRDSASSDFDLYRHNRPTVFDIKQFQFVYKVVQKYFEPYDAKCLPNEDVIVNSPNKSSGFPFNTIANAGGILEYMQNIRVDKLFAWTKRNVLATLTKVNLKNAISAKDRVRTVAGVSILSTCTNRCAHQHLLKNIAATRDATVVIGTPKFYGNWDTMLRRFMDGENRNLIGWDYPKCDRSMPNLHRIAAAWLFLNKHTSCCNHTERFYRLVNEHAQVISETCLVDGALLYKPGGTSSGDATTAYANSVFNLFQVVTASVQHMLKKNVPDLLRDVKVLDSNSVTIRNYRDVVHNIYECIYLCPYQEAVQQFYDLMLKVCPLMILSDDGVAAPMVEHPFALTLDDFKGVLFSQNNVVLTSDKCWTQSDLTKGPEEFCSQHTIFSEGSFYPVPDPSRIIAACIFVSKMEPTTDSYKIERLVSLAIDAWPLTLHHDENYKRVFPVILDFINKCADRIQDDVSFDVPDCVNVTPHVIRSESFYRNMYVKSAMLQAGATMCVVCQIPTILSCSSCVRPVPLCCMCMYHHVATCNHHRVATTTVMCCNHRGCACDDVNKLYFDGKDGLIRCLDHVTSTITLPIMDPSSETIFSLGRDKCVPNGDIHVFNQAILDGFSKVANYDTFGKPYHLKCQELVRAMEE\n",
    ">pisu.Nidovirales.puntigrus_tetrazona_nidovirus:SS0000011\n",
    "MDLCNAFTTADFFRLQEIALTLPEFADFDFCDNWYDPITNPDSHDLFYSLLSPRLLNILKCTCAYFDHCTAKSFAGVLTLDNQNLNGRWLDHGDFTPFPEPAIVVDSYYGYLLPILGRCNVFKDSFDVNNHDDFDYSKEKLRLFYEYFRWSVKYHPNCVDCCDSYCVLWCFNFNVLFGNNIPVNHLGPLVAKGQFHDLLSPVCCGYASIELGIIRNRYEYPTDPITTASLTFASPITHVAASKPCVDHRINPPCLGSTGKIISKVLPPAKFDIDFYNLLKEEITYEGSPITLKHFFFMQTRDSAVTDFGYYGYNRPTVCDPYKLCFAYSVSSLYFDCYEAGCLPNDKVVVANPKKSAGFPFSQLGTAGEVLEFLSSERLDELHAASKGAIIPTITKVNTKYAISAKDRARTVSGVSIVSTALNRMTHQMLLKSIAATRSATCCIGTPKFYGNWDAMMSRFCDGVSRTLIGWDYPKCDRSMPNILRAGASMAFLDKHQGCCTPTDKFYLMANELCQVLHDTAYLEGNFYYKPGGTSSGDASTAYANSHFNIMQVVSASVSRCLNSPASGYYDFQSSLRDSIYNPNRGFHDFAHTFFGYMQSTNPLLILSDDGVCGPSKDHPLALTLNDFKGTLYSQNNVVLTTDKCWMEENPSHGPEEFCSQHSIRHNGVLYPCPDASRILSACVFMSDEIRADPQLMLERYVSLAIDAYPLVYHKCEVHRKVFPVLLRHIRSVHASLKQAVSFTGGFDGDPDILNEDFYRRLYECKTTLQATSNTCIICESPTIIYCSSCPRKHMLCCSCAFFHVKDTGHTTIAHYKDITCCVSGCTSDITLLHYAADGSVRCVDHLASSKYSLPIYDTAAQKIFVMHHDACRYDPSCHTYNSGVAKGYSEVANYDVYSLPMSAQLACHENIQVLEEMSKRTYGSATVDVVVNNTFPPMVVVSWDGPIPPINKNSKFSATKNRSNYGEFTLTTASSGSSHYFYESLTNARLVQGCVLKLV\n",
    ">pisu.Nidovirales.crotalus_viridis_nidovirus:SS0000012\n",
    "MKTVNFKVGFDNFTVKCVDEKELARELIWYRRLKNKVKVVPHYSCKIGRNFALIRGPMTEKSLGDYVYKYFKADDSIGIDREDGVSFNERRNWYKNNLFNDELFGNLDVLYDNSNLLGGIISLDNIDCKGYLYDFGDFNGQESYLDKALADLCSFYSSIGLDKRNEFKKWFSVDYNVLKSGDWLDKVLMLNNSILNSSSKVTDLFIDVDEVCTSAYYNQLFGYCKIDQSKIEEVDLLLQYIYKLQDPALRGLEPVKSVSDAIVVHKYKDIKSDNMVKNVYYNIDSYEDLKNRGVDCDSALKYTYFQGDNVDVMVDFMYYNYQGDLFLQPHILKFLYKQTLKMFEVVRTCKRFNYEECKPNKSSLGPSEHLLTSLKQDYVYNHLNESVVDQIVKLAAETPLIFLTKISEKFALTKKARARTIAACSMFSSLLFRALHKPVTSSMVSNAQQYKVPSLIGVSKFYNNFDGYIKSRYGDLNDYKIFGSDYTKCDRSFPLIFRAAAAAMLFELGEWDFDNYYFINEIQAFMLDFVMVGDCILQKPGGTSSGDATTAFSNTLYNCMVHLYVQLASLVSCENECKDPVFKAAAVKLWLTGDSSDYEMMLDYYNSEVYRFNFLSDDSFILTNEGKCVLDIYNCNNFSLHLERLIHVNVDVNKAWEGKDIHEFCSSTIVKVEGAWQYVAEKDRLLASLLITGKESDLELDIIRTSAILAEAAVYYWVDNMFFMALYDHFILKCQMYFNKYGCHVLPLVM\n",
    ">pisu.Nidovirales.hippocampus_kuda_nidovirus:SS0000013\n",
    "MSCRIKFLFSFKLTEGSCKFVNLRGIYRNPNNKFSISLSPAKRAVDVSCQRGDGHRVSFIGYVNKSNTLQIDTANERIILKCVHESEYQAEYDAYNRVKSNTSVCRHAFLKIFGMLFVCRRDCTEYTVMDLCYAFRTANFPKLKEICLSMGLSEGLFVEGWYDSVEGYDAYIDFTQSLRPYFLNILSSTKHYIEGLKVAGLSGVLTLDNQNLNGKWLDHGDFVFTTEASIVVESYYSYMLPLLGVCNVFYDVFDHANCQYGYDYTNEKLRLFETYFSWPITYHPNCVDCSSDICLLTCTNFNILFSNMMDLAKVGPIITVTNYCELTLPTTMGINSIELGVIYNDGDYPSDPIIRTIIHYGTPPVHVLSSTALTDHRVQFYSVCSVNRNRSKLQPPAVIDFSFYNYVKDKLLFAGSSINLNHFFFMQRPDAAVNDFAYYRYNKPTVLDIRQAIFCYKVTCLYFGEYVARCLPDEEVIVANPHKSAGFPFSGFGTAGEWLDFVESHRLNELHACAKRCILPTITKTNVKVAISGKSRARTVAGVSILSTALNRMTHQHLLKDITSKRDATICIGTPKFYGNWDTMMSRFRDGKPRTLFGFDYPKCDRSMPNILRQAACMAFLAKHECCNVRERFYLLANEMAQVLTEYTYLEGGLYRKPGGTSSGDASTAYANSFFNVFQVVTSSVNRCLHSKCAPYVDFQSQLRANLYDCVDDPLFVSQYQKFLQHTCPLLILSDDGVAGPLVEHPLSLTLEDFRATLFYQNNVYLSDGKCWVQPDVNKGPDEFCSQHTIVHNGCLYPVPDPSRIIAACCFVSTAERSDPTMQLTRLVSLAIDAYPLIYHNDETYKLVFHCILAYIRQLHKSVHTEALAGICEAFSDYGGSVLDEGFYKRLYETRSTLQS\n",
    ">pisu.Nidovirales.syngnathus_typhle_nidovirus:SS0000014\n",
    "MLFVSRRQCTAFTVMDLCYSFRTANFVLIKEICLSMGMPESLFVEGWYDSVEGFDAYVEFVTALRPYFLRILECTKSYITGLKAVGLSGVLTLDNQNLNGKWLDHGDFVYTGESAIVVESYYSYMLPLLGVCNVFYDSFDHSNCKYGYDYTDEKLRLFNTYFDWTITYHPNCVDCVDDICLLTCTNFNILFSNMMDLALVGPIVTVTNYCELTLPTTMGINSIELGIIYNDGDYPSDPITRTIIHYGTPPVHVLSSTAMKDHRVDFYSVCSVNRNRCKMQAPSVIDFAFYNFVKDKLLFAGSSISLSHFFYMQRPDAASTDFAYYRYNKPTVLDIRQAIFCYKVTCAYFDDYSARCLPDEEVVVANPHKSAGFPFSTFGTAGEWLDFVEPHRLNDLHACAKRCIIPTITKTNVKVAISCKSRARTVAGVSILSTALNRMTHQYLLKDITAKRNKTICIGTPKFFGNWDAMMSRFRDGKSRTLCGWDYPKCDRSMPNIIRQAACMAFLAKHECCSTSERFYLMANEMSQVLTEYTYLEGGLYRKPGGTSSGDASTAYANSYFNIFQVVTSSVNRCLHKKCEPHLDFQAQLRSNLYDCVDDPMFVKQYREFLQHTCPLLILSDDGIAGPLTDHPLALTLEDFRATLFYQNNVYLSDSKCWVEPDVNKGPEEFCSQHTIVYNGCLYPIPNPSRIIAACCFVSTAEKSDSTMQLTRLVSLAIDAYPLIYHKDETFQLVFPCILNYIRQLHQTVHSEALAGICEAFSDYGGSVLDEDFYKRLYETRSVLQSSTGGCVICDSNTIL\n",
    ">pisu.Nidovirales.acanthemblemaria_nidovirus:SS0000015\n",
    "MLPLLGASQCFYNQLDITLVQNEYDYTKCKLALYEHYFRWHLTYHPNIVDCIDDMCVLTCSNFNVLLGMGIPWNLFGPLVSIDFLHGQQVPVCSGFHSLELGIILNEKPLPTEHLLRAAIVYGTPLLHTTTAIPTIDTRTQIPCVGSGKLISGKVTRPAFIDFEFYEFAKDALFFDGSSINQQHFYFLHPQDSTMEDFNYYRFNIPTVVDICQFKFVFSVTEHYFSMYSATCLKSEFVVVNNPNKSSGFPFNQLGNAGDIYSYLGHERIDELFNYSKSAIVPTITKVIVKSAISAKDRCRTVSGVGILSTMMCRCMHQCLLKEIAATRDATVVIGTPKFYGGWHSMLSRFVDKYERELIGWDYPKCDRSMPNMLRLASAWLFICKHKCCNLHDRIFRLGNELSEVLTETCLVQGVYYSKPGGTSSGDATTAYANSVFNIFQVVTTAVNHCLQQDLPEHRPFNVKLYNSIYRLQDKEFVSEFYNWMQVKVPLMILSDDGVAGPIANDPQCLNLDSFKSTLFYQNNVVLTPDKCWRQRDISKGPHEFCSQHTIMHNGIFYPVPDPSRILAACMYTNSMEKSDPHLMCERFVSLAIDAYPLVYSNNETYRQVFPTILGYINKVVKDRDLNIYSLLGDYWDNSGASILSEAFYS\n",
    ">pisu.Nidovirales.takifugu pardalis_nidovirus:SS0000016\n",
    "MRAKCTEYTVMDLCHAYRSADFERLQTILYTCGFTPLDFADNWHDSVEGFDAYNHWVSKMDTLFLNILTKTTSYYDHLTANSMIGVLTLDNQNLNGNWLDHGDFTCSTDSAVIVDSYYSYMLPLLGACRCFKSVFDHENKEHGYDYTAEKSRLYEQYFRWHLPYHPNCVDCPNTRCVLACSNFNTLFGMYINPDLLGPICAIENFCEIRTPTTIGCHSIEQGVTFSEGSFPNDVVTRTCIYYGTPPVHVAASTPLVDHRMSVHPIGSVGRPRSKMQQPGFVDQVFYEFAKNELFFPGSSIELKHFFYMQERSAGVADFDYYRYNRPTVLDICQALFCYEITLDFFSDYEAGCIPDEQVVVANPNKSCGFPFSQFGNAGDFLEFLGSDRINSLHKACKSNILPTITKSNVKCAVSGKSRARTVAGVGILSTAMSRCTHQKLLKGIAAARDKTICIGTPKFYGNWHRMMSRFTDGEDRILIGWDYPKCDRSMPNWIRQAASLGFLSKHECCNHSQQFYLCANEMCQVLSEATLLDGNLYRKPGGTSSGDGTTAFANSFFNIFQVATSCVNRCLTTRVEGHLDFQSLLRDTVYGYSPPEEFVPTFRAFMQHTCPLLILSDDGVAGPLQTHPLSLKLQDFTATLFYQNNVFLNETKCWTEPHVNLGPEEFCSQHTLVHNGVMYPCPSPSRILASCCFVSNAEKSDSVLMLTRLVSLAIDAYPLVHHPDETYRRVFPVMLKYIRHVYGIINKEAL\n",
    ">pisu.Nidovirales.caretta_caretta_nidovirus:SS0000017\n",
    "MCNKERTTLQVIRSDCSLYSLLDLACAFRSADFDLAIEIAKSLKIDTSSWPSNWYDPIEQHSSKYWSAFGPAVFKCLLAANDFGDFLIEKGIIATLTADNQCLNSKWLDFADYSHNLNCPGAVSMDSYYSYLLPLINVVNYITVDSPFTVTDVLRYDYSEEKMSLFKKYFRHWPYEYHPNCINCDSDRCTIHCINFNVLLSQPYSTTNGPLISKCHFNGQPIYATCGYVSAELGVVINSDNLLELPSEFVKVFAILCDASLHITTSEALLDCTEGIIPIGLISLRKNTVKPGEVNLDFYDHLEKYGILHPGTGINLKHYFFMQDSTAATADFSLYKYNKFTVVDILQYLFCYEVTNCYFEVYEGGCIPASEVIVRSPHKSAGFLLNKISTAGGFYDSISLAEQDMLYAYTKSNIIPTVTQINLKHSIAAKERARTVAGVSLLATMTTRQCQQKLLKSIAATRGGTIVIGTTKFYGGWDSMMRTLIDGVSNPIMYGWDYPKCDRSYPTILRLTSSILFSRFHNSCCSDSEKFFRLCNELAQVVTESSFCNGALFYKKGGQSSGDALTAYSNSCFNIMQVVSSTVGEIISRPTRNVPDNFKKFQRSIYSAVYTTSLHSPDLATVRLLYQHMKTYCHLMILSDDGVATVDGSLMESGMVPSLDSFKKLLYYQNNVVLADDKCWTETNMSVGPHEFCSQHSKLVTIDGEVVYLPMPDPSRIISALVYVDKVEKASSYVVMERFVAMAMEAWPLKHHDNVHVRKIFPCILDYISNLYNSLSADVLSSFGCDVEYDGGTFTSETFY\n",
    ">pisu.Nidovirales.microhyla_fissipes_nidovirus:SS0000018\n",
    "MFYNILKSLSIDYTNWPTNWYDPVECFKCYYYWDLILTPVLEVCLTNANLFTSYCNANGIVGVLTADNQCLNGLILDFGDFHLAPVGSVQMRSYYSYLMPVITMSTLFNKFNADEYDTSVYDFTEFKLKLHNMYFKHWEFDYHPNCIDCNNDLCQLHCGNFNTMYTQYIPTYYFGPLLGVEYYNGTPVYQTNGVSTKFFGVVLNKAVFNTSSQFLSLFKLLSDASLHILSSKKFVDLNIDMPASAFAALPNRSVPPAGFFNEQFYNFCVEQGLFTDNTPLDIKHFYFMQESTSMVNDYGLYIHNRATSVDICQFLFVVDYAFQYFKDYEGGCLHPDAVYVRNKHRSSGFVFSSFGDAGNFYESLTAEDFINLYEYTGRAVLPTVTKLNLKISLSAKSRARTVAGVALLSTICGRLYHQKALKSIAQSNDKPVVIGRTKYYGGWDDMLKSLLGVVDQPMFIGWDYPKCDRAMPNILRIAASLILSRLHVDCCTKENKFFRTVNEYVQVICEYVYCSGIFYNKPGGTSSGDATTATANSVFNIIQYLSSFISKIIGTNYTTHVKTDDCPIDIVKLQSSVYSAIYADGDKSSVIKELHSISPHFFNLMILSDDGVACCNKNYYDFISPSLSDFGNLLYYQNNVFMSQDKCWTQLDVTKGPTEFCSQHTLLVCIDDEWVYLPYPDPSRIMAAALFVNKADGSLPSDVVERFVSLAIDAYPLVHTGNPTYSKVFYLLLDYIKLNSIYIAETLFEH\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Revision 4 / RefSeq RdRp Set / Testing Diamond\n",
    "\n",
    "Create the \"High Confidence\" set of RdRp from RefSeq. Note it looks like this is already implemented as a sub-set of wolf18\n",
    "\n",
    "Run Diamond with different settings to measure performance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Set-up Folder\n",
    "# RefSeq sequence extraction\n",
    "mkdir rdrp0; cd rdrp0\n",
    "\n",
    "# Download rdrp0 work folder\n",
    "aws s3 sync s3://serratus-public/notebook/201226_rdrp0/ ./\n",
    "\n",
    "# RefSeq (rs) folder\n",
    "mkdir -p rs; cd rs\n",
    "\n",
    "\n",
    "## Set-up Database\n",
    "GENOME='rdrp0'\n",
    "cp ../rev1/rdrp0_r1.fa ./rdrp0.fa\n",
    "diamond makedb --in $GENOME.fa -d $GENOME\n",
    "\n",
    "## Set-up Query\n",
    "# Query is Viral GenBank v240\n",
    "# Extract out Viral RefSeq (NC_XXX)\n",
    "QUERY='vrs240.fa'\n",
    "\n",
    "wget https://ftp.ncbi.nlm.nih.gov/refseq/release/viral/viral.1.protein.faa.gz\n",
    "wget https://ftp.ncbi.nlm.nih.gov/refseq/release/viral/viral.2.protein.faa.gz\n",
    "gzip -d *.gz\n",
    "\n",
    "cat viral.1.protein.faa viral.2.protein.faa > $QUERY\n",
    "rm viral*faa\n",
    "\n",
    "## Run Diamond -- DEFAULT MODE\n",
    "QUERY='vrs240.fa'\n",
    "GENOME='rdrp0'\n",
    "OUTNAME='vrs240_r0'\n",
    "\n",
    "## test\n",
    "# head -n 100000 $IN > tmp.fa\n",
    "# IN='tmp.fa'\n",
    "# OUTNAME='tmp.pro'\n",
    "\n",
    "# Diamond blastx alignment\n",
    "time cat $QUERY |\\\n",
    "diamond blastp \\\n",
    "  -d \"$GENOME\".dmnd \\\n",
    "  --unal 0 \\\n",
    "  --masking 0 \\\n",
    "  -k 1 \\\n",
    "  -p 16 \\\n",
    "  -b 1 \\\n",
    "  -f 6 qseqid qstart qend qlen qstrand \\\n",
    "       sseqid sstart send slen \\\n",
    "       pident evalue cigar \\\n",
    "       qseq \\\n",
    "  > \"$OUTNAME\".pro\n",
    "\n",
    "# real    0m3.656s\n",
    "# user    0m41.907s\n",
    "# sys     0m3.899s\n",
    "\n",
    "# Fasta of local diamond matches\n",
    "#\n",
    "# query_acc query_start query_end query_len\n",
    "# db_acc db_start db_end db_len\n",
    "# pctid\n",
    "cut -f1,2,3,4,6,7,8,9,10,13 $OUTNAME.pro \\\n",
    "  | sed 's/^/>/g' \\\n",
    "  | sed 's/\\(\\t\\)\\([A-Z]*$\\)/\\n\\2/g' - \\\n",
    "  > $OUTNAME.fa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Run Diamond with ULTRA-SENSITIVE mode\n",
    "QUERY='vrs240.fa'\n",
    "GENOME='rdrp0'\n",
    "OUTNAME='vrs240_r0US'\n",
    "\n",
    "# Diamond blastp alignment\n",
    "time cat $QUERY |\\\n",
    "diamond blastp \\\n",
    "  -d \"$GENOME\".dmnd \\\n",
    "  --unal 0 \\\n",
    "  --masking 0 \\\n",
    "  --ultra-sensitive \\\n",
    "  -k 1 \\\n",
    "  -p 16 \\\n",
    "  -b 1 \\\n",
    "  -f 6 qseqid qstart qend qlen qstrand \\\n",
    "       sseqid sstart send slen \\\n",
    "       pident evalue cigar \\\n",
    "       qseq \\\n",
    "  > \"$OUTNAME\".pro\n",
    "\n",
    "\n",
    "# Fasta of local diamond matches\n",
    "cut -f1,2,3,4,6,7,8,9,10,13 $OUTNAME.pro \\\n",
    "  | sed 's/^/>/g' \\\n",
    "  | sed 's/\\(\\t\\)\\([A-Z]*$\\)/\\n\\2/g' - \\\n",
    "  > $OUTNAME.fa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Run Diamond with ULTRA-SENSITIVE mode with BLOSUM45\n",
    "QUERY='vrs240.fa'\n",
    "GENOME='rdrp0'\n",
    "OUTNAME='vrs240_r0USB45'\n",
    "\n",
    "# Diamond blastp alignment\n",
    "time cat $QUERY |\\\n",
    "diamond blastp \\\n",
    "  -d \"$GENOME\".dmnd \\\n",
    "  --unal 0 \\\n",
    "  --masking 0 \\\n",
    "  --ultra-sensitive \\\n",
    "  --matrix BLOSUM45 \\\n",
    "  -k 1 \\\n",
    "  -p 16 \\\n",
    "  -b 1 \\\n",
    "  -f 6 qseqid qstart qend qlen qstrand \\\n",
    "       sseqid sstart send slen \\\n",
    "       pident evalue cigar \\\n",
    "       qseq \\\n",
    "  > \"$OUTNAME\".pro"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Run Diamond with ULTRA-SENSITIVE mode with BLOSUM45\n",
    "QUERY='vrs240.fa'\n",
    "GENOME='rdrp0'\n",
    "OUTNAME='vrs240_r0USP250'\n",
    "\n",
    "# Diamond blastp alignment\n",
    "time cat $QUERY |\\\n",
    "diamond blastp \\\n",
    "  -d \"$GENOME\".dmnd \\\n",
    "  --unal 0 \\\n",
    "  --masking 0 \\\n",
    "  --ultra-sensitive \\\n",
    "  --matrix PAM250 \\\n",
    "  -k 1 \\\n",
    "  -p 16 \\\n",
    "  -b 1 \\\n",
    "  -f 6 qseqid qstart qend qlen qstrand \\\n",
    "       sseqid sstart send slen \\\n",
    "       pident evalue cigar \\\n",
    "       qseq \\\n",
    "  > \"$OUTNAME\".pro"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### RUNTIME -----------\n",
    "\n",
    "## DEFAULT\n",
    "# real    0m3.656s\n",
    "# user    0m41.907s\n",
    "# sys     0m3.899s\n",
    "\n",
    "## ULTRASENSITIVE\n",
    "# real    1m39.734s\n",
    "# user    24m28.415s\n",
    "# sys     0m5.550s\n",
    "\n",
    "## US-BLOSUM32\n",
    "# real    1m37.864s\n",
    "# user    24m5.467s\n",
    "# sys     0m5.500s\n",
    "\n",
    "## US-PAM250\n",
    "# real    1m33.032s\n",
    "# user    23m24.756s\n",
    "# sys     0m3.548s\n",
    "\n",
    "\n",
    "### HITS -----------\n",
    "# 4379 vrs240_r0.pro\n",
    "# 5056 vrs240_r0US.pro\n",
    "# 4919 vrs240_r0USB45.pro\n",
    "# 5359 vrs240_r0USP250.pro\n",
    "\n",
    "### EDGECASES -----------\n",
    "grep \"YP_009480680.1\" vrs240_r0USP250.pro  | cut -f1-4,6-10\n",
    "grep \"YP_009551960.1\" vrs240_r0USP250.pro  | cut -f1-4,6-10\n",
    "grep \"YP_008802665.1\" vrs240_r0USP250.pro  | cut -f1-4,6-10\n",
    "\n",
    "# Spot compares of edge-cases\n",
    "## DEFAULT\n",
    "# YP_009480680.1\t446\t565\t578\n",
    "# rdrp2.unc0336.Grapevine_associated_mycovirus_2:ADO60939\t1\t120\t325\n",
    "# 85.8\n",
    "## ULTRASENSITIVE\n",
    "# YP_009480680.1\t446\t565\t578\n",
    "# rdrp2.unc0336.Grapevine_associated_mycovirus_2:ADO60939\t1\t120\t325\n",
    "# 85.8\n",
    "## US BLOSUM45\n",
    "# YP_009480680.1  446     565     578\n",
    "# rdrp2.unc0336.Grapevine_associated_mycovirus_2:ADO60939    1       120     325\n",
    "# 85.8\n",
    "## US PAM250\n",
    "# YP_009480680.1  328     562     578\n",
    "# rdrp2.Fusarividae.Sclerotinia_sclerotiorum_fusarivirus_1:YP_009143301      5       233     506\n",
    "# 44.3\n",
    "\n",
    "## DEFAULT\n",
    "# YP_009551960.1\t492\t725\t1010\n",
    "# rdrp1.Narnaviridae.Rhizophagus_sp_HR1_mitovirus_like_ssRNA:BAN85985\t116\t338\t482\n",
    "# 35.4\n",
    "## ULTRASENSITIVE\n",
    "# YP_009551960.1\t491\t751\t1010\n",
    "# rdrp1.Narnaviridae.Macrophomina_phaseolina_mitovirus_3:AMM45292\t120\t383\t474\n",
    "# 33.6\n",
    "## US BLOSUM45\n",
    "# YP_009551960.1  493     725     1010\n",
    "# rdrp1.Narnaviridae.Fusarium_poae_mitovirus_4:YP_009272901  120     334     471\n",
    "# 34.8\n",
    "# US PAM250\n",
    "# YP_009551960.1  491     740     1010\n",
    "# rdrp1.Narnaviridae.Macrophomina_phaseolina_mitovirus_3:AMM45292    120     361     474\n",
    "# 32.7\n",
    "\n",
    "\n",
    "## DEFAULT\n",
    "# YP_008802665.1\t31\t147\t285\n",
    "# rdrp2.Picornaviridae.Kobuvirus_sewage_Aichi:BAO02685\t4\t133\t330\n",
    "# 33.8\n",
    "## ULTRASENSIVE\n",
    "# YP_008802665.1  31      147     285\n",
    "# rdrp2.Picornaviridae.Kobuvirus_sewage_Aichi:BAO02685    4       133     330\n",
    "# 33.8\n",
    "## US BLOSUM45\n",
    "# YP_008802665.1  31      151     285\n",
    "# rdrp2.Picornaviridae.Kobuvirus_sewage_Aichi:BAO02685       4       137     330\n",
    "# 32.8\n",
    "## US PAM250\n",
    "# YP_008802665.1  30      147     285\n",
    "# rdrp2.Picornaviridae.Kobuvirus_sewage_Aichi:BAO02685       3       133     330\n",
    "# 33.6\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# High confidence set of matches\n",
    "OUTNAME='vrs240_r0'\n",
    "OUTNAME2='vrs240_r0_HC'\n",
    "\n",
    "# Extract only matches that are:\n",
    "# >88% identity to rdrp0\n",
    "# >98% coverage of rdrp0\n",
    "\n",
    "cut -f1,2,3,4,6,7,8,9,10,13 $OUTNAME.pro \\\n",
    "  > pro.tmp\n",
    "\n",
    "rm $OUTNAME2.pro\n",
    "\n",
    "while read -r line; do\n",
    "  \n",
    "  pctid=$(echo $line | cut -f 9 -d' ' - )  \n",
    "  \n",
    "  r0_start=$(echo $line | cut -f 6 -d' ' - )\n",
    "  r0_end=$(echo $line | cut -f 7 -d' ' - )\n",
    "  r0_len=$(echo $line | cut -f 8 -d' ' - )\n",
    "  \n",
    "  r0_cov=$( echo \"$r0_end - $r0_start + 1\" | bc )\n",
    "  r0_cov2=$( echo \"scale=2;$r0_cov / $r0_len\" | bc )\n",
    "  \n",
    "  #echo $r0_start $r0_end $r0_len $pctid $r0_cov2\n",
    " \n",
    "  if [[ $pctid > 0.88 ]] && [[ $r0_cov2 > 0.98 ]]; then\n",
    "    echo $line >> $OUTNAME2.pro\n",
    "  fi\n",
    "  \n",
    "done < pro.tmp\n",
    "\n",
    "rm pro.tmp\n",
    "\n",
    "# Fasta of high confidence diamond matches\n",
    "cut -f1,2,3,10 -d' ' $OUTNAME2.pro \\\n",
    "  | sed 's/^/>/g' \\\n",
    "  | sed 's/\\( \\)\\([A-Z]*$\\)/\\n\\2/g' - \\\n",
    "  > $OUTNAME2.fa\n",
    "\n",
    "# List of accession with High Confidence Hits\n",
    "# Create a file of full-length records with these hits\n",
    "# and associated bed3 file\n",
    "cut -f1 -d' ' vrs240_r0_HC.pro | sed 's/>//g' - > vrs240_r0_HC.hits\n",
    "seqkit grep -r -f vrs240_r0_HC.hits vrs240.fa > vrs240_r0_HC_FullLength.fa\n",
    "\n",
    "cut -f1,2,3 -d' ' $OUTNAME2.pro \\\n",
    "  | sed 's/ /\\t/g' - \\\n",
    "  > $OUTNAME2.rdrp.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create Cross-Validation Set\n",
    "mkdir cvi; cd cvi\n",
    "RDRP=\"vrs240_r0_HC.fa\"\n",
    "\n",
    "# High-Confidence RDRP only\n",
    "ln -s ../$RDRP ./\n",
    "\n",
    "# USEARCH CVI\n",
    "usearch -threads 30 -calc_distmx $RDRP -maxdist 1 -termdist 1 -tabbedout vrs240.distmat\n",
    "\n",
    "# 90% CVI \n",
    "usearch -distmx_split_identity vrs240.distmat -mindist 0.05 -maxdist 0.15 \\\n",
    "  -tabbedout subsets.tmp\n",
    "sed 's/ /_/g' subsets.tmp > vrs240_r0_HC_id90.cvi\n",
    "\n",
    "# 75% CVI \n",
    "usearch -distmx_split_identity vrs240.distmat -mindist 0.2 -maxdist 0.3 \\\n",
    "  -tabbedout subsets.tmp\n",
    "sed 's/ /_/g' subsets.tmp > vrs240_r0_HC_id75.cvi\n",
    "\n",
    "# 45% CVI \n",
    "usearch -distmx_split_identity vrs240.distmat -mindist 0.4 -maxdist 0.6 \\\n",
    "  -tabbedout subsets.tmp\n",
    "sed 's/ /_/g' subsets.tmp > vrs240_r0_HC_id45.cvi\n",
    "\n",
    "# 25% CVI \n",
    "usearch -distmx_split_identity vrs240.distmat -mindist 0.7 -maxdist 0.8 \\\n",
    "  -tabbedout subsets.tmp\n",
    "sed 's/ /_/g' subsets.tmp > vrs240_r0_HC_id25.cvi\n",
    "\n",
    "# Replace spaces with undescores for RdRp coordinates within accession\n",
    "\n",
    "#1 Subset name.\n",
    "#2 Label1\n",
    "#3 Label2\n",
    "#4 Dist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Upload rs folder\n",
    "aws s3 sync ./ s3://serratus-public/notebook/201226_rdrp0/rs/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# From base amazon linux 2\n",
    "sudo yum install -y docker\n",
    "sudo yum install -y git\n",
    "sudo service docker start\n",
    "sudo docker run --rm --entrypoint /bin/bash -it serratus-align:latest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# HMMER\n",
    "wget http://eddylab.org/software/hmmer/hmmer-3.3.1.tar.gz ./"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Wiki Page Draft"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# rdrp: RNA-dependent RNA polymerse\n",
    "\n",
    "### Current Version: `rdrp0`\n",
    "\n",
    "[RNA-dependent RNA polymerse (RdRp)](https://en.wikipedia.org/wiki/RNA-dependent_RNA_polymerase) is the enzyme which catalyzes the synthesis of an RNA molecule from an RNA template. RdRp are a marker for RNA viruses and are essential for viruses to synthesize RNA genomes and mRNA. The similarity between viral RdRp demarcate species, genus, and family level taxonomy at approximately 90%,80%, and 40%, respectively. RdRp is a slowly evolving, meaning related viruses are most likely to show sequence similarity in this enzyme. \n",
    "\n",
    "To isolate every RdRp in the SRA, we constructed a collection of all available viral RdRp sequences called `rdrp#`. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Custom added sequences\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Permuted RdRp\n",
    "\n",
    "FJ977041 \n",
    "\n",
    ">ACV83739.1 replicase-associated polyprotein [Grapevine virus Q]\n",
    "MAAPATAYASPSAAFFALFQEQDLRCFRPLTLAHSLRYDAPVRPRQLPRLRSITVPITSLDEGFTPILIA\n",
    "RPSLPLLGGGLKELVEMLAPTTHRDTVASPILEAVAGPLRTSIQRYPYEVPAHAVPVLQRFGIEASGFGF\n",
    "KAHPHPVHKTIEIHLLFEHWLNLCRSPSAVLFMKQSKFEKLQHENANFEALANYNLTARDTTRYEQVAVA\n",
    "PPTQAVWFMHDALQYFSLSQVAAFFADCPHLEKLFASLVVPPESDFTNLSLFPEIYRYSFAGSRLNYQLE\n",
    "GNPGHSYSQPREALEWLKTTTIRCGNLYLTVTKLESWGPVHSLLIQRGKPSVHLEHDEVSFVGPDAVALP\n",
    "EAAALRQDLRHRLVPRTVYDALFVYVRAVRTLRTTDPVGFVRTQSNKAEYSWVTSAAWDNLQHFVTETAA\n",
    "HRVPNRHFFFNSTFAKCRYWCSQHKLGLLTVTTPPACGLTLFTGAKLASAMSSRLTALAVFHHWVVPPPT\n",
    "LFFTPKAPLLAIQLTRLPQPLFSSVPFLHKPLGKLSLRLLNFDSFLRRFFPDAPIPTWARLLTVAIALSP\n",
    "AVWLAIRHFIGPDAPQALNDHYVRFFHPDRWQLTFERQPRFVALDRTFPWPLPQAPEPTEPRDSDVPLET\n",
    "VPSPLPVVAPLPAPATSVPPVDTSATTASAVEPSLSTESLKTDEAPSGTTILQPRELKDTIFPLPAAALA\n",
    "VTPPEPAPAPAEPVSASTVLGTAPLSRDLHTGHVSTPATEPGLVEPEHSPLAADSSATGEVSEFFNLHPA\n",
    "DWIAPTATFLARRRGETISGAKYPAMDCLLAAVSAGANIPKDALWKTICSYFPDSMLREEDIAKHGLSTH\n",
    "HFAALAREHRLQATFHSAGNQFVLGVEHPSVSFHIDHTPESATAPGHFSLRADERQHSPRLLGGRAADLV\n",
    "HAALKFKVGSAVLPFQQAHDYTTNVARAKNLISNMKNGFDGVLANIDPAHTNESRDRLLSLDGAMDIAAP\n",
    "RDVKLIHIAGFPGCGKSYPIAQLLKSRAFKHFKIAVPTVELRNEWKGVLKVKPQDNWRISTWESSLLKSA\n",
    "RILVIDEIYKMPRGYLDLAIHADPTIDLVIALGDPLQGVYHSTHSDSSNHRLSSEVKHLQPYMDYYCLWS\n",
    "HRVPQDIGTFFGIKSTSTVPGFKSYQANIPGNLRQLANSQSAAKVLNQCGFSSVTIASSQGSTYSAPACI\n",
    "HLDRHSMSLSHAHSLVALTRSKSGVIFTGDKRVLEAPGGNLLFSSYFQEKKVDLRALFPTEFPCRPILLE\n",
    "PLKRRPTDLTGGAPFPFRDEARVFNPERRDDVFVEAAVVCGDGSSNAPQVSTHFLPETRRPLHFDLPSAK\n",
    "PEFAAHEAPAPLTDTFIEPVYPGETFENIAAHFLPAHDPEVKEILFKDQRSNQFPFIDQPFHVGAQPASL\n",
    "CAAVHHSKKDPTLLAASIEKRLRFRASDAPYQITAKDEILGSILFEAHCRAMRRDPNVRVPFDEALFAEC\n",
    "IALNEFAQLTSKTQAVIMANHERSDPDWRYTAVRIFAKNQHKVNSGSLFGPWKACQTLALMHDAVILLFG\n",
    "PVKKYQLIHDERDRPEHIFIYAGRTPQEMSEWCQKFLTPRSASSPVPVMVSGDDSLIGCHPHFVANDYTA\n",
    "FDQSQHGEAAVLERLKMERVNIPEWLITLHIMIKTHITTQFGPLTCMRLTGEPGTYFDNSDYNLAVIFLE\n",
    "YSMSGQWLSENPLWPAIKPLLALRFKKEKTRYGNFCGYYVGAAGAVRMPRALFAKILIAVEDASIADKMA\n",
    "SYATEFAIGHSLGDALWSLLPVEEVVYQSAVFDFFCRNAPRELKLLFKLGPVERSVVEAVQEFATWASYA\n",
    "FYRFLNSAQRKVLLTRSPQLHFPGDAPEVSQLQGELLQSFSMMQPTFPLTGGLLLPRAVDAPMSDDSPAG\n",
    "RARSQRDPDHRVDPQPPLPLAPSVQETSGGPAITVPFQWVALVVKSESTIFTVDPPRAKSLTQLIGPYCH\n",
    "ARLLSLEAILMPTLNAFQNPVTVHMVWTVNTVQPASGEELFYPGGQALTVGGPVSMSALATVPADVSRLN\n",
    "PVIKGAVAFLDTPRLTGTTMKCAKSETSPMAYVVIRGTLALSGPVGTRLSE \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# rdrp DECOY\n",
    "\n",
    "Non-rdrp sequences which return a high E-value to a known and verified rdrp. These are the sequences most likely to give us trouble at scale/assembly.\n",
    "\n",
    "\n",
    "- Hepatitis C Virus `E2/NS1` Envelope gene\n",
    "\n",
    "A short region of HCV E2 (AST36587.1) consistently yields hits to the HepC RdRP (called NS5B, BAK61626)\n",
    "\n",
    "```\n",
    "Query:   AST36587.1      61      99      99      +\n",
    "Subject: rdrp3.Flaviviridae.Hepatitis_C_virus_subtype_2b:BAK61626        376     414     608\n",
    "48.7    8.7e-07 43.9    39M     HHRFNSSGCPERLASCRHLTEFAQGWGPISHANGSGPDE\n",
    "```\n",
    "\n",
    "- Retrovirus sequences that made it through\n",
    "```\n",
    "AWS06671.1\tMaize-associated retrovirus 1\tRetroviridae\tArtverviricota\n",
    "AXY66749.1\tSaesbyeol virus\tRetroviridae\tArtverviricota\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# rdrp0 BLACKLIST\n",
    "\n",
    "rdrp sequences that are mis-annotated, contain junk (plasmid/other genes) or are of low quality. Excluded from SRA search. `rdrp1` "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "-`ADG27878` :rdrp2.Caliciviridae.Calicivirus_pig_NC_WGP93C_USA_2009:ADG27878\n",
    "\n",
    "Matches calici capsid protein and related stuff\n",
    "\n",
    "> Region          1030..1441\n",
    "                     /region_name=\"RT_like\"\n",
    "                     /note=\"RT_like: Reverse transcriptase (RT, RNA-dependent\n",
    "                     DNA polymerase)_like family. An RT gene is usually\n",
    "                     indicative of a mobile element such as a retrotransposon\n",
    "                     or retrovirus. RTs occur in a variety of mobile elements,\n",
    "                     including retrotransposons; cl02808\"\n",
    "                                      \n",
    "\n",
    "- **Missing/Strange motif A**\n",
    "\n",
    ">AKJ82635.1|Influenza_A_virus_A_swine_Italy_55925_2011_H3N2_/61-60\n",
    "....................................................................................................................................\n",
    ">ANW82746.1|Makokou_virus/1-16\n",
    "...................................................WS...P..G..........................D...N.....S...A......K....F......R....LF.T.A.M\n",
    "\n",
    ">BAT24481.1|Rosellinia_necatrix_partitivirus_6/127-126\n",
    "....................................................................................................................................\n",
    ">ADP00834.1|Turnip_mosaic_virus/193-207\n",
    "I.................................................TVR...S..S..........................L...S.....P...Y......L....I......N....A......V\n",
    ">AFK23478.1|Zucchini_yellow_mosaic_virus/188-209\n",
    "L............Y.....................C....H...A..DGSXXD...S..S..........................L...T.....P...A......L....L......N....A......V\n",
    ">AKH48980.1|unidentified_Reptarenavirus/210-232\n",
    "I............V.....................L....S...X..XXXXXX...X..X..........................N...S.....P...L......Q....Y......H....LM.....F\n",
    ">AHA83412.1|Xinyi_virus/236-262\n",
    "I............N.....................V....P...F..AQTPLNGG.P..G..........................D...N.....S...A......K....F......R....RF.T.A.I\n",
    ">AJZ74605.1|Zaire_ebolavirus/227-247\n",
    "S............F..........................V...T..DLEXYX...L..X..........................F...X.....Y...X......F....T......A....P......F\n",
    ">ACL68402.1|Hepatitis_C_virus/189-210\n",
    "M............G.....................F....X...Y..DTRCFD...S..T..........................V...T.....E...R......D....I......Q....V......E\n",
    ">YP_009330274.1|Beihai_sesarmid_crab_virus_3/207-223\n",
    "K............I.....................I....F...I..SFDMSE...F..S....................................................K......K....F......P\n",
    ">YP_009337854.1|Shahe_yuevirus_like_virus_1/208-224\n",
    "K............A.....................L....F...V..SFDMSE...F..S....................................................K......K....F......P\n",
    ">YP_009329866.1|Wuhan_large_pig_roundworm_virus_1/195-216\n",
    "F............A.....................Y....C...L..DXSSFD...S..S..........................I...N.....T...W......F....I......E....R......F\n",
    ">YP_009246481.1|Tilapia_lake_virus/146-163\n",
    "L............L.....................S....S...V..VETHAR...S..V..........................L...S.................................KV.....S\n",
    ">ACY46471.1|Influenza_A_virus_A_Singapore_ON368_2009_H1N1_/218-240\n",
    "F............T.....................I....T...G..DNTXWN...E..N..........................Q...X.....P...X......M....F......L....AM.....I\n",
    ">AQY77579.1|Atypical_porcine_pestivirus/193-214\n",
    "V............A.....................V....S...F..DTKAWD...T..Q..........................V...T.....R...E......D....L......R....L......V\n",
    "\n",
    "- **Missing/Strange motif B**\n",
    "\n",
    ">AEK49762.1|Influenza_A_virus_A_gadwall_California_8504_2008_H6N1_/321-334\n",
    "L.X..X..X...X.......XXXLS.T....V.......L...G\n",
    ">BAD89093.1|Ibaraki_virus/308-326\n",
    "R.YS.T..Y...RA.K..IQHWSVI.F....M.......H...N\n",
    ">ACY46471.1|Influenza_A_virus_A_Singapore_ON368_2009_H1N1_/321-338\n",
    "L.X..P..G...RM.M..GMFXXLS.T....V.......L...G\n",
    "AFJ05062.2|Cordyline_virus_4/271-279\n",
    "A.Q..R..R...TG.................I.......T...T\n",
    ">AEM63700.1|Varroa_destructor_virus_1/274-291\n",
    "C.G..I..P...SG.S..PITDILN.X....X.......X...X\n",
    ">AHW48492.1|West_Nile_virus/266-278\n",
    "X.X..X..X...X........XXXX.X....X.......X...X\n",
    ">AKH67623.1|Nora_virus/270-287\n",
    "R.G..N..K...SG.S..YTTTIDN.X....X.......X...X\n",
    "\n",
    ">API61900.1|Quarivirus_93C4/274-291\n",
    "N.G..T..P...AG.F..VPTAENN.S....L.......Y...G\n",
    ">YP_009330019.1|Hubei_astro_like_virus/268-285\n",
    "C.G..F..N...PG.C..LCTTHDN.T....L.......V...N\n",
    ">APG79231.1|Beihai_sipunculid_worm_virus_7/319-340\n",
    "K.L..S..D...FG.Q..GLSGFCA.SAESTV.......V...E\n",
    "\n",
    "- **Missing motif C**\n",
    "\n",
    "**REFSEQ SEQUENCE**\n",
    ">NP_042695.1|Cassava_common_mosaic_virus/293-299\n",
    "Q..........................D.S.A........M.......D...A\n",
    ">AKZ17743.1|Grapevine_Syrah_virus_1/314-314\n",
    "L....................................................\n",
    ">YP_009337147.1|Changjiang_hepe_like_virus_1/325-328\n",
    "T..........V....................................L...G\n",
    ">ACY46471.1|Influenza_A_virus_A_Singapore_ON368_2009_H1N1_/358-369\n",
    "G..........L...Q....S.SD...D.Y.X........L.......I...X\n",
    ">APG77679.1|Hubei_negev_like_virus_1/316-316\n",
    "....................................................P\n",
    ">APG77656.1|Hubei_virga_like_virus_22/327-334\n",
    "S..........T...N....I.KR........................Y...L\n",
    ">AIX97862.1|Goutanap_virus/317-318\n",
    "................................................F...L\n",
    ">YP_009001772.1|Wallerfield_virus/317-318\n",
    "................................................Y...T\n",
    ">YP_009351824.1|Biratnagar_virus/317-318\n",
    "................................................Y...T\n",
    ">APG75979.1|Wenzhou_toti_like_virus_2/301-310\n",
    "....................Q.GL...Y.VRM........K.......S...V\n",
    ">AFI24669.1|Dezidougou_virus/317-318\n",
    "................................................Y...T\n",
    ">YP_009344994.1|Wuhan_insect_virus_8/315-315\n",
    "....................................................N\n",
    ">AHX42605.1|Tanay_virus/319-320\n",
    "................................................Y...T\n",
    ">YP_009333224.1|Beihai_barnacle_virus_3/341-344\n",
    "E..........L....................................V...D\n",
    ">YP_009337768.1|Hubei_virga_like_virus_7/318-319\n",
    "................................................Y...S\n",
    ">API61901.1|Biggievirus_Mos11/317-318\n",
    "................................................Y...T\n",
    ">YP_009342435.1|Wuhan_house_centipede_virus_1/315-315\n",
    "....................................................I\n",
    ">AEV46286.1|Hepatitis_C_virus_genotype_3/287-301\n",
    "F..........L...F....G.EDGSWS.Y.Y........L.......R...T\n",
    ">APG77564.1|Beihai_hepe_like_virus_11/332-335\n",
    "A..........P....................................V...G\n",
    ">ACH97717.1|Hepatitis_C_virus/287-298\n",
    "M..........L...V....C.GD...X.X.X........X.......X...X\n",
    ">APD78607.1|St_Louis_encephalitis_virus/329-340\n",
    "X..........X...X....X.XX...X.X.X........X.......X...X\n",
    ">ANA85187.1|Zika_virus/327-338\n",
    "X..........X...X....X.XX...X.X.X........X.......X...X\n",
    ">AFI24675.1|Santana_virus/318-319\n",
    "................................................Y...T"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
